
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>bullkpy.tl &#8212; BULLKpy 0.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=f539c95a"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/bullkpy.tl';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Plotting" href="pl/index.html" />
    <link rel="prev" title="bullkpy.pp" href="bullkpy.pp.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.0.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">BULLKpy 0.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="bullkpy.html">bullkpy</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.get.html">bullkpy.get</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.io.html">bullkpy.io</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.pp.html">bullkpy.pp</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">bullkpy.tl</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="pl/index.html">Plotting</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="pl/violin.html">Violin plots</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="index.html" class="nav-link">API reference</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">bullkpy.tl</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="bullkpy-tl">
<h1>bullkpy.tl<a class="headerlink" href="#bullkpy-tl" title="Link to this heading">#</a></h1>
<section id="statistics">
<h2>Statistics<a class="headerlink" href="#statistics" title="Link to this heading">#</a></h2>
<p>method=”t-test”    -&gt; Welch t-test per gene, score = t statistic
method=”wilcoxon”  -&gt; Mann–Whitney U per gene, score = z-like from U (approx)</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<p>Stores results in adata.uns[key_added] with:
- params
- names[group], scores[group], logfoldchanges[group], pvals[group], pvals_adj[group]
- mean_group[group], mean_ref[group] (extra, useful for bulk)</p>
<p>.. py:function:: adjusted_rand_index(adata, *, true_key, pred_key)
:module: bullkpy.tl</p>
<p>.. py:function:: cluster_metrics(adata, *, true_key, cluster_key=’leiden’, use_rep=’X_pca’, layer=None, n_pcs=None, silhouette_on=’rep’, metric=’euclidean’, dropna=True)
:module: bullkpy.tl</p>
<p>Compute agreement + quality metrics between known labels and clustering.</p>
<p>Returns dict with: n_used, ari, nmi, cramers_v, silhouette</p>
<p>Notes:</p>
<ul class="simple">
<li><p>ARI/NMI/Cramér’s V compare <code class="docutils literal notranslate"><span class="pre">true_key</span></code> vs <code class="docutils literal notranslate"><span class="pre">cluster_key</span></code>.</p></li>
<li><p>Silhouette measures cluster separation on chosen data:</p>
<ul>
<li><p>silhouette_on=”rep”: adata.obsm[use_rep] (recommended)</p></li>
<li><p>silhouette_on=”X”: adata.X</p></li>
<li><p>silhouette_on=”layer”: adata.layers[layer]</p></li>
</ul>
</li>
</ul>
<p>.. py:function:: leiden_resolution_scan(adata, *, true_key, resolutions=None, base_key=’leiden’, store_key=’leiden_scan’, use_rep=’X_pca’, n_pcs=20, n_neighbors=15, metric=’euclidean’, recompute_neighbors=False, inplace=True, random_state=0)
:module: bullkpy.tl</p>
<p>Scan multiple Leiden resolutions and score vs a ground-truth annotation.
Scan many Leiden resolutions and score against a ‘true’ categorical label
using ARI / NMI / Cramér’s V.</p>
<p>.. rubric:: Notes</p>
<ul class="simple">
<li><p>Uses bk.tl.neighbors() to build adata.obsp[‘connectivities’] if missing.</p></li>
<li><p>Uses bk.tl.cluster(method=’leiden’) to compute clusters for each resolution.</p></li>
<li><p>Computes ARI, NMI (if sklearn available) + Cramér’s V (always).</p></li>
</ul>
<p>.. py:function:: score_genes(adata, genes, *, score_name=’score’, layer=None, gene_pool=None, ctrl_size=50, n_bins=25, random_state=0, scale=False)
:module: bullkpy.tl</p>
<p>Scanpy-like signature scoring (bulk-friendly).</p>
<p>Score per sample:
mean(expression[genes]) - mean(expression[control_genes])</p>
<p>Control genes are sampled from gene_pool (default: all genes) by matching
expression bins of signature genes (similar to scanpy.tl.score_genes).</p>
<p>Stores result in: adata.obs[score_name]</p>
<p>.. py:function:: score_genes_dict(adata, gene_sets, *, layer=None, prefix=’’, suffix=’_score’, gene_pool=None, ctrl_size=50, n_bins=25, random_state=0, scale=False)
:module: bullkpy.tl</p>
<p>Compute multiple signature scores and store each in adata.obs.</p>
<p>Output column name:
f”{prefix}{name}{suffix}”</p>
<p>.. py:function:: score_genes_cell_cycle(adata, *, s_genes, g2m_genes, layer=None, gene_pool=None, ctrl_size=50, n_bins=25, random_state=0, scale=False, s_score=’S_score’, g2m_score=’G2M_score’, phase=’phase’)
:module: bullkpy.tl</p>
<p>Score cell cycle phases similar to scanpy.tl.score_genes_cell_cycle.</p>
<p>Writes to:
- adata.obs[s_score]   (float)
- adata.obs[g2m_score] (float)
- adata.obs[phase]     (category: “S”, “G2M”, “G1”)</p>
<p>Phase calling rule (Scanpy-like):
- if S_score &gt; G2M_score and S_score &gt; 0 -&gt; “S”
- elif G2M_score &gt; S_score and G2M_score &gt; 0 -&gt; “G2M”
- else -&gt; “G1”</p>
<p>.. py:function:: top_gene_gene_correlations(adata, *, genes=None, layer=’log1p_cpm’, method=’pearson’, top_n=200, min_abs_r=None, use_abs=True, batch_key=None, batch_mode=’none’, covariates=None)
:module: bullkpy.tl</p>
<p>Find strongest gene–gene correlations among a gene set (or all genes if provided).
Returns long table with (gene1, gene2, r, pval, qval, n).</p>
<p>.. py:function:: top_gene_obs_correlations(adata, *, gene, obs=None, obs_keys=None, layer=’log1p_cpm’, method=’pearson’, top_n=50, min_abs_r=None, use_abs=True, batch_key=None, batch_mode=’none’, covariates=None)
:module: bullkpy.tl</p>
<p>Correlate one or more genes with numeric obs columns.</p>
<ul class="simple">
<li><p>obs: restrict to these obs columns</p></li>
<li><p>obs_keys: alternative explicit numeric obs selection</p></li>
</ul>
<p>.. py:function:: top_obs_obs_correlations(adata, *, focus, against=None, obs_keys=None, method=’pearson’, top_n=100, min_abs_r=None, use_abs=True, batch_key=None, batch_mode=’none’, covariates=None)
:module: bullkpy.tl</p>
<p>Strongest correlations between numeric obs columns.
If against is None: compare focus vs all other numeric obs.
Returns long table of pairs.</p>
<p>.. py:function:: gene_gene_correlations(adata, *, gene, genes=None, layer=’log1p_cpm’, method=’pearson’, top_n=50, min_abs_r=None, use_abs=True, batch_key=None, batch_mode=’none’, covariates=None)
:module: bullkpy.tl</p>
<p>Correlate one gene with all other genes or a subset.
Returns table: (gene, r, pval, qval, n).</p>
<p>.. py:function:: obs_obs_corr_matrix(adata, *, focus, against=None, obs_keys=None, method=’pearson’, drop_na=True, batch_key=None, batch_mode=’none’, covariates=None, return_p=False, return_q=False)
:module: bullkpy.tl</p>
<p>Correlation matrix between numeric obs columns (focus × against).
Returns wide matrix (rows=focus, cols=against).
Optionally returns p-values and q-values matrices.</p>
<p>.. py:function:: partial_corr(adata, *, x, y, covariates, method=’pearson’)
:module: bullkpy.tl</p>
<p>Partial correlation: residualize x and y on covariates (obs columns) then correlate.</p>
<p>.. py:function:: plot_corr_scatter(adata, *, x, y, hue=None, kind=’obs_obs’, layer=’log1p_cpm’, method=’pearson’, batch_key=None, batch_mode=’none’, covariates=None, figsize=(5.5, 4.5), s=18.0, alpha=0.8, cmap=’viridis’, palette=’tab20’, title=None, show=True)
:module: bullkpy.tl</p>
<p>Scatter plot + correlation in title.
kind=”obs_obs”: x,y are obs numeric columns
kind=”gene_obs”: x is gene, y is obs numeric (or vice versa)</p>
<p>.. py:function:: plot_corr_heatmap(mat, *, figsize=(7.5, 4.5), cmap=’vlag’, center=0.0, vmin=None, vmax=None, annot=False, fmt=’.2f’, title=None, show=True)
:module: bullkpy.tl</p>
<p>Heatmap for a wide correlation matrix (e.g., from obs_obs_corr_matrix).</p>
<p>.. py:function:: gene_categorical_association(adata, *, groupby, genes=None, layer=’log1p_cpm’, method=’kruskal’, effect_size=’epsilon2’, min_group_size=2, adjust=’fdr_bh’)
:module: bullkpy.tl</p>
<p>A) Association between gene expression (numeric) and a categorical obs column.</p>
<p>Returns tidy df with columns:
groupby, gene, statistic, pval, qval, effect, n_groups, n, means (optional columns)</p>
<p>.. py:function:: obs_categorical_association(adata, *, groupby, obs_keys=None, method=’kruskal’, effect_size=’epsilon2’, min_group_size=2, adjust=’fdr_bh’)
:module: bullkpy.tl</p>
<p>B) Association between numeric obs columns and a categorical obs column.</p>
<p>.. py:function:: association(adata, *, x, y, layer=’log1p_cpm’, method=’auto’)
:module: bullkpy.tl</p>
<p>Unified association dispatcher.
- x gene, y categorical -&gt; gene_categorical_association (single-gene)
- x numeric obs, y categorical -&gt; obs_categorical_association (single-obs)
- x categorical, y categorical -&gt; categorical_association</p>
<p>.. py:function:: pairwise_posthoc(df, *, group_col=’grp’, value_col=’y’, method=’mwu’, correction=’bh’, dropna=True)
:module: bullkpy.tl</p>
<p>Pairwise post-hoc tests between groups.</p>
<p>:Parameters: * <strong>df</strong> (<em>DataFrame with columns [group_col, value_col]</em>)
* <strong>method</strong> (<em>“mwu” (Mann-Whitney U, two-sided) or “ttest” (Welch t-test)</em>)
* <strong>correction</strong> (<em>currently only “bh” (Benjamini-Hochberg)</em>)
* <strong>dropna</strong> (<em>drop rows with NA in group/value</em>)</p>
<p>:returns: <em>DataFrame with columns</em> – group1, group2, n1, n2, pval, qval, effect, delta_mean, delta_median</p>
<p>.. py:function:: cat_cat_association(adata, *, key1, key2, adjust=’none’)
:module: bullkpy.tl</p>
<p>C) Association between two categorical obs columns.</p>
<p>:returns: key1, key2, chi2, dof, pval, qval, cramers_v, n</p>
<p>.. py:function:: rank_genes_categorical(adata, *, groupby, group=None, reference=’rest’, layer=None, genes=None, method=’mwu’, store_key=None)
:module: bullkpy.tl</p>
<p>Scanpy-like (but bulk-friendly) rank of genes associated with a categorical obs.</p>
<p>Two-group comparison:
- MWU (default): effect_size = rank-biserial correlation
- t-test: effect_size = Cohen’s d (approx)
Multi-group comparison:
- Kruskal or ANOVA: effect_size = (rough) eta^2</p>
<p>Returns a DataFrame with:
gene, pval, qval, effect_size, mean_group, mean_ref, log2FC</p>
<p>.. py:function:: posthoc_per_gene(adata, *, genes, groupby, layer=None, method=’mwu’)
:module: bullkpy.tl</p>
<p>For each gene, run pairwise posthoc tests across all categories in <code class="docutils literal notranslate"><span class="pre">groupby</span></code>.
Returns dict: gene -&gt; DataFrame (pairwise results).</p>
<p>Requires bk.tl.pairwise_posthoc(df, method=…) to exist.</p>
<p>.. py:function:: gsea_preranked(adata=None, *, res, gene_sets=’MSigDB_Hallmark_2020’, comparison=None, store_key=’gsea’, results_key=’results’, gene_col=’gene’, score_col=’t’, ascending=False, min_size=5, max_size=1000, permutation_num=1000, seed=6, processes=4, outdir=’gsea’, save_rnk=True, return_pre_res=True, add_comparison_column=True)
:module: bullkpy.tl</p>
<p>Run GSEApy prerank and return a tidy results table.</p>
</section>
<section id="key-improvements">
<h2>Key improvements<a class="headerlink" href="#key-improvements" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Writes to outdir/<comparison>/ (folder created with parents=True)</p></li>
<li><p>Adds a ‘comparison’ column to the returned results</p></li>
<li><p>gene_sets can be:</p>
<ul class="simple">
<li><p>Enrichr library name, e.g. “MSigDB_Hallmark_2020”</p></li>
<li><p>.gmt path</p></li>
<li><p>dict of gene sets</p></li>
<li><p>convenience aliases: “hallmark(s)”, “c2”/”curated”</p></li>
<li><p>list of libraries/paths (mix allowed)</p></li>
</ul>
</li>
<li><p>Optionally returns the raw gseapy ‘pre_res’ object so you can plot later.</p></li>
</ol>
<p>.. rubric:: Notes</p>
<ul class="simple">
<li><p>The returned ‘pre_res’ is NOT stored in adata.uns by default, because it is
not h5ad-serializable. Store tables only.</p></li>
</ul>
<p>.. py:function:: list_enrichr_libraries()
:module: bullkpy.tl</p>
<p>Return available Enrichr libraries as known by gseapy.
Useful to discover exact names like ‘MSigDB_Hallmark_2020’.</p>
<p>.. rubric:: Notes</p>
<ul class="simple">
<li><p>Requires internet access (Enrichr).</p></li>
</ul>
</section>
<p>.. py:module:: bullkpy.tl</p>
<p>.. py:function:: pca(adata, *, layer=’log1p_cpm’, n_comps=50, center=True, scale=False, use_highly_variable=False, key_added=’pca’)
:module: bullkpy.tl</p>
<p>.. py:function:: neighbors(adata, *, n_neighbors=15, n_pcs=20, use_rep=’X_pca’, metric=’euclidean’, key_added=’neighbors’)
:module: bullkpy.tl</p>
<p>Compute kNN graph on samples (obs) using PCA representation.</p>
<p>Stores:
- adata.obsp[“distances”] (CSR sparse)
- adata.obsp[“connectivities”] (CSR sparse; Gaussian kernel)
- adata.uns[key_added] with parameters</p>
<p>.. py:function:: cluster(adata, *, method=’leiden’, key_added=’clusters’, resolution=1.0, n_clusters=8, use_rep=’X_pca’, n_pcs=20, random_state=0)
:module: bullkpy.tl</p>
<p>Cluster samples.</p>
<ul class="simple">
<li><p>Leiden: uses <code class="docutils literal notranslate"><span class="pre">adata.obsp['connectivities']</span></code> (requires igraph + leidenalg).</p></li>
<li><p>KMeans fallback: clusters in PCA space (no extra deps).</p></li>
</ul>
<p>.. py:function:: umap(adata, *, n_neighbors=15, n_pcs=20, use_rep=’X_pca’, min_dist=0.5, spread=1.0, n_components=2, metric=’euclidean’, random_state=0, init=’spectral’)
:module: bullkpy.tl</p>
<p>Compute UMAP embedding from a representation (default: PCA).</p>
<p>This mirrors Scanpy practice: UMAP is computed from <code class="docutils literal notranslate"><span class="pre">X_pca</span></code> with
<code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code>/<code class="docutils literal notranslate"><span class="pre">min_dist</span></code>, consistent with the neighbors graph settings.</p>
<p>Stores:
- adata.obsm[‘X_umap’]
- adata.uns[‘umap’]</p>
<p>.. py:function:: umap_graph(adata, *, graph_key=’connectivities’, use_rep=’X_pca’, n_pcs=20, min_dist=0.5, spread=1.0, n_components=2, random_state=0, init=’spectral’, negative_sample_rate=5, n_epochs=None)
:module: bullkpy.tl</p>
<p>Compute UMAP embedding strictly from a precomputed neighbor graph.</p>
<p>Requires:
- adata.obsp[graph_key]  (typically ‘connectivities’ from bk.tl.neighbors)</p>
<p>Uses:
- adata.obsm[use_rep] only for initialization (spectral/random), not for graph construction.</p>
<p>Stores:
- adata.obsm[‘X_umap_graph’]
- adata.uns[‘umap_graph’]</p>
<p>.. py:function:: de(adata, *, groupby, group, reference, method=’welch_ttest’, layer_counts=’counts’, layer_expr=’log1p_cpm’, key_added=’de’, alpha=0.05, shrink_dispersion=True, prior_df=10.0)
:module: bullkpy.tl</p>
<p>Two-group differential expression.</p>
<p>Stores results in:
adata.uns[key_added][f”{groupby}:_{group}<em>vs</em>{reference}”] = DataFrame-like dict</p>
<p>.. method:: welch_ttest:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>  Fast, exploratory; runs Welch t-test on `layer_expr` (default log1p_cpm).
</pre></div>
</div>
<p>.. method:: nb_glm:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>  Proper bulk model; NB GLM on raw counts (layer_counts) with log(libsize) offset.
  Requires statsmodels.
</pre></div>
</div>
<p>.. py:function:: de_glm(adata, *, formula, contrast=None, layer_counts=’counts’, shrink_dispersion=True, prior_df=10.0, key_added=’de_glm’, add_intercept=True)
:module: bullkpy.tl</p>
<p>NB-GLM DE with design matrix from a formula and a contrast.</p>
<p>.. rubric:: Example</p>
<p>bk.tl.de_glm(adata, formula=”~ Subtype + Batch”, contrast=(“Subtype”,”Basal”,”Luminal”))</p>
<p>Stores:
adata.uns[key_added][contrast_name] = dict(method/params/results)</p>
<p>.. py:function:: pca_loadings(adata, *, key=’pca’, loadings_key=’PCs’, pcs=None, n_top=50, use_abs=False, include_negative=True, dropna=True, gene_col=’gene’, store_key=’pca_loadings’, export=None, export_sep=’\t’, export_gmt=None, min_abs_loading=None)
:module: bullkpy.tl</p>
<p>Rank genes by PCA loadings per PC and (optionally) export for enrichment.</p>
<p>Assumes PCA results like:
- adata.varm[loadings_key] == array (n_vars x n_comps)
- adata.uns[key] contains PCA params (optional)</p>
<p>:Parameters: * <strong>pcs</strong> – PCs to process, 1-based indexing (e.g. [1,2,3]). If None -&gt; all available.
* <strong>n_top</strong> – Number of top genes to return per PC (and per sign if include_negative=True and use_abs=False).
* <strong>use_abs</strong> – If True: rank by absolute loading (single list per PC).
If False: returns positive top list (and negative list if include_negative=True).
* <strong>include_negative</strong> – Only used when use_abs=False. If True returns both pos/neg lists.
* <strong>export</strong> – If provided, writes a long-format table (PC, sign, gene, loading, rank).
Also writes a wide-format file alongside: “<stem>.wide<suffix>”.
* <strong>export_gmt</strong> –</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>              If provided, writes GMT gene sets:
                PC1_pos, PC1_neg, PC2_pos, ...
              If use_abs=True: PC1_abs, PC2_abs, ...
            * **min_abs_loading** -- Optional filter: only keep genes with abs(loading) &gt;= threshold.
</pre></div>
</div>
<p>:returns: * <em>dict mapping keys like “PC1_pos”, “PC1_neg”, “PC1_abs” -&gt; DataFrame.</em>
* <em>Also stores results in adata.uns[store_key].</em></p>
<p>.. py:function:: rank_genes_groups(adata, *, groupby, groups=None, reference=’rest’, layer=’log1p_cpm’, method=’t-test’, corr_method=’benjamini-hochberg’, use_abs=False, n_genes=100, key_added=’rank_genes_groups’)
:module: bullkpy.tl</p>
<p>Scanpy-like ranking of genes per group.</p>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="bullkpy.pp.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">bullkpy.pp</p>
      </div>
    </a>
    <a class="right-next"
       href="pl/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Plotting</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistics">Statistics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-improvements">Key improvements</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Marcos Malumbres.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>