
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>bullkpy.pl &#8212; BULLKpy 0.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=f539c95a"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/bullkpy.pl';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="bullkpy.get" href="bullkpy.get.html" />
    <link rel="prev" title="bullkpy.tl" href="bullkpy.tl.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.0.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">BULLKpy 0.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="bullkpy.html">bullkpy</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.pp.html">bullkpy.pp</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.tl.html">bullkpy.tl</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">bullkpy.pl</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.get.html">bullkpy.get</a></li>
<li class="toctree-l1"><a class="reference internal" href="bullkpy.io.html">bullkpy.io</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="index.html" class="nav-link">API reference</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">bullkpy.pl</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="bullkpy-pl">
<h1>bullkpy.pl<a class="headerlink" href="#bullkpy-pl" title="Link to this heading">#</a></h1>
<p>.. py:module:: bullkpy.pl</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
.. py:function:: qc_metrics(adata, *, color=&#39;pct_counts_mt&#39;, vars_to_plot=(&#39;total_counts&#39;, &#39;n_genes_detected&#39;, &#39;pct_counts_mt&#39;, &#39;pct_counts_ribo&#39;), log1p_total_counts=True, log1p_n_genes=False, point_size=20.0, alpha=0.8, figsize=(10, 7), save=None, show=True)
   :module: bullkpy.pl

   Plot bulk RNA-seq QC metrics (robust to missing columns).
   :rtype: :sphinx_autodoc_typehints_type:`\:py\:class\:\`tuple\`\\ \\\[\:py\:class\:\`\~matplotlib.figure.Figure\`\, \:py\:class\:\`\~numpy.ndarray\`\]`


   - If total_counts + n_genes_detected exist: scatter (library size vs detected genes)
   - Otherwise: skip scatter and show histograms only


.. py:function:: library_size_vs_genes(adata, *, x=&#39;total_counts&#39;, y=&#39;n_genes_detected&#39;, groupby=None, min_counts=None, max_counts=None, min_genes=None, max_genes=None, logx=True, logy=True, s=18.0, alpha=0.85, linewidth=0.25, edgecolor=&#39;0.15&#39;, show_outliers=True, outlier_color=&#39;crimson&#39;, outlier_marker=&#39;x&#39;, outlier_size=28.0, title=None, xlabel=None, ylabel=None, figsize=(5.5, 4.5), legend=True, legend_loc=&#39;best&#39;, save=None, show=True)
   :module: bullkpy.pl

   QC scatter: library size vs detected genes with optional thresholds.

   :rtype: :sphinx_autodoc_typehints_type:`\:py\:class\:\`tuple\`\\ \\\[\:py\:class\:\`\~matplotlib.figure.Figure\`\, \:py\:class\:\`\~matplotlib.axes.\_axes.Axes\`\]`

   Requires columns in adata.obs:
     - x (default: total_counts)
     - y (default: n_genes_detected)

   Typical usage:
     bk.pl.library_size_vs_genes(adata, min_counts=1e6, min_genes=12000, groupby=&quot;Subtype&quot;)


.. py:function:: mt_fraction_vs_counts(adata, *, x=&#39;total_counts&#39;, y=&#39;pct_counts_mt&#39;, groupby=None, min_counts=None, max_counts=None, min_mt=None, max_mt=None, logx=True, logy=False, figsize=(5.5, 4.5), save=None, show=True)
   :module: bullkpy.pl

   Scatter QC: total counts vs mitochondrial fraction (with thresholds).


.. py:function:: genes_vs_mt_fraction(adata, *, x=&#39;pct_counts_mt&#39;, y=&#39;n_genes_detected&#39;, groupby=None, min_mt=None, max_mt=None, min_genes=None, max_genes=None, logx=False, logy=True, figsize=(5.5, 4.5), save=None, show=True)
   :module: bullkpy.pl

   Scatter QC: mt fraction vs detected genes (with thresholds).


.. py:function:: qc_scatter_panel(adata, *, groupby=None, min_counts=None, max_counts=None, min_genes=None, max_genes=None, min_mt=None, max_mt=None, total_counts_key=&#39;total_counts&#39;, n_genes_key=&#39;n_genes_detected&#39;, pct_mt_key=&#39;pct_counts_mt&#39;, log_counts=True, log_genes=True, figsize=(16.0, 4.6), save=None, show=True, share_legend=True)
   :module: bullkpy.pl

   1×3 QC panel: (genes vs counts) | (mt% vs counts) | (genes vs mt%).


.. py:function:: set_style(figsize=None)
   :module: bullkpy.pl

   Apply BULLKpy global plotting style (matplotlib + seaborn).

   This is intentionally lightweight and can be called repeatedly.


   :rtype: :sphinx_autodoc_typehints_type:`\:py\:obj\:\`None\``


.. py:function:: get_palette(n=None, palette=None)
   :module: bullkpy.pl

   Return a seaborn palette (list of RGB tuples).
   If n is None, returns the default palette object.


.. py:function:: qc_pairplot(adata, *, keys=(&#39;total_counts&#39;, &#39;n_genes_detected&#39;, &#39;pct_counts_mt&#39;), color=&#39;pct_counts_mt&#39;, log1p=(&#39;total_counts&#39;,), point_size=14.0, alpha=0.7, figsize=(8, 8), save=None, show=True)
   :module: bullkpy.pl

   Scatter-matrix (pairplot) of QC metrics in `adata.obs`.

   Diagonal: histograms
   Off-diagonal: scatter plots


.. py:function:: qc_by_group(adata, *, groupby, keys=(&#39;total_counts&#39;, &#39;n_genes_detected&#39;, &#39;pct_counts_mt&#39;, &#39;pct_counts_ribo&#39;), kind=&#39;violin&#39;, log1p=(&#39;total_counts&#39;,), figsize=(11, 4), rotate_xticks=45, save=None, show=True, show_n=True)
   :module: bullkpy.pl

   Plot QC metrics grouped by a metadata column in `adata.obs` (e.g. batch, cohort).


.. py:function:: pca_scatter(adata, *, basis=&#39;X_pca&#39;, components=(1, 2), color=None, layer=&#39;log1p_cpm&#39;, point_size=20.0, alpha=0.85, figsize=(6.5, 5.0), title=None, palette=&#39;Set1&#39;, cmap=&#39;viridis&#39;, highlight=None, grey_color=&#39;#D3D3D3&#39;, key=&#39;pca&#39;, save=None, show=True)
   :module: bullkpy.pl

   Scanpy-like PCA scatter plot.

   - color can be:
       - None
       - obs column (categorical or numeric)
       - gene name (continuous)
       - list of any of the above -&gt; multiple panels in one row
   - highlight (categoricals): show only selected classes in color, all others grey.


.. py:function:: umap(adata, *, basis=&#39;X_umap&#39;, color=None, layer=&#39;log1p_cpm&#39;, point_size=25.0, alpha=0.85, figsize=(6.5, 5.5), title=None, palette=&#39;Set1&#39;, cmap=&#39;viridis&#39;, highlight=None, grey_color=&#39;#D3D3D3&#39;, save=None, show=True)
   :module: bullkpy.pl

   Scanpy-like UMAP plotting.

   - color can be:
       - None
       - obs column (categorical or numeric)
       - gene name (continuous, from `layer`)
       - list of any of the above -&gt; multiple panels in one row
   - highlight (categoricals): show only selected classes in color, all others grey.


.. py:function:: volcano(res, *, gene_col=&#39;gene&#39;, fc_col=&#39;log2FC&#39;, p_col=&#39;pval&#39;, q_col=&#39;qval&#39;, use_qval=True, p_cutoff=0.05, fc_cutoff=0.0, title=None, figsize=(6.5, 5.5), alpha=0.85, point_size=18.0, color_ns=&#39;#BDBDBD&#39;, color_up=&#39;#D62728&#39;, color_down=&#39;#1F77B4&#39;, label_genes=(), top_n_labels=10, bottom_n_labels=10, label_fontsize=None, label_offset=(0.02, 0.02), show_thresholds=True, save=None, show=True)
   :module: bullkpy.pl

   Volcano plot for DE results.

   Coloring:
     - non-significant: grey
     - significant &amp; log2FC &gt; 0: red
     - significant &amp; log2FC &lt; 0: blue

   Labeling:
     - label_genes: explicit list
     - top_n_labels: top upregulated significant genes
     - bottom_n_labels: top downregulated significant genes (most negative log2FC)


.. py:function:: rankplot(adata=None, *, res=None, uns_key=None, contrast=None, n_genes=20, sort_by=&#39;qval&#39;, direction=&#39;both&#39;, fc_col=None, figsize=(7, 6), title=None, save=None, show=True)
   :module: bullkpy.pl

   Ranked horizontal barplot of top DE genes (up/down).

   - Upregulated bars: red
   - Downregulated bars: blue
   - sort_by supports: qval, pval, log2FC
   - For direction=&quot;both&quot;: top shows strongest upregulated; bottom shows strongest downregulated (last).


.. py:function:: ma(*, result, mean_col=&#39;mean_norm&#39;, fc_col=&#39;log2FC&#39;, gene_col=&#39;gene&#39;, pval_col=&#39;pval&#39;, qval_col=&#39;qval&#39;, alpha=0.05, use_qval=True, min_abs_fc=0.5, show_fc_lines=True, fc_line_kwargs=None, point_size=10.0, point_alpha=0.6, label_genes=(), top_n_labels=10, bottom_n_labels=10, label_only_large_fc=True, title=&#39;MA plot&#39;, xlabel=None, ylabel=None, figsize=(8.5, 5.5), save=None, show=True)
   :module: bullkpy.pl

   MA plot for DE results.

   Colors:
     - non-significant: grey
     - significant up (fc &gt;= +min_abs_fc): red
     - significant down (fc &lt;= -min_abs_fc): blue

   Adds automatic FC cutoff lines at ±min_abs_fc (optional).

   :type result: :sphinx_autodoc_typehints_type:`\:py\:class\:\`\~pandas.core.frame.DataFrame\``
   :param result: DataFrame with at least [mean_col, fc_col] and (gene_col recommended).
   :type alpha: :sphinx_autodoc_typehints_type:`\:py\:class\:\`float\``
   :param alpha: Significance threshold on q-value or p-value (see use_qval).
   :type min_abs_fc: :sphinx_autodoc_typehints_type:`\:py\:class\:\`float\``
   :param min_abs_fc: Minimum absolute fold-change (in the same units as fc_col, typically log2FC)
                      to consider a gene biologically relevant and to color as up/down.
   :type show_fc_lines: :sphinx_autodoc_typehints_type:`\:py\:class\:\`bool\``
   :param show_fc_lines: Draw horizontal lines at ±min_abs_fc.
   :type label_genes: :sphinx_autodoc_typehints_type:`\:py\:class\:\`\~typing.Sequence\`\\ \\\[\:py\:class\:\`str\`\]`
   :param label_genes: Explicit genes to label (if present).
   :param top_n_labels / bottom_n_labels: Additionally label top/bottom genes by fold-change among &quot;relevant&quot; hits
                                          (significant + abs(fc) &gt;= min_abs_fc if label_only_large_fc=True).


.. py:function:: dotplot(adata, *, var_names=None, var_groups=None, groupby=&#39;leiden&#39;, layer=&#39;log1p_cpm&#39;, fraction_layer=&#39;counts&#39;, expr_threshold=0.0, standard_scale=None, swap_axes=False, row_spacing=0.7, dendrogram_top=False, dendrogram_rows=False, row_dendrogram_position=&#39;right&#39;, cluster_rows=None, cluster_cols=None, cmap=&#39;Reds&#39;, vmin=None, vmax=None, dot_min=0.0, dot_max=1.0, gamma=0.5, smallest_dot=12.0, largest_dot=260.0, figsize=None, invert_yaxis=True, title=None, size_title=&#39;Fraction of samples\\nin group (%)&#39;, colorbar_title=&#39;Mean expression\\nin group&#39;, save=None, show=True)
   :module: bullkpy.pl


.. py:function:: heatmap_de(adata, *, contrast, de_key=&#39;de&#39;, results_key=&#39;results&#39;, layer=&#39;log1p_cpm&#39;, groupby=None, groups=None, top_n=50, mode=&#39;both&#39;, sort_by=&#39;qval&#39;, z_score=&#39;row&#39;, clip_z=3.0, cmap=&#39;vlag&#39;, center=0.0, col_colors=None, dendrogram_rows=True, dendrogram_cols=True, show_sample_labels=False, figsize=None, cbar_label=&#39;z-scored expression&#39;, save=None, show=True)
   :module: bullkpy.pl

   Heatmap of top DE genes using results stored at:
       adata.uns[de_key][contrast][results_key]

   Expected DE columns:
       [&#39;gene&#39;, &#39;log2FC&#39;, &#39;t&#39;, &#39;pval&#39;, &#39;qval&#39;, &#39;mean_group&#39;, &#39;mean_ref&#39;]

   - Heatmap values come from `layer` (default: log1p_cpm)
   - Gene selection uses DE table (default sort: qval)
   - Optionally subset/order samples by `groupby` and `groups`


.. py:function:: violin(adata, *, keys, groupby, layer=&#39;log1p_cpm&#39;, figsize=(8, 4), panel_size=None, show_points=True, point_size=2.0, point_alpha=0.35, palette=None, order=None, rotate_xticks=45, inner=&#39;quartile&#39;, cut=0.0, save=None, show=True)
   :module: bullkpy.pl

   Violin plots for:
     - adata.obs columns (QC, clinical, signatures)
     - and/or gene expression (genes in adata.var_names)

   If a key is not found in adata.obs but is found in adata.var_names,
   it will be plotted as gene expression from `layer` (or adata.X if layer is None).


.. py:function:: pca_variance_ratio(adata, *, key=&#39;pca&#39;, n_comps=None, cumulative=True, figsize=(6.5, 4.5), save=None, show=True)
   :module: bullkpy.pl

   Scree plot of PCA variance ratio (+ optional cumulative curve).
   Requires adata.uns[key][&#39;variance_ratio&#39;].


.. py:function:: corr_heatmap(adata, *, layer=&#39;log1p_cpm&#39;, method=&#39;pearson&#39;, use=&#39;samples&#39;, groupby=None, groups=None, col_colors=None, cmap=&#39;vlag&#39;, center=0.0, vmin=None, vmax=None, figsize=None, show_labels=False, dendrogram=True, save=None, show=True)
   :module: bullkpy.pl

   Correlation heatmap for sample QC (or gene-gene if use=&quot;genes&quot;).

   :type use: :sphinx_autodoc_typehints_type:`\:py\:data\:\`\~typing.Literal\`\\ \\\[\`\`\&#39;samples\&#39;\`\`\, \`\`\&#39;genes\&#39;\`\`\]`
   :param use: &quot;samples&quot; -&gt; correlation between samples (obs x obs) [default, QC]
               &quot;genes&quot;   -&gt; correlation between genes (var x var)
   :param groupby / groups: If provided, subset and order samples by adata.obs[groupby].
   :type col_colors: :sphinx_autodoc_typehints_type:`\:py\:data\:\`\~typing.Union\`\\ \\\[\:py\:class\:\`str\`\, \:py\:class\:\`\~typing.Sequence\`\\ \\\[\:py\:class\:\`str\`\]\, \:py\:obj\:\`None\`\]`
   :param col_colors: obs key(s) used to annotate columns (and rows, since it&#39;s symmetric) when plotting samples.
                      Values are mapped to colors automatically.
   :type dendrogram: :sphinx_autodoc_typehints_type:`\:py\:class\:\`bool\``
   :param dendrogram: If True, uses seaborn clustermap. If False, uses heatmap without clustering.


.. py:function:: pca_loadings_bar(adata, *, pc=1, n_top=15, loadings_key=&#39;PCs&#39;, use_abs=False, show_negative=True, gene_symbol_key=None, figsize=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Plot top PCA loadings for a single PC.
   :rtype: :sphinx_autodoc_typehints_type:`\:py\:class\:\`tuple\`\\ \\\[\:py\:class\:\`\~matplotlib.figure.Figure\`\, \:py\:class\:\`\~matplotlib.axes.\_axes.Axes\`\]`


   - If use_abs=True: shows top |loading| (all positive bars).
   - Else: shows top positive and (optionally) top negative loadings.


.. py:function:: pca_loadings_heatmap(adata, *, pcs=(1, 2, 3), n_top=15, loadings_key=&#39;PCs&#39;, use_abs=False, show_negative=True, gene_symbol_key=None, z_score=False, cluster_genes=True, cluster_pcs=False, cmap=&#39;vlag&#39;, center=0.0, figsize=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Heatmap of PCA loadings for union of top genes across selected PCs.

   - Selects top positive and (optional) top negative genes for each PC.
   - Builds a matrix [genes x PCs] of loadings.
   - Optional clustering and z-scoring.


.. py:function:: gene_plot(adata, *, gene, groupby, layer=&#39;log1p_cpm&#39;, groups=None, order=None, kind=&#39;both&#39;, show_points=True, point_size=3.0, point_alpha=0.7, test=&#39;auto&#39;, compare=None, add_pvalue=True, show_n=True, ylabel=None, title=None, figsize=None, save=None, show=True)
   :module: bullkpy.pl

   Plot one gene across groups (bulk): violin/box + optional p-value.

   - If compare=[&#39;A&#39;,&#39;B&#39;] -&gt; pairwise test + bracket annotation.
   - Else -&gt; overall test across all groups (ANOVA or Kruskal).

   :rtype: fig, ax, stats_dict


.. py:function:: rank_genes_groups(adata, *, key=&#39;rank_genes_groups&#39;, groups=None, n_genes=10, sort_by=&#39;scores&#39;, show=True, save=None, figsize=None, title=None)
   :module: bullkpy.pl

   Scanpy-like quick view: show top genes per group as a compact table plot.


   :rtype: :sphinx_autodoc_typehints_type:`\:py\:class\:\`tuple\`\\ \\\[\:py\:class\:\`\~matplotlib.figure.Figure\`\, \:py\:class\:\`\~matplotlib.axes.\_axes.Axes\`\]`


.. py:function:: rank_genes_groups_dotplot(adata, *, groupby, groups=None, key=&#39;rank_genes_groups&#39;, n_genes=5, sort_by=&#39;scores&#39;, unique=True, use_abs=False, values_to_plot=&#39;expression&#39;, layer=&#39;log1p_cpm&#39;, fraction_layer=&#39;counts&#39;, expr_threshold=0.0, min_in_group_fraction=None, max_in_group_fraction=None, standard_scale=&#39;auto&#39;, swap_axes=True, dendrogram_top=True, dendrogram_rows=False, row_dendrogram_position=&#39;right&#39;, row_spacing=0.75, cmap=&#39;Reds&#39;, save=None, show=True)
   :module: bullkpy.pl

   Scanpy-like dotplot of ranked genes per group.

   Enhancements:
     1) values_to_plot:
        - &quot;expression&quot;      -&gt; dot color = mean expression per group (your dotplot default)
        - &quot;logfoldchanges&quot;  -&gt; dot color = log2FC from rank_genes_groups
     2) min/max fraction filters (computed from fraction_layer + expr_threshold)
     3) standard_scale=&quot;auto&quot; -&gt; uses &quot;var&quot; for expression mode (Scanpy-like)


.. py:function:: sample_distances(adata, *, layer=&#39;log1p_cpm&#39;, metric=&#39;euclidean&#39;, method=&#39;average&#39;, use=&#39;samples&#39;, col_colors=None, palette=&#39;tab20&#39;, z_score=False, figsize=None, show_labels=False, save=None, show=True)
   :module: bullkpy.pl

   Sample (or gene) distance clustergram.

   - Computes pairwise distances (pdist) on X (samples x genes).
   - Uses seaborn.clustermap with hierarchical clustering.
   - Optionally annotate samples with metadata columns via col_colors.

   .. rubric:: Notes

   - For sample QC, use metric=&quot;correlation&quot; (distance = 1-corr) often works well.
   - z_score=True will z-score genes across samples before distance computation.


.. py:function:: sample_correlation_clustergram(adata, *, layer=&#39;log1p_cpm&#39;, method=&#39;pearson&#39;, linkage_method=&#39;average&#39;, col_colors=None, palette=&#39;tab20&#39;, figsize=None, show_labels=False, save=None, show=True)
   :module: bullkpy.pl

   Sample correlation clustergram (more interpretable than raw distances for bulk QC).
   Displays correlation, clusters by (1 - correlation).

   Heatmap values: correlation in [-1, 1]
   Clustering: on distance = 1 - correlation


.. py:function:: get_categorical_colors(adata, *, key, where=&#39;obs&#39;, categories=None, palette=None, force=False)
   :module: bullkpy.pl

   Stable mapping category -&gt; hex color for a categorical column.
   Cached in `bk.settings.categorical_colors[(where, key)]`.

   :rtype: :sphinx_autodoc_typehints_type:`\:py\:class\:\`dict\`\\ \\\[\:py\:class\:\`str\`\, \:py\:class\:\`str\`\]`

   If you set a per-column palette override via:
     settings.categorical_palette_overrides[(where, key)] = &quot;Set2&quot;
   it will be used automatically.


.. py:function:: categorical_colors_array(adata, *, key, where=&#39;obs&#39;, categories=None, palette=None, force=False)
   :module: bullkpy.pl

   Series of hex colors aligned to obs_names / var_names.


   :rtype: :sphinx_autodoc_typehints_type:`\:py\:class\:\`\~pandas.core.series.Series\``


.. py:function:: ari_resolution_heatmap(adata, *, df=None, store_key=&#39;leiden_scan&#39;, metric=&#39;ARI&#39;, show_n_clusters=True, cmap=&#39;viridis&#39;, vmin=None, vmax=None, figsize=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Heatmap-like summary of clustering quality vs Leiden resolution.

   Expects `df` with at least:
     - &#39;resolution&#39;
     - metric column: &#39;ARI&#39; or &#39;NMI&#39; or &#39;cramers_v&#39;
   Optional:
     - &#39;n_clusters&#39; (for a second row annotation)

   .. rubric:: Example

   df = bk.tl.leiden_resolution_scan(...)
   bk.pl.ari_resolution_heatmap(adata, df=df, metric=&quot;ARI&quot;)


.. py:function:: corrplot_obs(adata, *, x, y, color=None, hue=None, layer=None, palette=&#39;tab20&#39;, cmap=&#39;viridis&#39;, legend=True, method=&#39;both&#39;, add_regline=True, annotate=True, dropna=True, point_size=18.0, alpha=0.75, figsize=(5.5, 4.5), panel_size=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Scatter + correlations between two quantitative obs columns.

   Multi-panel:
     color=[&quot;DLL3&quot;,&quot;SOX10&quot;] makes one panel per color key in a single row.


.. py:function:: association_heatmap(df, *, index, columns, values, agg=&#39;mean&#39;, cmap=&#39;viridis&#39;, vmin=None, vmax=None, figsize=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Generic heatmap for association outputs.
   Example: genes x categories using effect or -log10(qval).


.. py:function:: boxplot_with_stats(adata, *, y, groupby, figsize=(7, 3.5), kind=&#39;violin&#39;, show_points=True, point_size=2.0, point_alpha=0.3, title=None, save=None, show=True)
   :module: bullkpy.pl

   Simple scanpy-like box/violin plot annotated with global p-value (Kruskal for &gt;2 groups, MWU for 2).


.. py:function:: categorical_confusion(adata, *, key1, key2, normalize=&#39;row&#39;, cmap=&#39;Blues&#39;, figsize=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Confusion-style heatmap for two categorical obs columns + Cramér’s V / ARI / NMI.


.. py:function:: rankplot_association(*, res, gene_col=&#39;gene&#39;, effect_col=&#39;effect&#39;, sort_by=&#39;qval&#39;, direction=&#39;both&#39;, n_items=20, figsize=(7, 6), title=None, save=None, show=True)
   :module: bullkpy.pl

   Ranked barplot for categorical association results.

   Behavior mirrors bk.pl.rankplot():
     • Upregulated / positive = red
     • Downregulated / negative = blue
     • Strongest up at top
     • Most downregulated LAST


.. py:function:: dotplot_association(df, *, feature_col, groupby_col=&#39;groupby&#39;, effect_col=&#39;effect&#39;, q_col=&#39;qval&#39;, top_n=50, figsize=None, cmap=&#39;RdBu_r&#39;, vmin=None, vmax=None, size_min=10.0, size_max=250.0, title=None, save=None, show=True)
   :module: bullkpy.pl

   Scanpy-like dotplot for association results across multiple contrasts/groupby runs.
     - dot color = effect
     - dot size  = -log10(qval)
   Works when df contains multiple groupby/contrasts (groupby_col), otherwise still works.


.. py:function:: heatmap_association(df, *, feature_col, groupby_col=&#39;groupby&#39;, value_col=&#39;effect&#39;, top_n=60, cmap=&#39;RdBu_r&#39;, center=0.0, figsize=None, title=None, save=None, show=True)
   :module: bullkpy.pl

   Heatmap of association values (effect by default), selecting top_n rows by best qval per column.


.. py:function:: gene_association(adata, *, gene, groupby, layer=&#39;log1p_cpm&#39;, kind=&#39;violin&#39;, order=None, rotate_xticklabels=45, figsize=None, panel_size=(4.2, 3.2), show_points=True, point_size=2.0, point_alpha=0.35, palette=&#39;Set2&#39;, annotate_posthoc=True, posthoc_method=&#39;mwu&#39;, posthoc_alpha=0.05, max_brackets=6, bracket_height=0.06, save=None, show=True)
   :module: bullkpy.pl

   Gene vs categorical obs association plot (Scanpy-like panels).

   - gene can be a string or list of genes -&gt; row of panels
   - violin/box + optional strip points
   - optional automatic pairwise post-hoc + significance brackets (BH corrected)

   Returns (fig, axes).


.. py:function:: volcano_categorical(results, *, x=&#39;effect_size&#39;, q_col=&#39;qval&#39;, label_top=12, q_thr=0.05, effect_thr=None, figsize=(6.8, 5.2), title=None, save=None, show=True)
   :module: bullkpy.pl

   Volcano plot for categorical associations.
   Expects columns: gene, qval, and x (effect_size or log2FC).


.. py:function:: gene_association_volcano(adata, *, groupby, group, reference=None, layer=&#39;log1p_cpm&#39;, genes=None, method=&#39;mwu&#39;, effect=&#39;delta_mean&#39;, alpha=0.05, top_labels=12, figsize=(6.5, 5.5), title=None, save=None, show=True)
   :module: bullkpy.pl

   Effect-size aware volcano for categorical association:
     x = effect size (delta_mean or delta_median between group vs reference/rest)
     y = -log10(q)

   This is NOT a full DE method — it&#39;s a fast association scan.


.. py:function:: gsea_leading_edge_heatmap(adata, pre_res, *, term_idx=None, terms=None, layer=&#39;log1p_cpm&#39;, use=&#39;samples&#39;, groupby=None, min_gene_freq=1, max_genes=200, z_score=&#39;row&#39;, clip_z=3.0, row_cluster=True, col_cluster=True, cmap=&#39;vlag&#39;, figsize=None, show_labels=False, gene_label_fontsize=7.0, save=None, show=True)
   :module: bullkpy.pl

   Heatmap of expression for leading-edge genes from selected pathways.

   - Builds a union of leading-edge genes across selected terms
   - Optionally keeps only genes that appear in &gt;= min_gene_freq pathways
   - Plots expression across:
       * samples (use=&quot;samples&quot;), OR
       * group means (use=&quot;group_mean&quot;, requires groupby)

   This helps you see whether the shared leading-edge genes form coherent
   expression patterns across samples/subtypes.


.. py:function:: leading_edge_jaccard_heatmap(pre_res, *, term_idx=None, terms=None, min_shared_genes=0, row_cluster=True, col_cluster=True, cmap=&#39;viridis&#39;, vmin=0.0, vmax=1.0, figsize=None, show_labels=True, label_fontsize=9.0, save=None, show=True)
   :module: bullkpy.pl

   Jaccard similarity between leading-edge gene sets for each pathway.

   Jaccard(A,B) = |A ∩ B| / |A ∪ B|

   Useful to see clusters of pathways driven by similar leading-edge genes.


.. py:function:: leading_edge_overlap_matrix(pre_res, *, term_idx=None, terms=None, min_gene_freq=2, sort_genes_by=&#39;freq&#39;, row_cluster=True, col_cluster=False, cmap=&#39;Greys&#39;, figsize=None, show_gene_labels=True, gene_label_fontsize=8.0, show_term_labels=True, save=None, show=True)
   :module: bullkpy.pl

   Pathway × gene binary matrix for leading-edge membership.

   This is the closest to the &quot;Leading Edge Viewer&quot; idea:
   you can see if a small set of genes drives many enriched pathways.


.. py:function:: oncoprint(adata, *, mut_cols, expr_genes=None, layer=&#39;log1p_cpm&#39;, sort_genes=True, sort_samples=&#39;mut_first&#39;, drop_all_wt=True, max_samples=None, groupby=None, group_order=None, group_blocks=True, within_group_sort=&#39;mut_first&#39;, show_sample_labels=False, mut_color=&#39;#222222&#39;, wt_color=&#39;#FFFFFF&#39;, grid_color=&#39;0.85&#39;, expr_cmap=&#39;viridis&#39;, expr_vmin=None, expr_vmax=None, expr_zscore=False, cell_size=None, row_height=0.35, expr_row_height=0.25, top_annotation_height=0.35, title=None, save_dpi=None, max_pixels=60000, save=None, show=True)
   :module: bullkpy.pl

   Binary oncoprint (mut vs wt) from adata.obs 0/1 columns.
   Implements mutation-first sample ordering and optional group blocks.


.. py:function:: gsea_bubbleplot(df_gsea, *, pathways, comparison_col=&#39;comparison&#39;, term_col=&#39;Term&#39;, nes_col=&#39;NES&#39;, fdr_col=&#39;FDR q-val&#39;, comparison_order=None, drop_empty_comparisons=True, size_from=&#39;fdr&#39;, min_q=1e-300, size_min=10.0, size_max=350.0, fdr_floor=1e-50, size_clip_quantile=0.99, cmap=&#39;RdBu_r&#39;, center=0.0, vmin=None, vmax=None, figsize=None, row_spacing=1.0, col_spacing=1.0, row_height=0.32, col_width=0.32, dot_edgecolor=&#39;0.15&#39;, dot_linewidth=0.35, show_grid=False, group_label_rotation=90, xtick_rotation=90, title=None, save=None, show=True)
   :module: bullkpy.pl

   Bubble plot matrix for GSEA results.

   Rows: comparisons (contrasts)
   Cols: pathways (terms)
   Color: NES (diverging, centered at `center`)
   Size: -log10(FDR q-val) with floor &amp; optional clipping

   `pathways` can be:
     - dict: {&quot;Immune&quot;: [term1, term2], &quot;Metabolism&quot;: [term3]}
     - list: [term1, term2, ...]

</pre></div>
</div>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="bullkpy.tl.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">bullkpy.tl</p>
      </div>
    </a>
    <a class="right-next"
       href="bullkpy.get.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">bullkpy.get</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/api/bullkpy.pl.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Marcos Malumbres.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>