
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Negative binomial GLM differential expression &#8212; BULLKpy 0.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=f539c95a"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/tl/de_glm';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Gene-Category association" href="gene_categorical_association.html" />
    <link rel="prev" title="Differential expression" href="de.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.0.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">BULLKpy 0.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../bullkpy.html">bullkpy</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../io/index.html">Input Output</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../io/read_counts.html">Read counts</a></li>



<li class="toctree-l2"><a class="reference internal" href="../io/add_metadata.html">Add metadata</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../pp/index.html">Pre-processing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../pp/batch_correct_combat.html">Batch correction with Combat</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/filter_genes.html">Filter genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/filter_samples.html">Filter samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/find_bad_obs_cols_by_write.html">Find bad Obs columns</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/highly_variable_genes.html">Highly variable genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/log1p.html">log1p</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/make_obs_h5ad_safe.html">Make Obs .h5ad safe</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/normalize_cpm.html">Normalize cpm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/qc_metrics.html">QC metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/sanitize_metadata.html">Sanitize metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/set_raw_counts.html">Set raw counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pp/_is_integerish.html">Is integerish</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">Tools</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="adjusted_rand_index.html">Adjusted rand index</a></li>
<li class="toctree-l2"><a class="reference internal" href="association.html">Association</a></li>
<li class="toctree-l2"><a class="reference internal" href="categorical_association.html">Categorical association</a></li>
<li class="toctree-l2"><a class="reference internal" href="cluster.html">Clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="cluster_metrics.html">Cluster metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="de.html">Differential expression</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Negative binomial GLM differential expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_categorical_association.html">Gene-Category association</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_gene_correlations.html">Gene-gene correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="gsea_preranked.html">GSEA preranked</a></li>
<li class="toctree-l2"><a class="reference internal" href="leiden_resolution_scan.html">Leiden resolution scan</a></li>
<li class="toctree-l2"><a class="reference internal" href="list_enrichr_libraries.html">List enrichr libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="neighbors.html">Neighbors</a></li>
<li class="toctree-l2"><a class="reference internal" href="obs_categorical_association.html">Obs-Category association</a></li>
<li class="toctree-l2"><a class="reference internal" href="obs_obs_corr_matrix.html">Obs-Obs correlation matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_posthoc.html">Pairwise postdoc</a></li>
<li class="toctree-l2"><a class="reference internal" href="partial_corr.html">Partial correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="pca.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="pca_loadings.html">PCA loadings</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot_corr_heatmap.html">Plot Correlation heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot_corr_scatter.html">Plot Correlation scatter</a></li>
<li class="toctree-l2"><a class="reference internal" href="posthoc_per_gene.html">Posthoc per gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="rank_genes_categorical.html">Rank genes categorical</a></li>
<li class="toctree-l2"><a class="reference internal" href="rank_genes_groups.html">Rank genes groups</a></li>
<li class="toctree-l2"><a class="reference internal" href="score_genes.html">Score genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="score_genes_cell_cycle.html">Score cell cycle genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="score_genes_dict.html">Score genes from a dictionary of genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="top_gene_gene_correlations.html">Top gene-gene correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="top_gene_obs_correlations.html">Top Gene-Obs correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="top_obs_obs_correlations.html">Top Obs-Obs correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="umap.html">UMAP embedding</a></li>
<li class="toctree-l2"><a class="reference internal" href="umap_graph.html">UMAP graph</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../pl/index.html">Plotting</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../pl/ari_resolution_heatmap.html">ARI resolution heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/association_heatmap.html">Association heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/boxplot_with_stats.html">Boxplot with statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/categorical_confusion.html">Categorical confusion</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/corr_heatmap.html">Correlation heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/corrplot_obs.html">Correlation plot Observations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/dotplot.html">Dotplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/dotplot_association.html">Dotplot association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gene_association.html">Gene Association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gene_association_volcano.html">Gene Association Volcano</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gene_plot.html">Gene Plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gsea_bubbleplot.html">GSEA bubbleplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gsea_leading_edge_heatmap.html">GSEA Leading edge heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/heatmap_association.html">Heatmap association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/heatmap_de.html">Heatmap Differential Expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/leading_edge_jaccard_heatmap.html">Leading edge Jaccard heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/leading_edge_overlap_matrix.html">Leading edge overlap matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/library_size_vs_genes.html">Library size versus genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/ma.html">MA plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/oncoprint.html">oncoprint</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/pca_loadings_bar.html">PCA loadings bar plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/pca_loadings_heatmap.html">PCA loadings heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/rankplot.html">Rankplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/rankplot_association.html">Rankplot association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/violin.html">Violin plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/volcano.html">Volcano plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/volcano_categorical.html">Volcano for categorial comparisons</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">API reference</a></li>
    
    
    <li class="breadcrumb-item"><a href="index.html" class="nav-link">Tools</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Negative binomial GLM differential expression</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="negative-binomial-glm-differential-expression">
<h1>Negative binomial GLM differential expression<a class="headerlink" href="#negative-binomial-glm-differential-expression" title="Link to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="bullkpy.tl.de_glm">
<span class="sig-prename descclassname"><span class="pre">bullkpy.tl.</span></span><span class="sig-name descname"><span class="pre">de_glm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formula</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contrast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shrink_dispersion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prior_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'de_glm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_intercept</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/bullkpy/tl/de_glm.html#de_glm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bullkpy.tl.de_glm" title="Link to this definition">#</a></dt>
<dd><p>NB-GLM DE with design matrix from a formula and a contrast.</p>
<p class="rubric">Example</p>
<p>bk.tl.de_glm(adata, formula=”~ Subtype + Batch”, contrast=(“Subtype”,”Basal”,”Luminal”))</p>
<dl class="simple">
<dt>Stores:</dt><dd><p>adata.uns[key_added][contrast_name] = dict(method/params/results)</p>
</dd>
</dl>
</dd></dl>

<p>Run <strong>negative binomial GLM differential expression</strong> (bulk-friendly) using a <strong>formula-defined design matrix</strong> and a <strong>two-level contrast</strong>. This is intended for raw (integer-like) counts and supports adding covariates (e.g., batch) in the model.</p>
<p>The implementation is “DESeq2-like” in spirit:</p>
<ul class="simple">
<li><p>it computes <strong>size factors</strong> and uses a <strong>log(size factor)</strong> offset</p></li>
<li><p>estimates <strong>gene-wise dispersion</strong> from normalized counts</p></li>
<li><p>optionally <strong>shrinks dispersion</strong> toward a trend</p></li>
<li><p>fits a <strong>gene-wise NB-GLM</strong> and performs a <strong>Wald test</strong> for a specified contrast</p></li>
<li><p>reports log2FC, Wald z, p-values, and BH FDR q-values</p></li>
</ul>
<section id="when-to-use">
<h2>When to use<a class="headerlink" href="#when-to-use" title="Link to this heading">#</a></h2>
<p>Use de_glm when you want:</p>
<ul class="simple">
<li><p>two-group DE while controlling for additional covariates (e.g., batch, sex, purity)</p></li>
<li><p>a count model (NB) instead of a t-test on log-expression</p></li>
<li><p>a workflow similar to bulk RNA-seq DE tools (DESeq2/edgeR style).</p></li>
</ul>
<p>You should provide raw counts in adata.layers[“counts”] (or another layer via layer_counts).</p>
</section>
<section id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Link to this heading">#</a></h2>
<section id="core-inputs">
<h3>Core inputs<a class="headerlink" href="#core-inputs" title="Link to this heading">#</a></h3>
<p><strong>adata</strong><br />
AnnData with samples in rows (n_obs) and genes in columns (n_vars).</p>
<p><strong>formula</strong><br />
A Patsy-like formula string defining the design matrix using adata.obs columns.
Examples:</p>
<ul class="simple">
<li><p>“~ Subtype”</p></li>
<li><p>“~ Subtype + Batch”</p></li>
<li><p>“~ Subtype + Batch + Sex”.
The design matrix is constructed via an internal helper (_design_matrix_from_formula).</p></li>
</ul>
<p><strong>contrast</strong><br />
Required in the current implementation.<br />
A tuple (var, level_a, level_b) meaning level_a − level_b for a categorical variable.<br />
Example:</p>
<ul class="simple">
<li><p>(“Subtype”, “Basal”, “Luminal”) → Basal vs Luminal (Basal − Luminal)
Internally a contrast vector c is built from the coefficient names (coef_names) using _contrast_vector.</p></li>
</ul>
<p><strong>layer_counts</strong><br />
Layer name containing the raw counts matrix (samples × genes).<br />
Default: “counts”.</p>
</section>
<section id="dispersion-options">
<h3>Dispersion options<a class="headerlink" href="#dispersion-options" title="Link to this heading">#</a></h3>
<p><strong>shrink_dispersion</strong><br />
If True, shrink gene-wise dispersion estimates toward a trend (stabilizes estimates when n is small).<br />
Default: True.</p>
<p><strong>prior_df</strong><br />
Prior degrees of freedom controlling the strength of dispersion shrinkage (higher → stronger shrinkage).<br />
Default: 10.0.</p>
</section>
<section id="output-storage">
<h3>Output storage<a class="headerlink" href="#output-storage" title="Link to this heading">#</a></h3>
<p><strong>key_added</strong><br />
Where results are stored in adata.uns.<br />
Default: “de_glm”.</p>
<p><strong>add_intercept</strong><br />
Whether to include an intercept term in the design matrix.<br />
Default: True.</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<p>Results are stored as:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">key_added</span><span class="p">][</span><span class="n">contrast_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;formula&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;contrast&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;coef_names&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;layer_counts&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;shrink_dispersion&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;prior_df&quot;</span><span class="p">:</span> <span class="o">...</span><span class="p">,</span>
  <span class="s2">&quot;results&quot;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">&gt;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Where:</p>
<ul class="simple">
<li><p>contrast_name is constructed as:</p>
<ul>
<li><p>f”{var}:{level_a}<em>vs</em>{level_b}”</p></li>
<li><p>Example: “Subtype:Basal_vs_Luminal”</p></li>
</ul>
</li>
</ul>
<p><strong>Result table columns</strong>.</p>
<p>The stored DataFrame (results) contains:</p>
<ul class="simple">
<li><p>gene : gene name (from adata.var_names)</p></li>
<li><p>log2FC : estimated log2 fold-change for the contrast (level_a − level_b)</p></li>
<li><p>wald_z : Wald z-statistic for the contrast</p></li>
<li><p>pval : two-sided p-value from Wald z</p></li>
<li><p>qval : BH-adjusted p-value (FDR)</p></li>
<li><p>dispersion : final dispersion used per gene (after optional shrink)</p></li>
<li><p>mean_norm : mean of normalized counts per gene (used in dispersion estimation).</p></li>
</ul>
<p>The table is sorted by qval ascending.</p>
</section>
<section id="statistical-details-high-level">
<h2>Statistical details (high level)<a class="headerlink" href="#statistical-details-high-level" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Design matrix.
Built from adata.obs using the provided formula (plus optional intercept).</p></li>
<li><p>Counts and offsets.</p></li>
</ol>
<ul class="simple">
<li><p>counts: Y = adata.layers[layer_counts]</p></li>
<li><p>size factors: sf = deseq2_size_factors(Y)</p></li>
<li><p>offset: log(sf) included in the GLM.</p></li>
</ul>
<ol class="arabic simple" start="3">
<li><p>Dispersion.</p></li>
</ol>
<ul class="simple">
<li><p>estimates dispersion from normalized counts (Y / sf)</p></li>
<li><p>optional shrinkage to a dispersion trend.</p></li>
</ul>
<ol class="arabic simple" start="4">
<li><p>Model fitting (per gene).
Fits:</p></li>
</ol>
<ul class="simple">
<li><p>NB(mu, alpha) with log link</p></li>
<li><p>log(mu) = Xβ + offset.</p></li>
</ul>
<ol class="arabic simple" start="5">
<li><p>Wald test for contrast
Tests cᵀβ = 0 using:</p></li>
</ol>
<ul class="simple">
<li><p>z = (cᵀβ) / sqrt(cᵀCov(β)c)</p></li>
<li><p>p-value from standard normal approximation</p></li>
</ul>
</section>
<section id="requirements-dependencies">
<h2>Requirements / dependencies<a class="headerlink" href="#requirements-dependencies" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>statsmodels is required:</p>
<ul>
<li><p>installed via pip install statsmodels</p></li>
</ul>
</li>
<li><p>scipy is used for the normal CDF in Wald p-values.</p></li>
</ul>
<p>If statsmodels is missing, the function raises an ImportError with install instructions.</p>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">#</a></h2>
<p>Basic DE with batch covariate</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bk</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">de_glm</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">formula</span><span class="o">=</span><span class="s2">&quot;~ Subtype + Batch&quot;</span><span class="p">,</span>
    <span class="n">contrast</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;Subtype&quot;</span><span class="p">,</span> <span class="s2">&quot;Basal&quot;</span><span class="p">,</span> <span class="s2">&quot;Luminal&quot;</span><span class="p">),</span>
    <span class="n">layer_counts</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Retrieve results</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">res</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;de_glm&quot;</span><span class="p">][</span><span class="s2">&quot;Subtype:Basal_vs_Luminal&quot;</span><span class="p">][</span><span class="s2">&quot;results&quot;</span><span class="p">]</span>
<span class="n">res</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<p>Turn off dispersion shrinkage</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bk</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">de_glm</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">formula</span><span class="o">=</span><span class="s2">&quot;~ Subtype + Batch&quot;</span><span class="p">,</span>
    <span class="n">contrast</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;Subtype&quot;</span><span class="p">,</span> <span class="s2">&quot;Basal&quot;</span><span class="p">,</span> <span class="s2">&quot;Luminal&quot;</span><span class="p">),</span>
    <span class="n">shrink_dispersion</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="notes-caveats">
<h2>Notes / caveats<a class="headerlink" href="#notes-caveats" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>The current implementation requires contrast (no “test all coefficients” mode yet).</p></li>
<li><p>This is a gene-wise GLM loop; it can be slower than vectorized approaches for very large gene counts.</p></li>
<li><p>Counts with total sum 0 for a gene are skipped (remain default p=1 / NaNs for some stats).</p></li>
<li><p>The contrast mechanism assumes a fairly standard dummy encoding for categorical variables (handled by the internal design/contrast helpers). If coefficient naming differs from expectations, _contrast_vector may fail or target the wrong coefficient—double-check coef_names stored in adata.uns.</p></li>
</ul>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="de.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Differential expression</p>
      </div>
    </a>
    <a class="right-next"
       href="gene_categorical_association.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Gene-Category association</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bullkpy.tl.de_glm"><code class="docutils literal notranslate"><span class="pre">de_glm()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#when-to-use">When to use</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#parameters">Parameters</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#core-inputs">Core inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dispersion-options">Dispersion options</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#output-storage">Output storage</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-details-high-level">Statistical details (high level)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#requirements-dependencies">Requirements / dependencies</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#examples">Examples</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#notes-caveats">Notes / caveats</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2026, Marcos Malumbres.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>