{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5a0f9a13-58f9-4746-93ce-4ed86242acdc",
   "metadata": {},
   "source": [
    "# TCGA data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c527d45c-8431-4f27-8a2d-5af68525b966",
   "metadata": {},
   "source": [
    "@mmm, December 24, 2025"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57c233a7-5da9-43cc-956c-0f8519b10a6f",
   "metadata": {},
   "source": [
    "Data and information of TCGA data.  \n",
    "Select RNAseq data and generate complete metadata.  \n",
    "See also previous metadata 2021 in \n",
    "- /Users/mmalumbres/Desktop/BioDATA/TCGA/210226 TCGA_metadata_analysis.ipynb\n",
    "- /Users/mmalumbres/Desktop/BioDATA/TCGA/210226 TCGA_metadata.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6b075459-4b45-4e00-9559-00750cb0e1d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DESKTOP = '/Users/mmalumbres/Desktop/'\n",
    "TCGA = \"/Users/mmalumbres/Desktop/BioDATA/TCGA/\"\n",
    "XENA = \"/Users/mmalumbres/Desktop/BioDATA/TCGA/Xena/\"\n",
    "PANCANATLAS = '/Users/mmalumbres/Desktop/BioDATA/TCGA/GDC TCGA PanCanAtlas/'\n",
    "PANCAN = '/Users/mmalumbres/Desktop/BioDATA/TCGA/GDC Pan-Cancer (PANCAN)/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "57fc758b-1043-40b5-84d4-6cea63a3ed26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eea9a39-904d-48e7-8345-750c05e82082",
   "metadata": {},
   "source": [
    "data obtained from:  \n",
    "- PanCanAtlas: https://gdc.cancer.gov/about-data/publications/pancanatlas\n",
    "- GDC TCGA PANCAN from Xena: https://xenabrowser.net/datapages/?cohort=GDC%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa793a6-dae1-464b-8e37-9cd93cf265ed",
   "metadata": {},
   "source": [
    "# Summary of data  \n",
    "\n",
    "### 1. RNAseq\n",
    "- 1.1. `xena`: XENA: RSEM_Hugo_norm_count: 58581 genes (GeneSymbol) x 10536 samples (TCGA-S9-A7J2-01).\n",
    "- 1.2. `transcripts`: PANCAN: 60488 genes (Ensembl_ID) × 11058 samples (TCGA-OR-A5JP-01A).\n",
    "- 1.3. `transcripts_PanCanAtlas`: PanCanAtlas: 20531 genes (other gene_id: ?|100134869) x 11070 samples (TCGA-OR-A5J1-01A-11R-A29S-07; unusual codes).\n",
    "\n",
    "### 2. Gene based\n",
    "- 2.1. `cnv`: > 2 million data multiple per gene and sample. DO NOT USE\n",
    "- 2.2. `masked_cnv` > 2 million data multiple per gene and sample. DO NOT USE\n",
    "- 2.1. `maf` (PANCANATLAS + mc3.v0.2.8.PUBLIC.maf.gz): > 3 millones de mutaciones. DO NOT USE\n",
    "\n",
    "### 3. Tumor data (e.g. TCGA-OR-A5J1-01)\n",
    "- 3.1. `basic_phenotype`: PANCAN. Basic clinical data per sample\n",
    "- 3.2. `survival`: PANCAN. survival\n",
    "- 3.3. `phenotype`: PANCAN phenotype\n",
    "- 3.4. `purity`: PANCANATLAS ABSOLUTE Genome purity and doublings. OK\n",
    "- 3.5. `paradigm`: PANCANATLAS PARADIGM Data interference of pathways. 19k pathways - 9830 samples. too many pathways! - CHECK!\n",
    "- 3.6. `gistic`: Copy Number per Gene: 19729 genes x 11369 samples. numeric values. OK but CHECK too large.\n",
    "- 3.7. `mc3`: Xena MC3 SNPs... all genes by sample. 40543 genes (RB1) x 9105 samples (TCGA-02-0003-01). OK. but select most frequent genes?? or use externally=\n",
    "\n",
    "### 4. Patient data (e.g. TCGA-OR-A5J1)\n",
    "- 4.1. `clinical`: PANCANATLAS PFI, PFS, OS, etc. OK\n",
    "- 4.2. `clinical_follow`: PANCANATLAS additional follow up. >10k patients. OK\n",
    "\n",
    "### 5. Others\n",
    "- 5.1. `quality`: sample quality: PANCANATLAS + \"merged_sample_quality_annotations.tsv\". multiple samples per tumor and platform. 79,286 rows. DO NOT USE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc898563-5204-4668-83aa-e45b5bcd1ece",
   "metadata": {},
   "source": [
    "# 1. Download data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bffe564a-089e-4616-842b-72807966d06f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## 1. RNAseq data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2439647-9461-4444-93a8-cda55438a3f0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 1.1. Xena"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "668b731f-a5a4-41b8-9663-eb7b005b4010",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58581, 10536)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-S9-A7J2-01</th>\n",
       "      <th>TCGA-G3-A3CH-11</th>\n",
       "      <th>TCGA-EK-A2RE-01</th>\n",
       "      <th>TCGA-44-6778-01</th>\n",
       "      <th>TCGA-F4-6854-01</th>\n",
       "      <th>TCGA-AB-2863-03</th>\n",
       "      <th>TCGA-C8-A1HL-01</th>\n",
       "      <th>TCGA-EW-A2FS-01</th>\n",
       "      <th>TCGA-05-4420-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-DJ-A2QC-01</th>\n",
       "      <th>TCGA-A8-A09K-01</th>\n",
       "      <th>TCGA-61-1907-01</th>\n",
       "      <th>TCGA-IB-7885-01</th>\n",
       "      <th>TCGA-B6-A0IA-01</th>\n",
       "      <th>TCGA-VQ-AA6F-01</th>\n",
       "      <th>TCGA-BR-8588-01</th>\n",
       "      <th>TCGA-24-2254-01</th>\n",
       "      <th>TCGA-DD-A115-01</th>\n",
       "      <th>TCGA-FV-A3I0-11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTD-2588J6.1</td>\n",
       "      <td>0.7385</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.1763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.9652</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.1745</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RP11-433M22.1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTD-2588J6.2</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 10536 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample  TCGA-S9-A7J2-01  TCGA-G3-A3CH-11  TCGA-EK-A2RE-01  \\\n",
       "0   CTD-2588J6.1           0.7385              0.0              0.0   \n",
       "1  RP11-433M22.1           0.0000              0.0              0.0   \n",
       "2   CTD-2588J6.2           0.0000              0.0              0.0   \n",
       "\n",
       "   TCGA-44-6778-01  TCGA-F4-6854-01  TCGA-AB-2863-03  TCGA-C8-A1HL-01  \\\n",
       "0              0.0              0.0           1.1763              0.0   \n",
       "1              0.0              0.0           0.0000              0.0   \n",
       "2              0.0              0.0           0.0000              0.0   \n",
       "\n",
       "   TCGA-EW-A2FS-01  TCGA-05-4420-01  ...  TCGA-DJ-A2QC-01  TCGA-A8-A09K-01  \\\n",
       "0              0.0           0.9652  ...              0.0              0.0   \n",
       "1              0.0           0.0000  ...              0.0              0.0   \n",
       "2              0.0           0.0000  ...              0.0              0.0   \n",
       "\n",
       "   TCGA-61-1907-01  TCGA-IB-7885-01  TCGA-B6-A0IA-01  TCGA-VQ-AA6F-01  \\\n",
       "0              0.0              0.0              0.0           1.1745   \n",
       "1              0.0              0.0              0.0           0.0000   \n",
       "2              0.0              0.0              0.0           0.0000   \n",
       "\n",
       "   TCGA-BR-8588-01  TCGA-24-2254-01  TCGA-DD-A115-01  TCGA-FV-A3I0-11  \n",
       "0              0.0              0.0              0.0              0.0  \n",
       "1              0.0              0.0              0.0              0.0  \n",
       "2              0.0              0.0              0.0              0.0  \n",
       "\n",
       "[3 rows x 10536 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RNAseq data\n",
    "XENA = '/Users/mmalumbres/Desktop/BioDATA/TCGA/data/Xena/'\n",
    "xena = pd.read_csv(XENA + \"Gene expression/tcga_RSEM_Hugo_norm_count.tsv\", sep=\"\\t\")\n",
    "print(xena.shape)\n",
    "xena.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dc84a11-bb87-4e60-ac63-68279ca6d8c8",
   "metadata": {},
   "source": [
    "### 1.2. PANCAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "cd4b634f-a5d5-46e1-ac58-86b0e8043342",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl_ID</th>\n",
       "      <th>TCGA-OR-A5JP-01A</th>\n",
       "      <th>TCGA-OR-A5JG-01A</th>\n",
       "      <th>TCGA-OR-A5K1-01A</th>\n",
       "      <th>TCGA-OR-A5JR-01A</th>\n",
       "      <th>TCGA-OR-A5KU-01A</th>\n",
       "      <th>TCGA-OR-A5L9-01A</th>\n",
       "      <th>TCGA-OR-A5JQ-01A</th>\n",
       "      <th>TCGA-OR-A5K4-01A</th>\n",
       "      <th>TCGA-OR-A5JL-01A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-VD-A8KI-01A</th>\n",
       "      <th>TCGA-V4-A9E9-01A</th>\n",
       "      <th>TCGA-V4-A9F7-01A</th>\n",
       "      <th>TCGA-V4-A9EO-01A</th>\n",
       "      <th>TCGA-V4-A9EU-01A</th>\n",
       "      <th>TCGA-WC-A87T-01A</th>\n",
       "      <th>TCGA-WC-AA9A-01A</th>\n",
       "      <th>TCGA-V4-A9EA-01A</th>\n",
       "      <th>TCGA-RZ-AB0B-01A</th>\n",
       "      <th>TCGA-V4-A9F8-01A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.13</td>\n",
       "      <td>10.769838</td>\n",
       "      <td>10.721099</td>\n",
       "      <td>10.253847</td>\n",
       "      <td>11.448116</td>\n",
       "      <td>10.843921</td>\n",
       "      <td>7.787903</td>\n",
       "      <td>11.971184</td>\n",
       "      <td>11.188589</td>\n",
       "      <td>11.762797</td>\n",
       "      <td>...</td>\n",
       "      <td>10.372865</td>\n",
       "      <td>10.380461</td>\n",
       "      <td>11.545930</td>\n",
       "      <td>10.696968</td>\n",
       "      <td>10.588715</td>\n",
       "      <td>11.394463</td>\n",
       "      <td>9.748193</td>\n",
       "      <td>9.533330</td>\n",
       "      <td>9.667112</td>\n",
       "      <td>8.405141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.11</td>\n",
       "      <td>11.361396</td>\n",
       "      <td>11.456868</td>\n",
       "      <td>11.002112</td>\n",
       "      <td>11.324181</td>\n",
       "      <td>10.714246</td>\n",
       "      <td>6.442943</td>\n",
       "      <td>11.385862</td>\n",
       "      <td>11.252665</td>\n",
       "      <td>10.682117</td>\n",
       "      <td>...</td>\n",
       "      <td>8.791163</td>\n",
       "      <td>8.451211</td>\n",
       "      <td>10.055282</td>\n",
       "      <td>9.879583</td>\n",
       "      <td>10.205793</td>\n",
       "      <td>10.062046</td>\n",
       "      <td>10.133142</td>\n",
       "      <td>6.629357</td>\n",
       "      <td>10.018200</td>\n",
       "      <td>7.781360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.12</td>\n",
       "      <td>9.152285</td>\n",
       "      <td>8.654636</td>\n",
       "      <td>7.781360</td>\n",
       "      <td>9.055282</td>\n",
       "      <td>7.948367</td>\n",
       "      <td>5.727920</td>\n",
       "      <td>9.057992</td>\n",
       "      <td>8.527477</td>\n",
       "      <td>9.226412</td>\n",
       "      <td>...</td>\n",
       "      <td>8.447083</td>\n",
       "      <td>7.483816</td>\n",
       "      <td>9.797662</td>\n",
       "      <td>9.525521</td>\n",
       "      <td>9.535275</td>\n",
       "      <td>9.702173</td>\n",
       "      <td>9.303781</td>\n",
       "      <td>6.475733</td>\n",
       "      <td>9.350939</td>\n",
       "      <td>7.499846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.15</td>\n",
       "      <td>7.693487</td>\n",
       "      <td>7.832890</td>\n",
       "      <td>6.918863</td>\n",
       "      <td>7.754888</td>\n",
       "      <td>6.459432</td>\n",
       "      <td>3.807355</td>\n",
       "      <td>7.592457</td>\n",
       "      <td>7.475733</td>\n",
       "      <td>6.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>6.584963</td>\n",
       "      <td>6.247928</td>\n",
       "      <td>7.912889</td>\n",
       "      <td>7.714246</td>\n",
       "      <td>7.768184</td>\n",
       "      <td>7.942515</td>\n",
       "      <td>7.787903</td>\n",
       "      <td>5.247928</td>\n",
       "      <td>7.906891</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60483</th>\n",
       "      <td>__no_feature</td>\n",
       "      <td>20.857955</td>\n",
       "      <td>20.913038</td>\n",
       "      <td>20.642255</td>\n",
       "      <td>21.594382</td>\n",
       "      <td>20.915012</td>\n",
       "      <td>20.335652</td>\n",
       "      <td>21.262439</td>\n",
       "      <td>20.937683</td>\n",
       "      <td>21.253776</td>\n",
       "      <td>...</td>\n",
       "      <td>20.078441</td>\n",
       "      <td>20.108500</td>\n",
       "      <td>21.910417</td>\n",
       "      <td>20.804356</td>\n",
       "      <td>21.859678</td>\n",
       "      <td>22.122194</td>\n",
       "      <td>21.561402</td>\n",
       "      <td>20.012948</td>\n",
       "      <td>21.125540</td>\n",
       "      <td>19.634309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60484</th>\n",
       "      <td>__ambiguous</td>\n",
       "      <td>21.196319</td>\n",
       "      <td>20.855109</td>\n",
       "      <td>20.250448</td>\n",
       "      <td>21.013543</td>\n",
       "      <td>20.282017</td>\n",
       "      <td>18.897063</td>\n",
       "      <td>21.192075</td>\n",
       "      <td>20.704917</td>\n",
       "      <td>21.144501</td>\n",
       "      <td>...</td>\n",
       "      <td>21.207991</td>\n",
       "      <td>21.127259</td>\n",
       "      <td>21.393229</td>\n",
       "      <td>21.368524</td>\n",
       "      <td>21.066775</td>\n",
       "      <td>21.344917</td>\n",
       "      <td>21.649232</td>\n",
       "      <td>20.622553</td>\n",
       "      <td>20.987240</td>\n",
       "      <td>20.894101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60485</th>\n",
       "      <td>__too_low_aQual</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60486</th>\n",
       "      <td>__not_aligned</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60487</th>\n",
       "      <td>__alignment_not_unique</td>\n",
       "      <td>23.986868</td>\n",
       "      <td>24.216696</td>\n",
       "      <td>23.260758</td>\n",
       "      <td>24.180528</td>\n",
       "      <td>23.185174</td>\n",
       "      <td>23.601889</td>\n",
       "      <td>24.098957</td>\n",
       "      <td>23.322809</td>\n",
       "      <td>23.827002</td>\n",
       "      <td>...</td>\n",
       "      <td>23.404797</td>\n",
       "      <td>23.971370</td>\n",
       "      <td>24.320447</td>\n",
       "      <td>23.739028</td>\n",
       "      <td>23.690098</td>\n",
       "      <td>24.414007</td>\n",
       "      <td>24.587663</td>\n",
       "      <td>23.902186</td>\n",
       "      <td>23.548168</td>\n",
       "      <td>23.445219</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60488 rows × 11058 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Ensembl_ID  TCGA-OR-A5JP-01A  TCGA-OR-A5JG-01A  \\\n",
       "0          ENSG00000000003.13         10.769838         10.721099   \n",
       "1           ENSG00000000005.5          2.584963          4.000000   \n",
       "2          ENSG00000000419.11         11.361396         11.456868   \n",
       "3          ENSG00000000457.12          9.152285          8.654636   \n",
       "4          ENSG00000000460.15          7.693487          7.832890   \n",
       "...                       ...               ...               ...   \n",
       "60483            __no_feature         20.857955         20.913038   \n",
       "60484             __ambiguous         21.196319         20.855109   \n",
       "60485         __too_low_aQual          0.000000          0.000000   \n",
       "60486           __not_aligned          0.000000          0.000000   \n",
       "60487  __alignment_not_unique         23.986868         24.216696   \n",
       "\n",
       "       TCGA-OR-A5K1-01A  TCGA-OR-A5JR-01A  TCGA-OR-A5KU-01A  TCGA-OR-A5L9-01A  \\\n",
       "0             10.253847         11.448116         10.843921          7.787903   \n",
       "1              1.000000          1.000000          1.000000          0.000000   \n",
       "2             11.002112         11.324181         10.714246          6.442943   \n",
       "3              7.781360          9.055282          7.948367          5.727920   \n",
       "4              6.918863          7.754888          6.459432          3.807355   \n",
       "...                 ...               ...               ...               ...   \n",
       "60483         20.642255         21.594382         20.915012         20.335652   \n",
       "60484         20.250448         21.013543         20.282017         18.897063   \n",
       "60485          0.000000          0.000000          0.000000          0.000000   \n",
       "60486          0.000000          0.000000          0.000000          0.000000   \n",
       "60487         23.260758         24.180528         23.185174         23.601889   \n",
       "\n",
       "       TCGA-OR-A5JQ-01A  TCGA-OR-A5K4-01A  TCGA-OR-A5JL-01A  ...  \\\n",
       "0             11.971184         11.188589         11.762797  ...   \n",
       "1              2.000000          1.584963          3.700440  ...   \n",
       "2             11.385862         11.252665         10.682117  ...   \n",
       "3              9.057992          8.527477          9.226412  ...   \n",
       "4              7.592457          7.475733          6.584963  ...   \n",
       "...                 ...               ...               ...  ...   \n",
       "60483         21.262439         20.937683         21.253776  ...   \n",
       "60484         21.192075         20.704917         21.144501  ...   \n",
       "60485          0.000000          0.000000          0.000000  ...   \n",
       "60486          0.000000          0.000000          0.000000  ...   \n",
       "60487         24.098957         23.322809         23.827002  ...   \n",
       "\n",
       "       TCGA-VD-A8KI-01A  TCGA-V4-A9E9-01A  TCGA-V4-A9F7-01A  TCGA-V4-A9EO-01A  \\\n",
       "0             10.372865         10.380461         11.545930         10.696968   \n",
       "1              0.000000          0.000000          0.000000          0.000000   \n",
       "2              8.791163          8.451211         10.055282          9.879583   \n",
       "3              8.447083          7.483816          9.797662          9.525521   \n",
       "4              6.584963          6.247928          7.912889          7.714246   \n",
       "...                 ...               ...               ...               ...   \n",
       "60483         20.078441         20.108500         21.910417         20.804356   \n",
       "60484         21.207991         21.127259         21.393229         21.368524   \n",
       "60485          0.000000          0.000000          0.000000          0.000000   \n",
       "60486          0.000000          0.000000          0.000000          0.000000   \n",
       "60487         23.404797         23.971370         24.320447         23.739028   \n",
       "\n",
       "       TCGA-V4-A9EU-01A  TCGA-WC-A87T-01A  TCGA-WC-AA9A-01A  TCGA-V4-A9EA-01A  \\\n",
       "0             10.588715         11.394463          9.748193          9.533330   \n",
       "1              0.000000          0.000000          0.000000          0.000000   \n",
       "2             10.205793         10.062046         10.133142          6.629357   \n",
       "3              9.535275          9.702173          9.303781          6.475733   \n",
       "4              7.768184          7.942515          7.787903          5.247928   \n",
       "...                 ...               ...               ...               ...   \n",
       "60483         21.859678         22.122194         21.561402         20.012948   \n",
       "60484         21.066775         21.344917         21.649232         20.622553   \n",
       "60485          0.000000          0.000000          0.000000          0.000000   \n",
       "60486          0.000000          0.000000          0.000000          0.000000   \n",
       "60487         23.690098         24.414007         24.587663         23.902186   \n",
       "\n",
       "       TCGA-RZ-AB0B-01A  TCGA-V4-A9F8-01A  \n",
       "0              9.667112          8.405141  \n",
       "1              0.000000          0.000000  \n",
       "2             10.018200          7.781360  \n",
       "3              9.350939          7.499846  \n",
       "4              7.906891          6.000000  \n",
       "...                 ...               ...  \n",
       "60483         21.125540         19.634309  \n",
       "60484         20.987240         20.894101  \n",
       "60485          0.000000          0.000000  \n",
       "60486          0.000000          0.000000  \n",
       "60487         23.548168         23.445219  \n",
       "\n",
       "[60488 rows x 11058 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RNAseq data\n",
    "transcripts = pd.read_csv(PANCAN + \"RNAseq/GDC-PANCAN.htseq_counts.tsv\", sep=\"\\t\")\n",
    "transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "f525112c-fdb2-40f9-bf3d-57a04722c8b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-OR-A5JP-01</th>\n",
       "      <th>TCGA-OR-A5JG-01</th>\n",
       "      <th>TCGA-OR-A5K1-01</th>\n",
       "      <th>TCGA-OR-A5JR-01</th>\n",
       "      <th>TCGA-OR-A5KU-01</th>\n",
       "      <th>TCGA-OR-A5L9-01</th>\n",
       "      <th>TCGA-OR-A5JQ-01</th>\n",
       "      <th>TCGA-OR-A5K4-01</th>\n",
       "      <th>TCGA-OR-A5JL-01</th>\n",
       "      <th>TCGA-OR-A5LS-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-VD-A8KI-01</th>\n",
       "      <th>TCGA-V4-A9E9-01</th>\n",
       "      <th>TCGA-V4-A9F7-01</th>\n",
       "      <th>TCGA-V4-A9EO-01</th>\n",
       "      <th>TCGA-V4-A9EU-01</th>\n",
       "      <th>TCGA-WC-A87T-01</th>\n",
       "      <th>TCGA-WC-AA9A-01</th>\n",
       "      <th>TCGA-V4-A9EA-01</th>\n",
       "      <th>TCGA-RZ-AB0B-01</th>\n",
       "      <th>TCGA-V4-A9F8-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>10.769838</td>\n",
       "      <td>10.721099</td>\n",
       "      <td>10.253847</td>\n",
       "      <td>11.448116</td>\n",
       "      <td>10.843921</td>\n",
       "      <td>7.787903</td>\n",
       "      <td>11.971184</td>\n",
       "      <td>11.188589</td>\n",
       "      <td>11.762797</td>\n",
       "      <td>10.932953</td>\n",
       "      <td>...</td>\n",
       "      <td>10.372865</td>\n",
       "      <td>10.380461</td>\n",
       "      <td>11.545930</td>\n",
       "      <td>10.696968</td>\n",
       "      <td>10.588715</td>\n",
       "      <td>11.394463</td>\n",
       "      <td>9.748193</td>\n",
       "      <td>9.533330</td>\n",
       "      <td>9.667112</td>\n",
       "      <td>8.405141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>2.584963</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>11.361396</td>\n",
       "      <td>11.456868</td>\n",
       "      <td>11.002112</td>\n",
       "      <td>11.324181</td>\n",
       "      <td>10.714246</td>\n",
       "      <td>6.442943</td>\n",
       "      <td>11.385862</td>\n",
       "      <td>11.252665</td>\n",
       "      <td>10.682117</td>\n",
       "      <td>10.590587</td>\n",
       "      <td>...</td>\n",
       "      <td>8.791163</td>\n",
       "      <td>8.451211</td>\n",
       "      <td>10.055282</td>\n",
       "      <td>9.879583</td>\n",
       "      <td>10.205793</td>\n",
       "      <td>10.062046</td>\n",
       "      <td>10.133142</td>\n",
       "      <td>6.629357</td>\n",
       "      <td>10.018200</td>\n",
       "      <td>7.781360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11057 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TCGA-OR-A5JP-01  TCGA-OR-A5JG-01  TCGA-OR-A5K1-01  \\\n",
       "Ensembl                                                              \n",
       "ENSG00000000003        10.769838        10.721099        10.253847   \n",
       "ENSG00000000005         2.584963         4.000000         1.000000   \n",
       "ENSG00000000419        11.361396        11.456868        11.002112   \n",
       "\n",
       "                 TCGA-OR-A5JR-01  TCGA-OR-A5KU-01  TCGA-OR-A5L9-01  \\\n",
       "Ensembl                                                              \n",
       "ENSG00000000003        11.448116        10.843921         7.787903   \n",
       "ENSG00000000005         1.000000         1.000000         0.000000   \n",
       "ENSG00000000419        11.324181        10.714246         6.442943   \n",
       "\n",
       "                 TCGA-OR-A5JQ-01  TCGA-OR-A5K4-01  TCGA-OR-A5JL-01  \\\n",
       "Ensembl                                                              \n",
       "ENSG00000000003        11.971184        11.188589        11.762797   \n",
       "ENSG00000000005         2.000000         1.584963         3.700440   \n",
       "ENSG00000000419        11.385862        11.252665        10.682117   \n",
       "\n",
       "                 TCGA-OR-A5LS-01  ...  TCGA-VD-A8KI-01  TCGA-V4-A9E9-01  \\\n",
       "Ensembl                           ...                                     \n",
       "ENSG00000000003        10.932953  ...        10.372865        10.380461   \n",
       "ENSG00000000005         4.584963  ...         0.000000         0.000000   \n",
       "ENSG00000000419        10.590587  ...         8.791163         8.451211   \n",
       "\n",
       "                 TCGA-V4-A9F7-01  TCGA-V4-A9EO-01  TCGA-V4-A9EU-01  \\\n",
       "Ensembl                                                              \n",
       "ENSG00000000003        11.545930        10.696968        10.588715   \n",
       "ENSG00000000005         0.000000         0.000000         0.000000   \n",
       "ENSG00000000419        10.055282         9.879583        10.205793   \n",
       "\n",
       "                 TCGA-WC-A87T-01  TCGA-WC-AA9A-01  TCGA-V4-A9EA-01  \\\n",
       "Ensembl                                                              \n",
       "ENSG00000000003        11.394463         9.748193         9.533330   \n",
       "ENSG00000000005         0.000000         0.000000         0.000000   \n",
       "ENSG00000000419        10.062046        10.133142         6.629357   \n",
       "\n",
       "                 TCGA-RZ-AB0B-01  TCGA-V4-A9F8-01  \n",
       "Ensembl                                            \n",
       "ENSG00000000003         9.667112         8.405141  \n",
       "ENSG00000000005         0.000000         0.000000  \n",
       "ENSG00000000419        10.018200         7.781360  \n",
       "\n",
       "[3 rows x 11057 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts.columns = transcripts.columns.str[:15]\n",
    "transcripts.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "204e269a-b3db-4c96-a8ac-d53791196650",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60488, 11057)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-OR-A5JP-01A</th>\n",
       "      <th>TCGA-OR-A5JG-01A</th>\n",
       "      <th>TCGA-OR-A5K1-01A</th>\n",
       "      <th>TCGA-OR-A5JR-01A</th>\n",
       "      <th>TCGA-OR-A5KU-01A</th>\n",
       "      <th>TCGA-OR-A5L9-01A</th>\n",
       "      <th>TCGA-OR-A5JQ-01A</th>\n",
       "      <th>TCGA-OR-A5K4-01A</th>\n",
       "      <th>TCGA-OR-A5JL-01A</th>\n",
       "      <th>TCGA-OR-A5LS-01A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-VD-A8KI-01A</th>\n",
       "      <th>TCGA-V4-A9E9-01A</th>\n",
       "      <th>TCGA-V4-A9F7-01A</th>\n",
       "      <th>TCGA-V4-A9EO-01A</th>\n",
       "      <th>TCGA-V4-A9EU-01A</th>\n",
       "      <th>TCGA-WC-A87T-01A</th>\n",
       "      <th>TCGA-WC-AA9A-01A</th>\n",
       "      <th>TCGA-V4-A9EA-01A</th>\n",
       "      <th>TCGA-RZ-AB0B-01A</th>\n",
       "      <th>TCGA-V4-A9F8-01A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>10.769838</td>\n",
       "      <td>10.721099</td>\n",
       "      <td>10.253847</td>\n",
       "      <td>11.448116</td>\n",
       "      <td>10.843921</td>\n",
       "      <td>7.787903</td>\n",
       "      <td>11.971184</td>\n",
       "      <td>11.188589</td>\n",
       "      <td>11.762797</td>\n",
       "      <td>10.932953</td>\n",
       "      <td>...</td>\n",
       "      <td>10.372865</td>\n",
       "      <td>10.380461</td>\n",
       "      <td>11.545930</td>\n",
       "      <td>10.696968</td>\n",
       "      <td>10.588715</td>\n",
       "      <td>11.394463</td>\n",
       "      <td>9.748193</td>\n",
       "      <td>9.533330</td>\n",
       "      <td>9.667112</td>\n",
       "      <td>8.405141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>2.584963</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>11.361396</td>\n",
       "      <td>11.456868</td>\n",
       "      <td>11.002112</td>\n",
       "      <td>11.324181</td>\n",
       "      <td>10.714246</td>\n",
       "      <td>6.442943</td>\n",
       "      <td>11.385862</td>\n",
       "      <td>11.252665</td>\n",
       "      <td>10.682117</td>\n",
       "      <td>10.590587</td>\n",
       "      <td>...</td>\n",
       "      <td>8.791163</td>\n",
       "      <td>8.451211</td>\n",
       "      <td>10.055282</td>\n",
       "      <td>9.879583</td>\n",
       "      <td>10.205793</td>\n",
       "      <td>10.062046</td>\n",
       "      <td>10.133142</td>\n",
       "      <td>6.629357</td>\n",
       "      <td>10.018200</td>\n",
       "      <td>7.781360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11057 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TCGA-OR-A5JP-01A  TCGA-OR-A5JG-01A  TCGA-OR-A5K1-01A  \\\n",
       "Ensembl                                                                 \n",
       "ENSG00000000003         10.769838         10.721099         10.253847   \n",
       "ENSG00000000005          2.584963          4.000000          1.000000   \n",
       "ENSG00000000419         11.361396         11.456868         11.002112   \n",
       "\n",
       "                 TCGA-OR-A5JR-01A  TCGA-OR-A5KU-01A  TCGA-OR-A5L9-01A  \\\n",
       "Ensembl                                                                 \n",
       "ENSG00000000003         11.448116         10.843921          7.787903   \n",
       "ENSG00000000005          1.000000          1.000000          0.000000   \n",
       "ENSG00000000419         11.324181         10.714246          6.442943   \n",
       "\n",
       "                 TCGA-OR-A5JQ-01A  TCGA-OR-A5K4-01A  TCGA-OR-A5JL-01A  \\\n",
       "Ensembl                                                                 \n",
       "ENSG00000000003         11.971184         11.188589         11.762797   \n",
       "ENSG00000000005          2.000000          1.584963          3.700440   \n",
       "ENSG00000000419         11.385862         11.252665         10.682117   \n",
       "\n",
       "                 TCGA-OR-A5LS-01A  ...  TCGA-VD-A8KI-01A  TCGA-V4-A9E9-01A  \\\n",
       "Ensembl                            ...                                       \n",
       "ENSG00000000003         10.932953  ...         10.372865         10.380461   \n",
       "ENSG00000000005          4.584963  ...          0.000000          0.000000   \n",
       "ENSG00000000419         10.590587  ...          8.791163          8.451211   \n",
       "\n",
       "                 TCGA-V4-A9F7-01A  TCGA-V4-A9EO-01A  TCGA-V4-A9EU-01A  \\\n",
       "Ensembl                                                                 \n",
       "ENSG00000000003         11.545930         10.696968         10.588715   \n",
       "ENSG00000000005          0.000000          0.000000          0.000000   \n",
       "ENSG00000000419         10.055282          9.879583         10.205793   \n",
       "\n",
       "                 TCGA-WC-A87T-01A  TCGA-WC-AA9A-01A  TCGA-V4-A9EA-01A  \\\n",
       "Ensembl                                                                 \n",
       "ENSG00000000003         11.394463          9.748193          9.533330   \n",
       "ENSG00000000005          0.000000          0.000000          0.000000   \n",
       "ENSG00000000419         10.062046         10.133142          6.629357   \n",
       "\n",
       "                 TCGA-RZ-AB0B-01A  TCGA-V4-A9F8-01A  \n",
       "Ensembl                                              \n",
       "ENSG00000000003          9.667112          8.405141  \n",
       "ENSG00000000005          0.000000          0.000000  \n",
       "ENSG00000000419         10.018200          7.781360  \n",
       "\n",
       "[3 rows x 11057 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts = transcripts.drop(\"Ensembl_ID\", axis=1).set_index(\"Ensembl\")\n",
    "print(transcripts.shape)\n",
    "transcripts.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcee189e-2e17-4914-9496-681e54e48280",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 1.3. PanCanAtlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ac790704-3ff5-4cf0-949f-6c3c4a20e718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20531, 11070)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TCGA-OR-A5J1-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J2-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J3-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J5-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J6-01A-31R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J7-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J8-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J9-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5JA-01A-11R-A29S-07</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-CG-4449-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4462-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4465-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4466-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4469-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4472-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4474-01A-02R-1157-13</th>\n",
       "      <th>TCGA-CG-4475-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4476-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4477-01A-01R-1157-13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>?|100130426</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?|100133144</td>\n",
       "      <td>3.2661</td>\n",
       "      <td>2.6815</td>\n",
       "      <td>1.7301</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>1.4422</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4.4556</td>\n",
       "      <td>...</td>\n",
       "      <td>4.358154</td>\n",
       "      <td>5.676995</td>\n",
       "      <td>5.219350</td>\n",
       "      <td>14.846708</td>\n",
       "      <td>20.115492</td>\n",
       "      <td>6.997533</td>\n",
       "      <td>18.311906</td>\n",
       "      <td>12.057112</td>\n",
       "      <td>18.628740</td>\n",
       "      <td>17.874417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?|100134869</td>\n",
       "      <td>3.9385</td>\n",
       "      <td>8.9948</td>\n",
       "      <td>6.5650</td>\n",
       "      <td>1.5492</td>\n",
       "      <td>4.4709</td>\n",
       "      <td>6.0529</td>\n",
       "      <td>2.2876</td>\n",
       "      <td>1.3599</td>\n",
       "      <td>5.0581</td>\n",
       "      <td>...</td>\n",
       "      <td>2.656360</td>\n",
       "      <td>3.342794</td>\n",
       "      <td>2.423442</td>\n",
       "      <td>5.055287</td>\n",
       "      <td>11.626054</td>\n",
       "      <td>13.654193</td>\n",
       "      <td>7.417109</td>\n",
       "      <td>11.585177</td>\n",
       "      <td>11.482418</td>\n",
       "      <td>14.919338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11070 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id  TCGA-OR-A5J1-01A-11R-A29S-07  TCGA-OR-A5J2-01A-11R-A29S-07  \\\n",
       "0  ?|100130426                        0.0000                        0.0000   \n",
       "1  ?|100133144                        3.2661                        2.6815   \n",
       "2  ?|100134869                        3.9385                        8.9948   \n",
       "\n",
       "   TCGA-OR-A5J3-01A-11R-A29S-07  TCGA-OR-A5J5-01A-11R-A29S-07  \\\n",
       "0                        0.0000                        0.0000   \n",
       "1                        1.7301                        0.0000   \n",
       "2                        6.5650                        1.5492   \n",
       "\n",
       "   TCGA-OR-A5J6-01A-31R-A29S-07  TCGA-OR-A5J7-01A-11R-A29S-07  \\\n",
       "0                        0.0000                        0.0000   \n",
       "1                        0.0000                        1.1673   \n",
       "2                        4.4709                        6.0529   \n",
       "\n",
       "   TCGA-OR-A5J8-01A-11R-A29S-07  TCGA-OR-A5J9-01A-11R-A29S-07  \\\n",
       "0                        0.0000                        0.0000   \n",
       "1                        1.4422                        0.0000   \n",
       "2                        2.2876                        1.3599   \n",
       "\n",
       "   TCGA-OR-A5JA-01A-11R-A29S-07  ...  TCGA-CG-4449-01A-01R-1157-13  \\\n",
       "0                        0.0000  ...                           NaN   \n",
       "1                        4.4556  ...                      4.358154   \n",
       "2                        5.0581  ...                      2.656360   \n",
       "\n",
       "   TCGA-CG-4462-01A-01R-1157-13  TCGA-CG-4465-01A-01R-1157-13  \\\n",
       "0                           NaN                           NaN   \n",
       "1                      5.676995                      5.219350   \n",
       "2                      3.342794                      2.423442   \n",
       "\n",
       "   TCGA-CG-4466-01A-01R-1157-13  TCGA-CG-4469-01A-01R-1157-13  \\\n",
       "0                           NaN                           NaN   \n",
       "1                     14.846708                     20.115492   \n",
       "2                      5.055287                     11.626054   \n",
       "\n",
       "   TCGA-CG-4472-01A-01R-1157-13  TCGA-CG-4474-01A-02R-1157-13  \\\n",
       "0                           NaN                           NaN   \n",
       "1                      6.997533                     18.311906   \n",
       "2                     13.654193                      7.417109   \n",
       "\n",
       "   TCGA-CG-4475-01A-01R-1157-13  TCGA-CG-4476-01A-01R-1157-13  \\\n",
       "0                           NaN                           NaN   \n",
       "1                     12.057112                     18.628740   \n",
       "2                     11.585177                     11.482418   \n",
       "\n",
       "   TCGA-CG-4477-01A-01R-1157-13  \n",
       "0                           NaN  \n",
       "1                     17.874417  \n",
       "2                     14.919338  \n",
       "\n",
       "[3 rows x 11070 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RNAseq data\n",
    "transcripts_PanCanAtlas = pd.read_csv(PANCANATLAS + \"EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv\", sep=\"\\t\")\n",
    "print(transcripts_PanCanAtlas.shape)\n",
    "transcripts_PanCanAtlas.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "c7564d8d-90fb-4dd3-b037-37ebcb151a08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <th>TCGA-OR-A5J6-01</th>\n",
       "      <th>TCGA-OR-A5J7-01</th>\n",
       "      <th>TCGA-OR-A5J8-01</th>\n",
       "      <th>TCGA-OR-A5J9-01</th>\n",
       "      <th>TCGA-OR-A5JA-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-CG-4449-01</th>\n",
       "      <th>TCGA-CG-4462-01</th>\n",
       "      <th>TCGA-CG-4465-01</th>\n",
       "      <th>TCGA-CG-4466-01</th>\n",
       "      <th>TCGA-CG-4469-01</th>\n",
       "      <th>TCGA-CG-4472-01</th>\n",
       "      <th>TCGA-CG-4474-01</th>\n",
       "      <th>TCGA-CG-4475-01</th>\n",
       "      <th>TCGA-CG-4476-01</th>\n",
       "      <th>TCGA-CG-4477-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>?|100130426</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?|100133144</td>\n",
       "      <td>3.2661</td>\n",
       "      <td>2.6815</td>\n",
       "      <td>1.7301</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>1.4422</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4.4556</td>\n",
       "      <td>...</td>\n",
       "      <td>4.358154</td>\n",
       "      <td>5.676995</td>\n",
       "      <td>5.219350</td>\n",
       "      <td>14.846708</td>\n",
       "      <td>20.115492</td>\n",
       "      <td>6.997533</td>\n",
       "      <td>18.311906</td>\n",
       "      <td>12.057112</td>\n",
       "      <td>18.628740</td>\n",
       "      <td>17.874417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?|100134869</td>\n",
       "      <td>3.9385</td>\n",
       "      <td>8.9948</td>\n",
       "      <td>6.5650</td>\n",
       "      <td>1.5492</td>\n",
       "      <td>4.4709</td>\n",
       "      <td>6.0529</td>\n",
       "      <td>2.2876</td>\n",
       "      <td>1.3599</td>\n",
       "      <td>5.0581</td>\n",
       "      <td>...</td>\n",
       "      <td>2.656360</td>\n",
       "      <td>3.342794</td>\n",
       "      <td>2.423442</td>\n",
       "      <td>5.055287</td>\n",
       "      <td>11.626054</td>\n",
       "      <td>13.654193</td>\n",
       "      <td>7.417109</td>\n",
       "      <td>11.585177</td>\n",
       "      <td>11.482418</td>\n",
       "      <td>14.919338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11070 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id  TCGA-OR-A5J1-01  TCGA-OR-A5J2-01  TCGA-OR-A5J3-01  \\\n",
       "0  ?|100130426           0.0000           0.0000           0.0000   \n",
       "1  ?|100133144           3.2661           2.6815           1.7301   \n",
       "2  ?|100134869           3.9385           8.9948           6.5650   \n",
       "\n",
       "   TCGA-OR-A5J5-01  TCGA-OR-A5J6-01  TCGA-OR-A5J7-01  TCGA-OR-A5J8-01  \\\n",
       "0           0.0000           0.0000           0.0000           0.0000   \n",
       "1           0.0000           0.0000           1.1673           1.4422   \n",
       "2           1.5492           4.4709           6.0529           2.2876   \n",
       "\n",
       "   TCGA-OR-A5J9-01  TCGA-OR-A5JA-01  ...  TCGA-CG-4449-01  TCGA-CG-4462-01  \\\n",
       "0           0.0000           0.0000  ...              NaN              NaN   \n",
       "1           0.0000           4.4556  ...         4.358154         5.676995   \n",
       "2           1.3599           5.0581  ...         2.656360         3.342794   \n",
       "\n",
       "   TCGA-CG-4465-01  TCGA-CG-4466-01  TCGA-CG-4469-01  TCGA-CG-4472-01  \\\n",
       "0              NaN              NaN              NaN              NaN   \n",
       "1         5.219350        14.846708        20.115492         6.997533   \n",
       "2         2.423442         5.055287        11.626054        13.654193   \n",
       "\n",
       "   TCGA-CG-4474-01  TCGA-CG-4475-01  TCGA-CG-4476-01  TCGA-CG-4477-01  \n",
       "0              NaN              NaN              NaN              NaN  \n",
       "1        18.311906        12.057112        18.628740        17.874417  \n",
       "2         7.417109        11.585177        11.482418        14.919338  \n",
       "\n",
       "[3 rows x 11070 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts_PanCanAtlas.columns = transcripts_PanCanAtlas.columns.str[:15]\n",
    "transcripts_PanCanAtlas.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc52b2b2-7134-4bab-8fd6-643991650cf0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Common samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "54000252-9338-439f-8f0b-89ed3da81bfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample', 'TCGA-S9-A7J2-01', 'TCGA-G3-A3CH-11', 'TCGA-EK-A2RE-01',\n",
       "       'TCGA-44-6778-01', 'TCGA-F4-6854-01', 'TCGA-AB-2863-03',\n",
       "       'TCGA-C8-A1HL-01', 'TCGA-EW-A2FS-01', 'TCGA-05-4420-01',\n",
       "       ...\n",
       "       'TCGA-DJ-A2QC-01', 'TCGA-A8-A09K-01', 'TCGA-61-1907-01',\n",
       "       'TCGA-IB-7885-01', 'TCGA-B6-A0IA-01', 'TCGA-VQ-AA6F-01',\n",
       "       'TCGA-BR-8588-01', 'TCGA-24-2254-01', 'TCGA-DD-A115-01',\n",
       "       'TCGA-FV-A3I0-11'],\n",
       "      dtype='object', length=10536)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xena.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "176b6f8e-ccbf-4346-b710-08aa1d69de37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCGA-OR-A5JP-01', 'TCGA-OR-A5JG-01', 'TCGA-OR-A5K1-01',\n",
       "       'TCGA-OR-A5JR-01', 'TCGA-OR-A5KU-01', 'TCGA-OR-A5L9-01',\n",
       "       'TCGA-OR-A5JQ-01', 'TCGA-OR-A5K4-01', 'TCGA-OR-A5JL-01',\n",
       "       'TCGA-OR-A5LS-01',\n",
       "       ...\n",
       "       'TCGA-VD-A8KI-01', 'TCGA-V4-A9E9-01', 'TCGA-V4-A9F7-01',\n",
       "       'TCGA-V4-A9EO-01', 'TCGA-V4-A9EU-01', 'TCGA-WC-A87T-01',\n",
       "       'TCGA-WC-AA9A-01', 'TCGA-V4-A9EA-01', 'TCGA-RZ-AB0B-01',\n",
       "       'TCGA-V4-A9F8-01'],\n",
       "      dtype='object', length=11057)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "be9f6b6a-aab3-4464-828f-942fd2da2fea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10334"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# xena & PANCAN\n",
    "common = xena.columns.isin(transcripts.columns)\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "b5150973-9b44-439b-9118-e55703e9010d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10904"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PANCAN & PanCanAtlas\n",
    "common = transcripts.columns.isin(transcripts_PanCanAtlas.columns)\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcf7f013-c58d-4a2f-b413-b526301a9b8e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## 2. Gene & Chromosome data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48e5efb1-e60e-440f-bb30-319ed3793c91",
   "metadata": {},
   "source": [
    "### 2.1. Copy number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e991c8e9-482e-4e38-a9c9-d189618e1109",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8235503, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5JE-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>62920</td>\n",
       "      <td>15381037</td>\n",
       "      <td>-0.5540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5JE-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>15383915</td>\n",
       "      <td>21811974</td>\n",
       "      <td>-0.1369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5JE-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>21815733</td>\n",
       "      <td>21898797</td>\n",
       "      <td>0.2401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample Chrom     Start       End   value\n",
       "0  TCGA-OR-A5JE-01A     1     62920  15381037 -0.5540\n",
       "1  TCGA-OR-A5JE-01A     1  15383915  21811974 -0.1369\n",
       "2  TCGA-OR-A5JE-01A     1  21815733  21898797  0.2401"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv = pd.read_csv(PANCAN + \"copy number/GDC-PANCAN.cnv.tsv\", sep=\"\\t\")\n",
    "print(cnv.shape)\n",
    "cnv.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "727196fd-60ef-4710-97f3-893f3d4df02b",
   "metadata": {},
   "source": [
    "### 2.2. Masked CNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "26080e83-7b3b-40ea-a901-cede80473262",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3490756, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5JK-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>3301765</td>\n",
       "      <td>119984772</td>\n",
       "      <td>-0.2971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5JK-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>146716041</td>\n",
       "      <td>197788606</td>\n",
       "      <td>0.0862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5JK-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>197788716</td>\n",
       "      <td>197799377</td>\n",
       "      <td>-1.3311</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample Chrom      Start        End   value\n",
       "0  TCGA-OR-A5JK-01A     1    3301765  119984772 -0.2971\n",
       "1  TCGA-OR-A5JK-01A     1  146716041  197788606  0.0862\n",
       "2  TCGA-OR-A5JK-01A     1  197788716  197799377 -1.3311"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "masked_cnv = pd.read_csv(PANCAN + \"copy number/GDC-PANCAN.masked_cnv.tsv\", sep=\"\\t\")\n",
    "print(masked_cnv.shape)\n",
    "masked_cnv.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fcf4203-54fc-409c-99c3-441fe0e22a1a",
   "metadata": {},
   "source": [
    "### 2.3. Somatic Mutation Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c486cbc1-a41b-4ebc-847a-15adf67f3ac2",
   "metadata": {},
   "source": [
    "All > 2 million rows.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b7e60528-35ba-4eb4-9ed3-963416887c8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2854561, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Amino_Acid_Change</th>\n",
       "      <th>effect</th>\n",
       "      <th>filter</th>\n",
       "      <th>dna_vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5J6-01A</td>\n",
       "      <td>OMA1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>58539205</td>\n",
       "      <td>58539205</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>p.T30T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.117647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5J6-01A</td>\n",
       "      <td>SH3RF3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>109437072</td>\n",
       "      <td>109437072</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>p.S585W</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.212871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5J6-01A</td>\n",
       "      <td>SCN2A</td>\n",
       "      <td>chr2</td>\n",
       "      <td>165344781</td>\n",
       "      <td>165344781</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>p.H930L</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.220401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample_ID    gene chrom      start        end ref alt  \\\n",
       "0  TCGA-OR-A5J6-01A    OMA1  chr1   58539205   58539205   T   C   \n",
       "1  TCGA-OR-A5J6-01A  SH3RF3  chr2  109437072  109437072   C   G   \n",
       "2  TCGA-OR-A5J6-01A   SCN2A  chr2  165344781  165344781   A   T   \n",
       "\n",
       "  Amino_Acid_Change              effect filter   dna_vaf  \n",
       "0            p.T30T  synonymous_variant   PASS  0.117647  \n",
       "1           p.S585W    missense_variant   PASS  0.212871  \n",
       "2           p.H930L    missense_variant   PASS  0.220401  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "varscan2 = pd.read_csv(PANCAN + \"somatic mutation/GDC-PANCAN.varscan2_snv.tsv\", sep=\"\\t\")\n",
    "print(varscan2.shape)\n",
    "varscan2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "071b9e5a-8e63-4e64-8153-c09df2ffa3fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2202277, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Amino_Acid_Change</th>\n",
       "      <th>effect</th>\n",
       "      <th>filter</th>\n",
       "      <th>dna_vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5K8-01A</td>\n",
       "      <td>RP1-163M9.7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16707615</td>\n",
       "      <td>16707615</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.542857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5K8-01A</td>\n",
       "      <td>HHIPL2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>222547868</td>\n",
       "      <td>222547868</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>p.E59D</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.811321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5K8-01A</td>\n",
       "      <td>GALNT2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>230279475</td>\n",
       "      <td>230279475</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.777778</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample_ID         gene chrom      start        end ref alt  \\\n",
       "0  TCGA-OR-A5K8-01A  RP1-163M9.7  chr1   16707615   16707615   C   T   \n",
       "1  TCGA-OR-A5K8-01A       HHIPL2  chr1  222547868  222547868   C   A   \n",
       "2  TCGA-OR-A5K8-01A       GALNT2  chr1  230279475  230279475   C   G   \n",
       "\n",
       "  Amino_Acid_Change                   effect filter   dna_vaf  \n",
       "0               NaN  downstream_gene_variant   PASS  0.542857  \n",
       "1            p.E59D         missense_variant   PASS  0.811321  \n",
       "2               NaN      3_prime_UTR_variant   PASS  0.777778  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "somaticsniper = pd.read_csv(PANCAN + \"somatic mutation/GDC-PANCAN.somaticsniper_snv.tsv\", sep=\"\\t\")\n",
    "print(somaticsniper.shape)\n",
    "somaticsniper.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c80b531f-7ec6-4a3e-b874-d698cee56ab3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3175929, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Amino_Acid_Change</th>\n",
       "      <th>effect</th>\n",
       "      <th>filter</th>\n",
       "      <th>dna_vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-P6-A5OH-01A</td>\n",
       "      <td>RER1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2402218</td>\n",
       "      <td>2402219</td>\n",
       "      <td>-</td>\n",
       "      <td>CAAGGGCATCCTTGTGGCTATGGTCTGTACTTTCTTCTACGCTTT</td>\n",
       "      <td>p.F138_D139insYAFKGILVAMVCTFF</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.172414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-P6-A5OH-01A</td>\n",
       "      <td>WASF2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>27410171</td>\n",
       "      <td>27410171</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.A287V</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.255319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-P6-A5OH-01A</td>\n",
       "      <td>MACF1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>39455094</td>\n",
       "      <td>39455094</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>p.H6923Q</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.072072</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample_ID   gene chrom     start       end ref  \\\n",
       "0  TCGA-P6-A5OH-01A   RER1  chr1   2402218   2402219   -   \n",
       "1  TCGA-P6-A5OH-01A  WASF2  chr1  27410171  27410171   G   \n",
       "2  TCGA-P6-A5OH-01A  MACF1  chr1  39455094  39455094   T   \n",
       "\n",
       "                                             alt  \\\n",
       "0  CAAGGGCATCCTTGTGGCTATGGTCTGTACTTTCTTCTACGCTTT   \n",
       "1                                              A   \n",
       "2                                              A   \n",
       "\n",
       "               Amino_Acid_Change             effect filter   dna_vaf  \n",
       "0  p.F138_D139insYAFKGILVAMVCTFF  inframe_insertion   PASS  0.172414  \n",
       "1                        p.A287V   missense_variant   PASS  0.255319  \n",
       "2                       p.H6923Q   missense_variant   PASS  0.072072  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2 = pd.read_csv(PANCAN + \"somatic mutation/GDC-PANCAN.mutect2_snv.tsv\", sep=\"\\t\")\n",
    "print(mutect2.shape)\n",
    "mutect2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ffc60f10-cbce-4d4d-88f4-152cc5fa5b1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2684788, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Amino_Acid_Change</th>\n",
       "      <th>effect</th>\n",
       "      <th>filter</th>\n",
       "      <th>dna_vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5J7-01A</td>\n",
       "      <td>CTBS</td>\n",
       "      <td>chr1</td>\n",
       "      <td>84570701</td>\n",
       "      <td>84570701</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>p.G66E</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.146667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5J7-01A</td>\n",
       "      <td>ATF6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>161791444</td>\n",
       "      <td>161791444</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>p.P131A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.491525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5J7-01A</td>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>234309160</td>\n",
       "      <td>234309160</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>p.A154E</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.218182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample_ID     gene chrom      start        end ref alt  \\\n",
       "0  TCGA-OR-A5J7-01A     CTBS  chr1   84570701   84570701   C   T   \n",
       "1  TCGA-OR-A5J7-01A     ATF6  chr1  161791444  161791444   C   G   \n",
       "2  TCGA-OR-A5J7-01A  SLC35F3  chr1  234309160  234309160   C   A   \n",
       "\n",
       "  Amino_Acid_Change            effect filter   dna_vaf  \n",
       "0            p.G66E  missense_variant   PASS  0.146667  \n",
       "1           p.P131A  missense_variant   PASS  0.491525  \n",
       "2           p.A154E  missense_variant   PASS  0.218182  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muse = pd.read_csv(PANCAN + \"somatic mutation/GDC-PANCAN.muse_snv.tsv\", sep=\"\\t\")\n",
    "print(muse.shape)\n",
    "muse.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23f81031-5ec7-41b0-9ae9-bdf7007e90a1",
   "metadata": {},
   "source": [
    "## 3. Sample data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6558741-f19b-4081-9625-597b97ecccd6",
   "metadata": {},
   "source": [
    "### 3.1. Basic clinical data by sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "93d944bc-afab-4329-a1c9-97cd29bc4cf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19188, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>program</th>\n",
       "      <th>sample_type_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>project_id</th>\n",
       "      <th>Age at Diagnosis in Years</th>\n",
       "      <th>Gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-69-7978-01A</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>59.0</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-AR-A24Z-01A</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-D1-A103-01A</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>87.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample program  sample_type_id    sample_type project_id  \\\n",
       "0  TCGA-69-7978-01A    TCGA               1  Primary Tumor  TCGA-LUAD   \n",
       "1  TCGA-AR-A24Z-01A    TCGA               1  Primary Tumor  TCGA-BRCA   \n",
       "2  TCGA-D1-A103-01A    TCGA               1  Primary Tumor  TCGA-UCEC   \n",
       "\n",
       "   Age at Diagnosis in Years  Gender  \n",
       "0                       59.0    Male  \n",
       "1                       57.0  Female  \n",
       "2                       87.0  Female  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basic_phenotype = pd.read_csv(PANCAN + \"basic-phenotype/GDC-PANCAN.basic_phenotype.tsv\", sep=\"\\t\")\n",
    "print(basic_phenotype.shape)\n",
    "basic_phenotype.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "535bf702-e72d-4ae8-bff0-56cc6c4e0e6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16186, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>program</th>\n",
       "      <th>sample_type_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>project_id</th>\n",
       "      <th>Age at Diagnosis in Years</th>\n",
       "      <th>Gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-69-7978-01</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>59.0</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-AR-A24Z-01</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-D1-A103-01</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>87.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample program  sample_type_id    sample_type project_id  \\\n",
       "0  TCGA-69-7978-01    TCGA               1  Primary Tumor  TCGA-LUAD   \n",
       "1  TCGA-AR-A24Z-01    TCGA               1  Primary Tumor  TCGA-BRCA   \n",
       "2  TCGA-D1-A103-01    TCGA               1  Primary Tumor  TCGA-UCEC   \n",
       "\n",
       "   Age at Diagnosis in Years  Gender  \n",
       "0                       59.0    Male  \n",
       "1                       57.0  Female  \n",
       "2                       87.0  Female  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basic_phenotype[\"sample\"] = basic_phenotype[\"sample\"].str[:15]\n",
    "basic_phenotype = basic_phenotype.drop_duplicates(subset=\"sample\")\n",
    "print(basic_phenotype.shape)\n",
    "basic_phenotype.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a2afc53-5468-4208-af6b-663829f8c089",
   "metadata": {},
   "source": [
    "### 3.2. Survival per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a06dc3f3-6740-4ec6-89b3-faedbc25fe8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18492, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>OS</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS.time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5KZ-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-OR-A5KZ</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5LC-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-OR-A5LC</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-P6-A5OF-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-P6-A5OF</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample  OS      _PATIENT  OS.time\n",
       "0  TCGA-OR-A5KZ-01A   1  TCGA-OR-A5KZ      125\n",
       "1  TCGA-OR-A5LC-01A   1  TCGA-OR-A5LC      159\n",
       "2  TCGA-P6-A5OF-01A   1  TCGA-P6-A5OF      207"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "survival = pd.read_csv(PANCAN + \"phenotype/GDC-PANCAN.survival.tsv\", sep=\"\\t\")\n",
    "print(survival.shape)\n",
    "survival.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "01627529-54f5-49a6-9f57-8bbf232244e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15719, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>OS</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS.time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5KZ-01</td>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-OR-A5KZ</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5LC-01</td>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-OR-A5LC</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-P6-A5OF-01</td>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-P6-A5OF</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample  OS      _PATIENT  OS.time\n",
       "0  TCGA-OR-A5KZ-01   1  TCGA-OR-A5KZ      125\n",
       "1  TCGA-OR-A5LC-01   1  TCGA-OR-A5LC      159\n",
       "2  TCGA-P6-A5OF-01   1  TCGA-P6-A5OF      207"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "survival[\"sample\"] = survival[\"sample\"].str[:15]\n",
    "survival = survival.drop_duplicates(subset=\"sample\")\n",
    "print(survival.shape)\n",
    "survival.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc108828-4799-47d8-bd26-a4f7e87384c5",
   "metadata": {},
   "source": [
    "### 3.3. Phenotype per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e9bbbf1d-69b7-4bba-a6a2-e96adda42d24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14318, 58)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>demographic.age_at_index</th>\n",
       "      <th>demographic.created_datetime</th>\n",
       "      <th>demographic.days_to_birth</th>\n",
       "      <th>demographic.days_to_death</th>\n",
       "      <th>demographic.demographic_id</th>\n",
       "      <th>demographic.ethnicity</th>\n",
       "      <th>demographic.gender</th>\n",
       "      <th>demographic.race</th>\n",
       "      <th>demographic.state</th>\n",
       "      <th>...</th>\n",
       "      <th>exposures.years_smoked</th>\n",
       "      <th>id</th>\n",
       "      <th>project.name</th>\n",
       "      <th>project.project_id</th>\n",
       "      <th>tissue_source_site.name</th>\n",
       "      <th>samples.is_ffpe</th>\n",
       "      <th>samples.sample_id</th>\n",
       "      <th>samples.sample_type</th>\n",
       "      <th>samples.sample_type_id</th>\n",
       "      <th>samples.tissue_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-AX-A064-01A</td>\n",
       "      <td>81.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-29625.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7a2e57b2-e19b-5407-967c-2f076d5ee7fc</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>released</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>633e6994-3e8b-4c0b-9b79-06cf4ab6f9bc</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Gynecologic Oncology Group</td>\n",
       "      <td>False</td>\n",
       "      <td>3d85e546-d9f4-49ba-adbd-301118cd1307</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-AP-A0L8-01A</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-25733.0</td>\n",
       "      <td>1484.0</td>\n",
       "      <td>dc7a4a86-4e5d-5763-939f-1ef9f9df5bf5</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>released</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9aa8c2b4-db32-4899-bf0c-d578cb175e90</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>False</td>\n",
       "      <td>d1a7a58f-7a35-4a94-a337-051ae56ded00</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-D1-A177-01A</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-25602.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dd1d384b-f3e5-53cb-b3e1-cff1a4d84850</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>released</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>e8c58d8b-51d7-4599-af79-469ee6f269db</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Mayo Clinic</td>\n",
       "      <td>False</td>\n",
       "      <td>97587fd4-0f6f-4af5-ac0a-7328abb40e9c</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample  demographic.age_at_index demographic.created_datetime  \\\n",
       "0  TCGA-AX-A064-01A                      81.0                          NaN   \n",
       "1  TCGA-AP-A0L8-01A                      70.0                          NaN   \n",
       "2  TCGA-D1-A177-01A                      70.0                          NaN   \n",
       "\n",
       "   demographic.days_to_birth  demographic.days_to_death  \\\n",
       "0                   -29625.0                        NaN   \n",
       "1                   -25733.0                     1484.0   \n",
       "2                   -25602.0                        NaN   \n",
       "\n",
       "             demographic.demographic_id   demographic.ethnicity  \\\n",
       "0  7a2e57b2-e19b-5407-967c-2f076d5ee7fc            not reported   \n",
       "1  dc7a4a86-4e5d-5763-939f-1ef9f9df5bf5            not reported   \n",
       "2  dd1d384b-f3e5-53cb-b3e1-cff1a4d84850  not hispanic or latino   \n",
       "\n",
       "  demographic.gender           demographic.race demographic.state  ...  \\\n",
       "0             female                      white          released  ...   \n",
       "1             female  black or african american          released  ...   \n",
       "2             female                      white          released  ...   \n",
       "\n",
       "  exposures.years_smoked                                    id  \\\n",
       "0                    NaN  633e6994-3e8b-4c0b-9b79-06cf4ab6f9bc   \n",
       "1                    NaN  9aa8c2b4-db32-4899-bf0c-d578cb175e90   \n",
       "2                    NaN  e8c58d8b-51d7-4599-af79-469ee6f269db   \n",
       "\n",
       "                           project.name  project.project_id  \\\n",
       "0  Uterine Corpus Endometrial Carcinoma           TCGA-UCEC   \n",
       "1  Uterine Corpus Endometrial Carcinoma           TCGA-UCEC   \n",
       "2  Uterine Corpus Endometrial Carcinoma           TCGA-UCEC   \n",
       "\n",
       "      tissue_source_site.name  samples.is_ffpe  \\\n",
       "0  Gynecologic Oncology Group            False   \n",
       "1                       MSKCC            False   \n",
       "2                 Mayo Clinic            False   \n",
       "\n",
       "                      samples.sample_id samples.sample_type  \\\n",
       "0  3d85e546-d9f4-49ba-adbd-301118cd1307       Primary Tumor   \n",
       "1  d1a7a58f-7a35-4a94-a337-051ae56ded00       Primary Tumor   \n",
       "2  97587fd4-0f6f-4af5-ac0a-7328abb40e9c       Primary Tumor   \n",
       "\n",
       "   samples.sample_type_id  samples.tissue_type  \n",
       "0                       1         Not Reported  \n",
       "1                       1         Not Reported  \n",
       "2                       1         Not Reported  \n",
       "\n",
       "[3 rows x 58 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype = pd.read_csv(PANCAN + \"phenotype/GDC-PANCAN.TCGA_phenotype.tsv\", sep=\"\\t\")\n",
    "print(phenotype.shape)\n",
    "phenotype.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "196ef055-dd80-4e96-905a-f6fc6efb38d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13926, 58)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>demographic.age_at_index</th>\n",
       "      <th>demographic.created_datetime</th>\n",
       "      <th>demographic.days_to_birth</th>\n",
       "      <th>demographic.days_to_death</th>\n",
       "      <th>demographic.demographic_id</th>\n",
       "      <th>demographic.ethnicity</th>\n",
       "      <th>demographic.gender</th>\n",
       "      <th>demographic.race</th>\n",
       "      <th>demographic.state</th>\n",
       "      <th>...</th>\n",
       "      <th>exposures.years_smoked</th>\n",
       "      <th>id</th>\n",
       "      <th>project.name</th>\n",
       "      <th>project.project_id</th>\n",
       "      <th>tissue_source_site.name</th>\n",
       "      <th>samples.is_ffpe</th>\n",
       "      <th>samples.sample_id</th>\n",
       "      <th>samples.sample_type</th>\n",
       "      <th>samples.sample_type_id</th>\n",
       "      <th>samples.tissue_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-AX-A064-01</td>\n",
       "      <td>81.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-29625.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7a2e57b2-e19b-5407-967c-2f076d5ee7fc</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>released</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>633e6994-3e8b-4c0b-9b79-06cf4ab6f9bc</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Gynecologic Oncology Group</td>\n",
       "      <td>False</td>\n",
       "      <td>3d85e546-d9f4-49ba-adbd-301118cd1307</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-AP-A0L8-01</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-25733.0</td>\n",
       "      <td>1484.0</td>\n",
       "      <td>dc7a4a86-4e5d-5763-939f-1ef9f9df5bf5</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>released</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9aa8c2b4-db32-4899-bf0c-d578cb175e90</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>MSKCC</td>\n",
       "      <td>False</td>\n",
       "      <td>d1a7a58f-7a35-4a94-a337-051ae56ded00</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-D1-A177-01</td>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-25602.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dd1d384b-f3e5-53cb-b3e1-cff1a4d84850</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>released</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>e8c58d8b-51d7-4599-af79-469ee6f269db</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Mayo Clinic</td>\n",
       "      <td>False</td>\n",
       "      <td>97587fd4-0f6f-4af5-ac0a-7328abb40e9c</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample  demographic.age_at_index demographic.created_datetime  \\\n",
       "0  TCGA-AX-A064-01                      81.0                          NaN   \n",
       "1  TCGA-AP-A0L8-01                      70.0                          NaN   \n",
       "2  TCGA-D1-A177-01                      70.0                          NaN   \n",
       "\n",
       "   demographic.days_to_birth  demographic.days_to_death  \\\n",
       "0                   -29625.0                        NaN   \n",
       "1                   -25733.0                     1484.0   \n",
       "2                   -25602.0                        NaN   \n",
       "\n",
       "             demographic.demographic_id   demographic.ethnicity  \\\n",
       "0  7a2e57b2-e19b-5407-967c-2f076d5ee7fc            not reported   \n",
       "1  dc7a4a86-4e5d-5763-939f-1ef9f9df5bf5            not reported   \n",
       "2  dd1d384b-f3e5-53cb-b3e1-cff1a4d84850  not hispanic or latino   \n",
       "\n",
       "  demographic.gender           demographic.race demographic.state  ...  \\\n",
       "0             female                      white          released  ...   \n",
       "1             female  black or african american          released  ...   \n",
       "2             female                      white          released  ...   \n",
       "\n",
       "  exposures.years_smoked                                    id  \\\n",
       "0                    NaN  633e6994-3e8b-4c0b-9b79-06cf4ab6f9bc   \n",
       "1                    NaN  9aa8c2b4-db32-4899-bf0c-d578cb175e90   \n",
       "2                    NaN  e8c58d8b-51d7-4599-af79-469ee6f269db   \n",
       "\n",
       "                           project.name  project.project_id  \\\n",
       "0  Uterine Corpus Endometrial Carcinoma           TCGA-UCEC   \n",
       "1  Uterine Corpus Endometrial Carcinoma           TCGA-UCEC   \n",
       "2  Uterine Corpus Endometrial Carcinoma           TCGA-UCEC   \n",
       "\n",
       "      tissue_source_site.name  samples.is_ffpe  \\\n",
       "0  Gynecologic Oncology Group            False   \n",
       "1                       MSKCC            False   \n",
       "2                 Mayo Clinic            False   \n",
       "\n",
       "                      samples.sample_id samples.sample_type  \\\n",
       "0  3d85e546-d9f4-49ba-adbd-301118cd1307       Primary Tumor   \n",
       "1  d1a7a58f-7a35-4a94-a337-051ae56ded00       Primary Tumor   \n",
       "2  97587fd4-0f6f-4af5-ac0a-7328abb40e9c       Primary Tumor   \n",
       "\n",
       "   samples.sample_type_id  samples.tissue_type  \n",
       "0                       1         Not Reported  \n",
       "1                       1         Not Reported  \n",
       "2                       1         Not Reported  \n",
       "\n",
       "[3 rows x 58 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype[\"sample\"] = phenotype[\"sample\"].str[:15]\n",
    "phenotype = phenotype.drop_duplicates(subset=\"sample\")\n",
    "print(phenotype.shape)\n",
    "phenotype.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3170d65a-b533-48ce-bbb8-a6e62bf00ae8",
   "metadata": {},
   "source": [
    "### 3.4. Genome Purity and doublings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dc94e45-2832-4c56-b593-f630284dbf5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ABSOLUTE purity/ploidy file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "971bbd67-bf38-4446-8b4d-b5f6dacd2589",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10786, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>array</th>\n",
       "      <th>sample</th>\n",
       "      <th>call status</th>\n",
       "      <th>purity</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>Genome doublings</th>\n",
       "      <th>Coverage for 80% power</th>\n",
       "      <th>Cancer DNA fraction</th>\n",
       "      <th>Subclonal genome fraction</th>\n",
       "      <th>solution</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5J1-01</td>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.90</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.02</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5J2-01</td>\n",
       "      <td>TCGA-OR-A5J2-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.16</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5J3-01</td>\n",
       "      <td>TCGA-OR-A5J3-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.93</td>\n",
       "      <td>1.27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.11</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             array                        sample call status  purity  ploidy  \\\n",
       "0  TCGA-OR-A5J1-01  TCGA-OR-A5J1-01A-11D-A29H-01      called    0.90    2.00   \n",
       "1  TCGA-OR-A5J2-01  TCGA-OR-A5J2-01A-11D-A29H-01      called    0.89    1.30   \n",
       "2  TCGA-OR-A5J3-01  TCGA-OR-A5J3-01A-11D-A29H-01      called    0.93    1.27   \n",
       "\n",
       "   Genome doublings  Coverage for 80% power  Cancer DNA fraction  \\\n",
       "0               0.0                     9.0                 0.90   \n",
       "1               0.0                     6.0                 0.84   \n",
       "2               0.0                     5.0                 0.89   \n",
       "\n",
       "   Subclonal genome fraction solution  \n",
       "0                       0.02      new  \n",
       "1                       0.16      new  \n",
       "2                       0.11      new  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "purity = pd.read_csv(PANCANATLAS + \"TCGA_mastercalls.abs_tables_JSedit.fixed.txt\", sep=\"\\t\")\n",
    "print(purity.shape)\n",
    "purity.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12325aaa-90da-4b2a-87ef-4519b8a08c48",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 3.5. PARADIGM Pathway Inference Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9efa168e-6f7f-4440-82cd-a8a8a2310570",
   "metadata": {},
   "source": [
    "PARADIGM pathway analysis of mRNASeq expression data. The method for inferring patient-specific genetic activities incorporating curated pathway interactions among genes. A gene is modeled by a factor graph as a set of interconnected variables encoding the expression and known activity of a gene and its products, allowing the incorporation of many types of omic data as evidence. The method predicts the degree to which a pathway's activities (e.g. internal gene states, interactions or high-level 'outputs') are altered in the patient using probabilistic inference.\n",
    "- https://academic.oup.com/bioinformatics/article/26/12/i237/282591\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "321391ab-a62d-45d9-800c-7e961dcaa0f7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19503, 9830)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>TCGA-06-2557</th>\n",
       "      <th>TCGA-85-8287</th>\n",
       "      <th>TCGA-DH-A7UV</th>\n",
       "      <th>TCGA-YG-AA3O</th>\n",
       "      <th>TCGA-EE-A29Q</th>\n",
       "      <th>TCGA-FP-A9TM</th>\n",
       "      <th>TCGA-94-A5I4</th>\n",
       "      <th>TCGA-BP-4799</th>\n",
       "      <th>TCGA-CZ-4856</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-EK-A2RL</th>\n",
       "      <th>TCGA-XC-AA0X</th>\n",
       "      <th>TCGA-E2-A150</th>\n",
       "      <th>TCGA-BR-A4J4</th>\n",
       "      <th>TCGA-EL-A3H3</th>\n",
       "      <th>TCGA-AJ-A3NE</th>\n",
       "      <th>TCGA-C8-A12N</th>\n",
       "      <th>TCGA-DU-7302</th>\n",
       "      <th>TCGA-44-7669</th>\n",
       "      <th>TCGA-WK-A8Y0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UBE2Q1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-1.570240</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.104788</td>\n",
       "      <td>-1.57024</td>\n",
       "      <td>-2.16255</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.288492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRIK_3_homomer_(complex)</td>\n",
       "      <td>0.364151</td>\n",
       "      <td>-1.22353</td>\n",
       "      <td>0.334967</td>\n",
       "      <td>-1.00429</td>\n",
       "      <td>0.334967</td>\n",
       "      <td>-1.07048</td>\n",
       "      <td>-1.22353</td>\n",
       "      <td>0.334967</td>\n",
       "      <td>0.334967</td>\n",
       "      <td>...</td>\n",
       "      <td>0.213701</td>\n",
       "      <td>0.334967</td>\n",
       "      <td>0.364151</td>\n",
       "      <td>-1.07048</td>\n",
       "      <td>-1.07048</td>\n",
       "      <td>0.213701</td>\n",
       "      <td>0.334967</td>\n",
       "      <td>0.239549</td>\n",
       "      <td>-1.00429</td>\n",
       "      <td>-1.004290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UBE2Q2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.44315</td>\n",
       "      <td>-1.10958</td>\n",
       "      <td>0.192315</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.192315</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.10958</td>\n",
       "      <td>-1.44315</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.192315</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 9830 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  TCGA-06-2557  TCGA-85-8287  TCGA-DH-A7UV  \\\n",
       "0                    UBE2Q1      0.000000       0.00000     -1.570240   \n",
       "1  GRIK_3_homomer_(complex)      0.364151      -1.22353      0.334967   \n",
       "2                    UBE2Q2      0.000000       0.00000      0.000000   \n",
       "\n",
       "   TCGA-YG-AA3O  TCGA-EE-A29Q  TCGA-FP-A9TM  TCGA-94-A5I4  TCGA-BP-4799  \\\n",
       "0       0.00000      0.104788      -1.57024      -2.16255      0.000000   \n",
       "1      -1.00429      0.334967      -1.07048      -1.22353      0.334967   \n",
       "2       0.00000      0.000000      -1.44315      -1.10958      0.192315   \n",
       "\n",
       "   TCGA-CZ-4856  ...  TCGA-EK-A2RL  TCGA-XC-AA0X  TCGA-E2-A150  TCGA-BR-A4J4  \\\n",
       "0      0.000000  ...      0.000000      0.000000      0.000000       0.00000   \n",
       "1      0.334967  ...      0.213701      0.334967      0.364151      -1.07048   \n",
       "2      0.000000  ...      0.192315      0.000000      0.000000      -1.10958   \n",
       "\n",
       "   TCGA-EL-A3H3  TCGA-AJ-A3NE  TCGA-C8-A12N  TCGA-DU-7302  TCGA-44-7669  \\\n",
       "0       0.00000      0.000000      0.000000      0.000000       0.00000   \n",
       "1      -1.07048      0.213701      0.334967      0.239549      -1.00429   \n",
       "2      -1.44315      0.000000      0.192315      0.000000       0.00000   \n",
       "\n",
       "   TCGA-WK-A8Y0  \n",
       "0      0.288492  \n",
       "1     -1.004290  \n",
       "2      0.000000  \n",
       "\n",
       "[3 rows x 9830 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Data compressed in `merge_merged_reals.tar.gz`\n",
    "paradigm = pd.read_csv(PANCANATLAS + \"merge_merged_reals.txt\", sep=\"\\t\")\n",
    "print(paradigm.shape)\n",
    "paradigm.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c582b307-780a-4045-ae4a-ea4d4a3daea0",
   "metadata": {},
   "source": [
    "### 3.6. Copy number per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0e088c70-1126-4375-99ab-1927214db982",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19729, 11369)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>TCGA-OR-A5KO-01A</th>\n",
       "      <th>TCGA-OR-A5KZ-01A</th>\n",
       "      <th>TCGA-OR-A5LA-01A</th>\n",
       "      <th>TCGA-OR-A5LP-01A</th>\n",
       "      <th>TCGA-OR-A5LR-01A</th>\n",
       "      <th>TCGA-OR-A5LJ-01A</th>\n",
       "      <th>TCGA-OR-A5KY-01A</th>\n",
       "      <th>TCGA-OR-A5L9-01A</th>\n",
       "      <th>TCGA-OR-A5LH-01A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-VD-A8KM-01A</th>\n",
       "      <th>TCGA-VD-AA8Q-01A</th>\n",
       "      <th>TCGA-V3-A9ZX-01A</th>\n",
       "      <th>TCGA-V4-A9EC-01A</th>\n",
       "      <th>TCGA-V4-A9EK-01A</th>\n",
       "      <th>TCGA-V4-A9EU-01A</th>\n",
       "      <th>TCGA-V4-A9F1-01A</th>\n",
       "      <th>TCGA-WC-A885-01A</th>\n",
       "      <th>TCGA-WC-A87W-01A</th>\n",
       "      <th>TCGA-YZ-A983-01A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11369 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Unnamed: 0  TCGA-OR-A5KO-01A  TCGA-OR-A5KZ-01A  TCGA-OR-A5LA-01A  \\\n",
       "0  ENSG00000000003.13                 0                 0                 0   \n",
       "1   ENSG00000000005.5                 0                 0                 0   \n",
       "2  ENSG00000000419.11                 0                 0                 0   \n",
       "\n",
       "   TCGA-OR-A5LP-01A  TCGA-OR-A5LR-01A  TCGA-OR-A5LJ-01A  TCGA-OR-A5KY-01A  \\\n",
       "0                 0                 0                 0                 0   \n",
       "1                 0                 0                 0                 0   \n",
       "2                 0                 0                 0                 0   \n",
       "\n",
       "   TCGA-OR-A5L9-01A  TCGA-OR-A5LH-01A  ...  TCGA-VD-A8KM-01A  \\\n",
       "0                 0                 0  ...                 0   \n",
       "1                 0                 0  ...                 0   \n",
       "2                 0                 0  ...                 0   \n",
       "\n",
       "   TCGA-VD-AA8Q-01A  TCGA-V3-A9ZX-01A  TCGA-V4-A9EC-01A  TCGA-V4-A9EK-01A  \\\n",
       "0                 0                 0                 0                 0   \n",
       "1                 0                 0                 0                 0   \n",
       "2                 0                 0                 0                 0   \n",
       "\n",
       "   TCGA-V4-A9EU-01A  TCGA-V4-A9F1-01A  TCGA-WC-A885-01A  TCGA-WC-A87W-01A  \\\n",
       "0                 0                 0                 0                 0   \n",
       "1                 0                 0                 0                 0   \n",
       "2                 0                 0                 0                 0   \n",
       "\n",
       "   TCGA-YZ-A983-01A  \n",
       "0                 0  \n",
       "1                 0  \n",
       "2                 0  \n",
       "\n",
       "[3 rows x 11369 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gistic = pd.read_csv(PATH + \"copy number gene/GDC-PANCAN.gistic.tsv\", sep=\"\\t\")\n",
    "print(gistic.shape)\n",
    "gistic.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8091c35-fda8-41a5-a62d-95ca5366340b",
   "metadata": {},
   "source": [
    "### 3.7. Xena SNP  \n",
    "processed subset derived from the MC3 data with nonsilent calls summarized at the gene level per sample. samples without any PASS nonsilent mutations may be excluded. Entries are usually:\n",
    "- 0 : no nonsilent mutation present\n",
    "- 1 : nonsilent mutation present\n",
    "- Only genes with at least one mutation are included"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e5d26864-d8b0-45e1-b3be-c132fc2e6406",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9104, 40543)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>UBE2Q2</th>\n",
       "      <th>CHMP1B</th>\n",
       "      <th>PSMA2P1</th>\n",
       "      <th>SHQ1P1</th>\n",
       "      <th>CPHL1P</th>\n",
       "      <th>SSXP10</th>\n",
       "      <th>REM1</th>\n",
       "      <th>TCOF1</th>\n",
       "      <th>NSRP1</th>\n",
       "      <th>OPA6</th>\n",
       "      <th>...</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>OR1E5</th>\n",
       "      <th>RP11-390F4.3</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>PTRF</th>\n",
       "      <th>DIAPH2-AS1</th>\n",
       "      <th>SELV</th>\n",
       "      <th>NFIX</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0003-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0033-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 40543 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample           UBE2Q2  CHMP1B  PSMA2P1  SHQ1P1  CPHL1P  SSXP10  REM1  TCOF1  \\\n",
       "TCGA-02-0003-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0033-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0047-01       0       0        0       0       0       0     0      0   \n",
       "\n",
       "sample           NSRP1  OPA6  ...  TULP2  OR1E5  RP11-390F4.3  GNGT2  GNGT1  \\\n",
       "TCGA-02-0003-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0033-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0047-01      0     0  ...      0      0             0      0      0   \n",
       "\n",
       "sample           PTRF  DIAPH2-AS1  SELV  NFIX  SELP  \n",
       "TCGA-02-0003-01     0           0     0     0     0  \n",
       "TCGA-02-0033-01     0           0     0     0     0  \n",
       "TCGA-02-0047-01     0           0     0     0     0  \n",
       "\n",
       "[3 rows x 40543 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc3 = pd.read_csv(XENA + \"SNP/mc3.v0.2.8.PUBLIC.nonsilentGene.xena.txt\", sep=\"\\t\")\n",
    "mc3 = mc3.set_index(\"sample\").T\n",
    "print(mc3.shape)\n",
    "mc3.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba3ac8e5-6530-41fc-8203-d9295914abbb",
   "metadata": {},
   "source": [
    "## 4. Patient data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74816b32-d3bb-4dad-bef8-bf6bb30c653d",
   "metadata": {},
   "source": [
    "### 4.1. Clinical data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae76adf8-a6c7-4f87-80a0-7be90496dac2",
   "metadata": {},
   "source": [
    "https://gdc.cancer.gov/about-data/publications/pancanatlas recommends using TCGA-CDR-SupplementalTableS1.xlsx (see PATH)\n",
    "- A curated resource of the clinical annotations for TCGA data and provides recommendations for use of clinical endpoints\n",
    "- It is strongly recommended that this file be used for clinical elements and survival outcome data first; more details please see the TCGA-CDR paper: Liu et al. Cell 2018: https://www.cell.com/cell/fulltext/S0092-8674(18)30229-0\n",
    "- Columns are explained in the sheets: TCGA-CDR_Notes and ExtraEndpoints_Notes in PATH+TCGA-CDR-SupplementalTableS1.xlsx\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ce94c8af-51a8-4c1e-906c-f35a6a4e629f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11160, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>type</th>\n",
       "      <th>PFI.1</th>\n",
       "      <th>PFI.time.1</th>\n",
       "      <th>PFI.2</th>\n",
       "      <th>PFI.time.2</th>\n",
       "      <th>PFS</th>\n",
       "      <th>PFS.time</th>\n",
       "      <th>DSS_cr</th>\n",
       "      <th>DSS.time.cr</th>\n",
       "      <th>DFI.cr</th>\n",
       "      <th>DFI.time.cr</th>\n",
       "      <th>PFI.cr</th>\n",
       "      <th>PFI.time.cr</th>\n",
       "      <th>PFI.1.cr</th>\n",
       "      <th>PFI.time.1.cr</th>\n",
       "      <th>PFI.2.cr</th>\n",
       "      <th>PFI.time.2.cr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-OR-A5J1</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "      <td>1</td>\n",
       "      <td>1355</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "      <td>1</td>\n",
       "      <td>754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TCGA-OR-A5J2</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>1</td>\n",
       "      <td>1677</td>\n",
       "      <td>#N/D</td>\n",
       "      <td>#N/D</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-OR-A5J3</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>2091</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 bcr_patient_barcode type PFI.1 PFI.time.1 PFI.2 PFI.time.2 PFS  \\\n",
       "0           1        TCGA-OR-A5J1  ACC     1        754     1        754   1   \n",
       "1           2        TCGA-OR-A5J2  ACC     1        289     1        289   1   \n",
       "2           3        TCGA-OR-A5J3  ACC     1         53     1         53   1   \n",
       "\n",
       "  PFS.time DSS_cr DSS.time.cr DFI.cr DFI.time.cr PFI.cr PFI.time.cr PFI.1.cr  \\\n",
       "0      754      1        1355      1         754      1         754        1   \n",
       "1      289      1        1677   #N/D        #N/D      1         289        1   \n",
       "2       53      0        2091      1          53      1          53        1   \n",
       "\n",
       "  PFI.time.1.cr PFI.2.cr PFI.time.2.cr  \n",
       "0           754        1           754  \n",
       "1           289        1           289  \n",
       "2            53        1            53  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical = pd.read_csv(PANCANATLAS + \"Clinical end points.tsv\", sep=\"\\t\")\n",
    "print(clinical.shape)\n",
    "clinical.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c83d7101-89e6-42bf-8fb5-a0769ec2188a",
   "metadata": {},
   "source": [
    "### 4.2. Clinical data with follow-up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c6123099-84cf-4755-819e-79b2a566ea29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10956, 746)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yw/q_pg1y917pz9nzm99n51ctzh0000gp/T/ipykernel_1379/1446671467.py:1: DtypeWarning: Columns (8,9,27,42,44,45,46,47,48,49,50,51,52,53,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,333,334,335,336,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407,408,409,410,411,412,413,414,415,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,664,665,666,667,668,669,670,671,672,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694,695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714,715,716,717,718,719,720,721,722,723,724,725,726,727,728,729,730,731,732,733,734,735,736,737,738,739,740,741,742,743,744) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinical_follow = pd.read_csv(PANCANATLAS + \"clinical_PANCAN_patient_with_followup.tsv\", sep=\"\\t\",\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bcr_patient_uuid</th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>acronym</th>\n",
       "      <th>gender</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_followup</th>\n",
       "      <th>days_to_initial_pathologic_diagnosis</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>...</th>\n",
       "      <th>total_bilirubin_upper_limit</th>\n",
       "      <th>platelet_result_count</th>\n",
       "      <th>fibrosis_ishak_score</th>\n",
       "      <th>fetoprotein_outcome_value</th>\n",
       "      <th>fetoprotein_outcome_upper_limit</th>\n",
       "      <th>fetoprotein_outcome_lower_limit</th>\n",
       "      <th>inter_norm_ratio_lower_limit</th>\n",
       "      <th>family_cancer_type_txt</th>\n",
       "      <th>bilirubin_upper_limit</th>\n",
       "      <th>days_to_last_known_alive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B3164F7B-C826-4E08-9EE6-8FF96D29B913</td>\n",
       "      <td>TCGA-OR-A5J1</td>\n",
       "      <td>ACC</td>\n",
       "      <td>MALE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-21496</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8E7C2E31-D085-4B75-A970-162526DD07A0</td>\n",
       "      <td>TCGA-OR-A5J2</td>\n",
       "      <td>ACC</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-16090</td>\n",
       "      <td>1677</td>\n",
       "      <td>[Not Available]</td>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DFD687BC-6E69-42F7-AF94-D17FC150D1A1</td>\n",
       "      <td>TCGA-OR-A5J3</td>\n",
       "      <td>ACC</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Alive</td>\n",
       "      <td>-8624</td>\n",
       "      <td>[Not Applicable]</td>\n",
       "      <td>2091.0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 746 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       bcr_patient_uuid bcr_patient_barcode acronym  gender  \\\n",
       "0  B3164F7B-C826-4E08-9EE6-8FF96D29B913        TCGA-OR-A5J1     ACC    MALE   \n",
       "1  8E7C2E31-D085-4B75-A970-162526DD07A0        TCGA-OR-A5J2     ACC  FEMALE   \n",
       "2  DFD687BC-6E69-42F7-AF94-D17FC150D1A1        TCGA-OR-A5J3     ACC  FEMALE   \n",
       "\n",
       "  vital_status days_to_birth     days_to_death days_to_last_followup  \\\n",
       "0         Dead        -21496            1355.0       [Not Available]   \n",
       "1         Dead        -16090              1677       [Not Available]   \n",
       "2        Alive         -8624  [Not Applicable]                2091.0   \n",
       "\n",
       "  days_to_initial_pathologic_diagnosis age_at_initial_pathologic_diagnosis  \\\n",
       "0                                    0                                  58   \n",
       "1                                    0                                  44   \n",
       "2                                    0                                  23   \n",
       "\n",
       "   ... total_bilirubin_upper_limit platelet_result_count fibrosis_ishak_score  \\\n",
       "0  ...                         NaN                   NaN                  NaN   \n",
       "1  ...                         NaN                   NaN                  NaN   \n",
       "2  ...                         NaN                   NaN                  NaN   \n",
       "\n",
       "  fetoprotein_outcome_value fetoprotein_outcome_upper_limit  \\\n",
       "0                       NaN                             NaN   \n",
       "1                       NaN                             NaN   \n",
       "2                       NaN                             NaN   \n",
       "\n",
       "  fetoprotein_outcome_lower_limit inter_norm_ratio_lower_limit  \\\n",
       "0                             NaN                          NaN   \n",
       "1                             NaN                          NaN   \n",
       "2                             NaN                          NaN   \n",
       "\n",
       "  family_cancer_type_txt bilirubin_upper_limit days_to_last_known_alive  \n",
       "0                    NaN                   NaN                      NaN  \n",
       "1                    NaN                   NaN                      NaN  \n",
       "2                    NaN                   NaN                      NaN  \n",
       "\n",
       "[3 rows x 746 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_follow = pd.read_csv(PANCANATLAS + \"clinical_PANCAN_patient_with_followup.tsv\", sep=\"\\t\",\n",
    "                             encoding=\"latin1\")\n",
    "print(clinical_follow.shape)\n",
    "clinical_follow.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcd5fce2-b0e2-4f9b-80fb-9964922aae86",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d9adc131-830c-47f6-8a56-d9446c8e28a7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## 5. Others"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8cc47c6-b024-415e-bc2c-9301935a058c",
   "metadata": {},
   "source": [
    "### 5.1. Quality annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a6728e1a-8c00-42bb-b9a0-4303534a258a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(79286, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_barcode</th>\n",
       "      <th>aliquot_barcode</th>\n",
       "      <th>cancer type</th>\n",
       "      <th>platform</th>\n",
       "      <th>patient_annotation</th>\n",
       "      <th>sample_annotation</th>\n",
       "      <th>aliquot_annotation</th>\n",
       "      <th>aliquot_annotation_updated</th>\n",
       "      <th>AWG_excluded_because_of_pathology</th>\n",
       "      <th>AWG_pathology_exclusion_reason</th>\n",
       "      <th>Reviewed_by_EPC</th>\n",
       "      <th>Do_not_use</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-01-0628</td>\n",
       "      <td>TCGA-01-0628-11A-01D-0356-01</td>\n",
       "      <td>OV</td>\n",
       "      <td>Genome_Wide_SNP_6</td>\n",
       "      <td>Organ-Specific Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-01-0628</td>\n",
       "      <td>TCGA-01-0628-11A-01D-0383-05</td>\n",
       "      <td>OV</td>\n",
       "      <td>HumanMethylation27</td>\n",
       "      <td>Organ-Specific Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-01-0630</td>\n",
       "      <td>TCGA-01-0630-11A-01D-0356-01</td>\n",
       "      <td>OV</td>\n",
       "      <td>Genome_Wide_SNP_6</td>\n",
       "      <td>Organ-Specific Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient_barcode               aliquot_barcode cancer type  \\\n",
       "0    TCGA-01-0628  TCGA-01-0628-11A-01D-0356-01          OV   \n",
       "1    TCGA-01-0628  TCGA-01-0628-11A-01D-0383-05          OV   \n",
       "2    TCGA-01-0630  TCGA-01-0630-11A-01D-0356-01          OV   \n",
       "\n",
       "             platform      patient_annotation sample_annotation  \\\n",
       "0   Genome_Wide_SNP_6  Organ-Specific Control               NaN   \n",
       "1  HumanMethylation27  Organ-Specific Control               NaN   \n",
       "2   Genome_Wide_SNP_6  Organ-Specific Control               NaN   \n",
       "\n",
       "  aliquot_annotation aliquot_annotation_updated  \\\n",
       "0                NaN                        NaN   \n",
       "1                NaN                        NaN   \n",
       "2                NaN                        NaN   \n",
       "\n",
       "   AWG_excluded_because_of_pathology AWG_pathology_exclusion_reason  \\\n",
       "0                                0.0                            NaN   \n",
       "1                                0.0                            NaN   \n",
       "2                                0.0                            NaN   \n",
       "\n",
       "   Reviewed_by_EPC  Do_not_use  \n",
       "0              0.0       False  \n",
       "1              0.0       False  \n",
       "2              0.0       False  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# \n",
    "quality = pd.read_csv(PANCANATLAS + \"merged_sample_quality_annotations.tsv\", sep=\"\\t\")\n",
    "print(quality.shape)\n",
    "quality.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ed2aaea-0dab-4bcf-abc6-d3c45846062b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ecf4e0c-35c3-4e20-b2ee-8bd4dcfaf1f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68e93a3-70a7-479e-a4bf-34c94613657b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778d84d9-6b6f-4507-8a5f-17073bd3cb44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f5f1b21-b896-4eb6-853c-75a4348f0857",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6795a650-d462-4326-a955-12fa0cc51185",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99541ec8-d047-4484-ae35-9aa59ea1f675",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fd6685ca-b781-415e-b74a-3a8999ce1c5d",
   "metadata": {},
   "source": [
    "# 2. Merge metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a57b96e6-a387-4fe2-8f2b-b9469e94ad02",
   "metadata": {},
   "source": [
    "## 2.1. Original metadata 2021.  \n",
    "Metadata per sample (TCGA-S9-A7J2-01). 10535 samples x 234 metadata columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "45e9d6f2-f16b-4a14-87dd-b26671690365",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 234)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>chr13</th>\n",
       "      <th>chr14</th>\n",
       "      <th>chr15</th>\n",
       "      <th>chr16</th>\n",
       "      <th>chr17</th>\n",
       "      <th>chr18</th>\n",
       "      <th>chr19</th>\n",
       "      <th>chr20</th>\n",
       "      <th>chr21</th>\n",
       "      <th>chr22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 234 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                 25.0    MALE  WHITE   \n",
       "1                                 53.0    MALE  ASIAN   \n",
       "2                                 26.0  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  ...  chr13 chr14  \\\n",
       "0                 Oligodendroglioma                 G3  ...    0.0   0.0   \n",
       "1          Hepatocellular Carcinoma                 G2  ...    NaN   NaN   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...    NaN   0.0   \n",
       "\n",
       "   chr15 chr16 chr17  chr18  chr19 chr20 chr21 chr22  \n",
       "0    0.0   0.0   0.0    0.0    NaN   0.0   0.0   0.0  \n",
       "1    NaN   NaN   NaN    NaN    NaN   NaN   NaN   NaN  \n",
       "2    0.0   0.0   0.0    0.0    0.0   0.0   0.0   0.0  \n",
       "\n",
       "[3 rows x 234 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(XENA + \"210301_TCGA_RNAseq_metadata.tsv\", sep=\"\\t\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4615c5d0-b58e-49da-9446-2b34e300b687",
   "metadata": {},
   "source": [
    "## 2.2. Add new metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "6ce60b12-1920-4f68-9442-68337b24e630",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10534"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Basic_phenotype\n",
    "common = metadata[\"Sample_ID\"].isin(basic_phenotype[\"sample\"])\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "3098394e-3536-4474-bd45-fe55c6f79009",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 241)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>chr20</th>\n",
       "      <th>chr21</th>\n",
       "      <th>chr22</th>\n",
       "      <th>sample</th>\n",
       "      <th>program</th>\n",
       "      <th>sample_type_id_y</th>\n",
       "      <th>sample_type_y</th>\n",
       "      <th>project_id</th>\n",
       "      <th>Age at Diagnosis in Years</th>\n",
       "      <th>Gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 241 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                 25.0    MALE  WHITE   \n",
       "1                                 53.0    MALE  ASIAN   \n",
       "2                                 26.0  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  ...  chr20 chr21  \\\n",
       "0                 Oligodendroglioma                 G3  ...    0.0   0.0   \n",
       "1          Hepatocellular Carcinoma                 G2  ...    NaN   NaN   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...    0.0   0.0   \n",
       "\n",
       "   chr22           sample program  sample_type_id_y        sample_type_y  \\\n",
       "0    0.0  TCGA-S9-A7J2-01    TCGA               1.0        Primary Tumor   \n",
       "1    NaN  TCGA-G3-A3CH-11    TCGA              11.0  Solid Tissue Normal   \n",
       "2    0.0  TCGA-EK-A2RE-01    TCGA               1.0        Primary Tumor   \n",
       "\n",
       "  project_id Age at Diagnosis in Years  Gender  \n",
       "0   TCGA-LGG                      25.0    Male  \n",
       "1  TCGA-LIHC                      53.0    Male  \n",
       "2  TCGA-CESC                      26.0  Female  \n",
       "\n",
       "[3 rows x 241 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Basic_phenotype\n",
    "metadata = metadata.merge(basic_phenotype, how=\"left\", left_on=\"Sample_ID\", right_on=\"sample\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "f0623884-a3b0-4846-bee7-ca5c7e7e9b33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10320"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Survival\n",
    "common = metadata[\"Sample_ID\"].isin(survival[\"sample\"])\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "bf9dd510-5e70-42c3-aaf1-ad3e8d649b0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 245)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>program</th>\n",
       "      <th>sample_type_id_y</th>\n",
       "      <th>sample_type_y</th>\n",
       "      <th>project_id</th>\n",
       "      <th>Age at Diagnosis in Years</th>\n",
       "      <th>Gender</th>\n",
       "      <th>sample_y</th>\n",
       "      <th>OS_y</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS.time_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>62.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>780.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 245 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                 25.0    MALE  WHITE   \n",
       "1                                 53.0    MALE  ASIAN   \n",
       "2                                 26.0  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  ...  program  \\\n",
       "0                 Oligodendroglioma                 G3  ...     TCGA   \n",
       "1          Hepatocellular Carcinoma                 G2  ...     TCGA   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...     TCGA   \n",
       "\n",
       "  sample_type_id_y        sample_type_y project_id Age at Diagnosis in Years  \\\n",
       "0              1.0        Primary Tumor   TCGA-LGG                      25.0   \n",
       "1             11.0  Solid Tissue Normal  TCGA-LIHC                      53.0   \n",
       "2              1.0        Primary Tumor  TCGA-CESC                      26.0   \n",
       "\n",
       "   Gender         sample_y OS_y      _PATIENT OS.time_y  \n",
       "0    Male  TCGA-S9-A7J2-01  0.0  TCGA-S9-A7J2      62.0  \n",
       "1    Male  TCGA-G3-A3CH-11  0.0  TCGA-G3-A3CH     780.0  \n",
       "2  Female  TCGA-EK-A2RE-01  0.0  TCGA-EK-A2RE      57.0  \n",
       "\n",
       "[3 rows x 245 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Survival\n",
    "metadata = metadata.merge(survival, how=\"left\", left_on=\"Sample_ID\", right_on=\"sample\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "0f69d892-d320-456b-bdbe-e8dcd77a1102",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10534"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Phenotype\n",
    "common = metadata[\"Sample_ID\"].isin(phenotype[\"sample\"])\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "7e3a9ffc-7b68-4a7b-baab-3291f2725e28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 303)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>exposures.years_smoked</th>\n",
       "      <th>id</th>\n",
       "      <th>project.name</th>\n",
       "      <th>project.project_id</th>\n",
       "      <th>tissue_source_site.name</th>\n",
       "      <th>samples.is_ffpe</th>\n",
       "      <th>samples.sample_id</th>\n",
       "      <th>samples.sample_type</th>\n",
       "      <th>samples.sample_type_id</th>\n",
       "      <th>samples.tissue_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>194e8231-e8d4-477f-8c62-e11f2484ecce</td>\n",
       "      <td>Brain Lower Grade Glioma</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>Dept of Neurosurgery at University of Heidelberg</td>\n",
       "      <td>False</td>\n",
       "      <td>abd56cd9-8545-4874-91b4-c77834f3e3df</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3a2d7482-c246-411e-841f-71d96e353328</td>\n",
       "      <td>Liver Hepatocellular Carcinoma</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Alberta Health Services</td>\n",
       "      <td>False</td>\n",
       "      <td>6b915dde-b8ac-4cc6-839b-e226e914b499</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9aa36ac2-8418-4109-a3c1-63ca8742baab</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma and Endocervi...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Gynecologic Oncology Group</td>\n",
       "      <td>False</td>\n",
       "      <td>7bcf54b6-3ffa-4014-92c6-51a7099e6f02</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 303 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                 25.0    MALE  WHITE   \n",
       "1                                 53.0    MALE  ASIAN   \n",
       "2                                 26.0  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  ...  \\\n",
       "0                 Oligodendroglioma                 G3  ...   \n",
       "1          Hepatocellular Carcinoma                 G2  ...   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...   \n",
       "\n",
       "   exposures.years_smoked                                    id  \\\n",
       "0                     NaN  194e8231-e8d4-477f-8c62-e11f2484ecce   \n",
       "1                     NaN  3a2d7482-c246-411e-841f-71d96e353328   \n",
       "2                     NaN  9aa36ac2-8418-4109-a3c1-63ca8742baab   \n",
       "\n",
       "                                        project.name project.project_id  \\\n",
       "0                           Brain Lower Grade Glioma           TCGA-LGG   \n",
       "1                     Liver Hepatocellular Carcinoma          TCGA-LIHC   \n",
       "2  Cervical Squamous Cell Carcinoma and Endocervi...          TCGA-CESC   \n",
       "\n",
       "                            tissue_source_site.name  samples.is_ffpe  \\\n",
       "0  Dept of Neurosurgery at University of Heidelberg            False   \n",
       "1                           Alberta Health Services            False   \n",
       "2                        Gynecologic Oncology Group            False   \n",
       "\n",
       "                      samples.sample_id  samples.sample_type  \\\n",
       "0  abd56cd9-8545-4874-91b4-c77834f3e3df        Primary Tumor   \n",
       "1  6b915dde-b8ac-4cc6-839b-e226e914b499  Solid Tissue Normal   \n",
       "2  7bcf54b6-3ffa-4014-92c6-51a7099e6f02        Primary Tumor   \n",
       "\n",
       "  samples.sample_type_id samples.tissue_type  \n",
       "0                    1.0        Not Reported  \n",
       "1                   11.0        Not Reported  \n",
       "2                    1.0        Not Reported  \n",
       "\n",
       "[3 rows x 303 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# phenotype\n",
    "metadata = metadata.merge(phenotype, how=\"left\", left_on=\"Sample_ID\", right_on=\"sample\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "01a61efd-1aca-4382-87dd-02244b6c0abf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9461"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Purity\n",
    "common = metadata[\"Sample_ID\"].isin(purity[\"array\"])\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "204f92f5-0717-4a13-b8bb-277fd851f270",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 313)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yw/q_pg1y917pz9nzm99n51ctzh0000gp/T/ipykernel_1379/1044728146.py:2: FutureWarning: Passing 'suffixes' which cause duplicate columns {'sample_x'} in the result is deprecated and will raise a MergeError in a future version.\n",
      "  metadata = metadata.merge(purity, how=\"left\", left_on=\"Sample_ID\", right_on=\"array\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>array</th>\n",
       "      <th>sample_y</th>\n",
       "      <th>call status</th>\n",
       "      <th>purity</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>Genome doublings</th>\n",
       "      <th>Coverage for 80% power</th>\n",
       "      <th>Cancer DNA fraction</th>\n",
       "      <th>Subclonal genome fraction</th>\n",
       "      <th>solution</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2-01A-11D-A349-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.89</td>\n",
       "      <td>3.89</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.01</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE-01A-11D-A18H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.86</td>\n",
       "      <td>1.97</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 313 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                 25.0    MALE  WHITE   \n",
       "1                                 53.0    MALE  ASIAN   \n",
       "2                                 26.0  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  ...            array  \\\n",
       "0                 Oligodendroglioma                 G3  ...  TCGA-S9-A7J2-01   \n",
       "1          Hepatocellular Carcinoma                 G2  ...              NaN   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...  TCGA-EK-A2RE-01   \n",
       "\n",
       "                       sample_y  call status purity ploidy  Genome doublings  \\\n",
       "0  TCGA-S9-A7J2-01A-11D-A349-01       called   0.89   3.89               1.0   \n",
       "1                           NaN          NaN    NaN    NaN               NaN   \n",
       "2  TCGA-EK-A2RE-01A-11D-A18H-01       called   0.86   1.97               0.0   \n",
       "\n",
       "   Coverage for 80% power Cancer DNA fraction Subclonal genome fraction  \\\n",
       "0                    17.0                0.94                      0.01   \n",
       "1                     NaN                 NaN                       NaN   \n",
       "2                     9.0                0.85                      0.00   \n",
       "\n",
       "  solution  \n",
       "0      new  \n",
       "1      NaN  \n",
       "2      new  \n",
       "\n",
       "[3 rows x 313 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# purity\n",
    "metadata = metadata.merge(purity, how=\"left\", left_on=\"Sample_ID\", right_on=\"array\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "82ce01f4-dada-49bf-8a7a-88c28c6db1e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PARADIGM & GISTIC not included... too large"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "371f2141-ffb7-4697-b0e7-34e72d318fae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10496"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clinical\n",
    "common = metadata[\"Patient_ID\"].isin(clinical[\"bcr_patient_barcode\"])\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "9af442a3-82f1-497d-b4e5-07c91ad32bfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 332)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>DSS_cr</th>\n",
       "      <th>DSS.time.cr</th>\n",
       "      <th>DFI.cr</th>\n",
       "      <th>DFI.time.cr</th>\n",
       "      <th>PFI.cr</th>\n",
       "      <th>PFI.time.cr</th>\n",
       "      <th>PFI.1.cr</th>\n",
       "      <th>PFI.time.1.cr</th>\n",
       "      <th>PFI.2.cr</th>\n",
       "      <th>PFI.time.2.cr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>#N/D</td>\n",
       "      <td>#N/D</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>780</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>#N/D</td>\n",
       "      <td>#N/D</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 332 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                 25.0    MALE  WHITE   \n",
       "1                                 53.0    MALE  ASIAN   \n",
       "2                                 26.0  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  ...  DSS_cr  \\\n",
       "0                 Oligodendroglioma                 G3  ...       0   \n",
       "1          Hepatocellular Carcinoma                 G2  ...       0   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...       0   \n",
       "\n",
       "  DSS.time.cr  DFI.cr DFI.time.cr PFI.cr  PFI.time.cr  PFI.1.cr PFI.time.1.cr  \\\n",
       "0          62    #N/D        #N/D      0           62         0            62   \n",
       "1         780       1         116      1          116         1           116   \n",
       "2          57    #N/D        #N/D      0           57         0            57   \n",
       "\n",
       "  PFI.2.cr PFI.time.2.cr  \n",
       "0        0            62  \n",
       "1        1           116  \n",
       "2        0            57  \n",
       "\n",
       "[3 rows x 332 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clinical\n",
    "metadata = metadata.merge(clinical, how=\"left\", left_on=\"Patient_ID\", right_on=\"bcr_patient_barcode\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "da5dfe4a-c3b1-4693-a0c9-5070de55f339",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10496"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clinical with followup\n",
    "common = metadata[\"Patient_ID\"].isin(clinical[\"bcr_patient_barcode\"])\n",
    "common.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "d65f6a65-97fa-4dd6-849b-b529183f1fbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10535, 1078)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis_x</th>\n",
       "      <th>gender_x</th>\n",
       "      <th>race_x</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage_x</th>\n",
       "      <th>histological_type_x</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>...</th>\n",
       "      <th>total_bilirubin_upper_limit</th>\n",
       "      <th>platelet_result_count</th>\n",
       "      <th>fibrosis_ishak_score</th>\n",
       "      <th>fetoprotein_outcome_value</th>\n",
       "      <th>fetoprotein_outcome_upper_limit</th>\n",
       "      <th>fetoprotein_outcome_lower_limit</th>\n",
       "      <th>inter_norm_ratio_lower_limit</th>\n",
       "      <th>family_cancer_type_txt</th>\n",
       "      <th>bilirubin_upper_limit</th>\n",
       "      <th>days_to_last_known_alive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>26.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1078 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC   \n",
       "\n",
       "   age_at_initial_pathologic_diagnosis_x gender_x race_x  \\\n",
       "0                                   25.0     MALE  WHITE   \n",
       "1                                   53.0     MALE  ASIAN   \n",
       "2                                   26.0   FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage_x  \\\n",
       "0                         NaN              NaN   \n",
       "1                  Stage IIIA              NaN   \n",
       "2                         NaN        Stage IIA   \n",
       "\n",
       "                histological_type_x histological_grade  ...  \\\n",
       "0                 Oligodendroglioma                 G3  ...   \n",
       "1          Hepatocellular Carcinoma                 G2  ...   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2  ...   \n",
       "\n",
       "   total_bilirubin_upper_limit platelet_result_count  fibrosis_ishak_score  \\\n",
       "0                          NaN                   NaN                   NaN   \n",
       "1                          NaN                   NaN                   NaN   \n",
       "2                          NaN                   NaN                   NaN   \n",
       "\n",
       "  fetoprotein_outcome_value fetoprotein_outcome_upper_limit  \\\n",
       "0                       NaN                             NaN   \n",
       "1                       NaN                             NaN   \n",
       "2                       NaN                             NaN   \n",
       "\n",
       "   fetoprotein_outcome_lower_limit  inter_norm_ratio_lower_limit  \\\n",
       "0                              NaN                           NaN   \n",
       "1                              NaN                           NaN   \n",
       "2                              NaN                           NaN   \n",
       "\n",
       "  family_cancer_type_txt bilirubin_upper_limit days_to_last_known_alive  \n",
       "0                    NaN                   NaN                      NaN  \n",
       "1                    NaN                   NaN                      NaN  \n",
       "2                    NaN                   NaN                      NaN  \n",
       "\n",
       "[3 rows x 1078 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clinical with followup\n",
    "metadata = metadata.merge(clinical_follow, how=\"left\", left_on=\"Patient_ID\", right_on=\"bcr_patient_barcode\")\n",
    "print(metadata.shape)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "43e5071b-0b16-4c2f-b266-96f41e36eee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.to_csv(DESKTOP + \"251225_TCGA_metadata.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d76ade24-1f1c-43ca-ab2b-621488222904",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "25ba8b02-90d3-4e19-a9f2-3058ef5449b6",
   "metadata": {},
   "source": [
    "## 2.3. Add mutant genes from XENA SNP MC3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ee6242e-d179-4d75-b3aa-269472cb2f7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_BULLKpy = \"/Users/mmalumbres/Library/CloudStorage/OneDrive-VHIO/BioInformatics/BioProjects/MM01_BULLKpy/bullkpy-skeleton/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9a9a5bcd-04e5-4a5b-85d3-15bf465e5217",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10534, 1085)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>initial_pathologic_dx_year</th>\n",
       "      <th>...</th>\n",
       "      <th>kmeans_20</th>\n",
       "      <th>kmeans_25</th>\n",
       "      <th>kmeans_30</th>\n",
       "      <th>Proliferation_score</th>\n",
       "      <th>T_cell_score</th>\n",
       "      <th>Neuroendocrine_score</th>\n",
       "      <th>score_prolif</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>27</td>\n",
       "      <td>-1.281150</td>\n",
       "      <td>-3.294353</td>\n",
       "      <td>2.421188</td>\n",
       "      <td>-2.581529</td>\n",
       "      <td>-0.847903</td>\n",
       "      <td>-0.972469</td>\n",
       "      <td>G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>-1.226665</td>\n",
       "      <td>0.957415</td>\n",
       "      <td>-0.122859</td>\n",
       "      <td>-3.387491</td>\n",
       "      <td>-0.637804</td>\n",
       "      <td>-1.100115</td>\n",
       "      <td>G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>1.434147</td>\n",
       "      <td>-0.939168</td>\n",
       "      <td>-0.041094</td>\n",
       "      <td>2.428172</td>\n",
       "      <td>1.475654</td>\n",
       "      <td>1.156046</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1085 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  gender   race  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG    MALE  WHITE   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC    MALE  ASIAN   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  \\\n",
       "0                 Oligodendroglioma                 G3   \n",
       "1          Hepatocellular Carcinoma                 G2   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2   \n",
       "\n",
       "   initial_pathologic_dx_year  ... kmeans_20  kmeans_25 kmeans_30  \\\n",
       "0                      2013.0  ...        10         13        27   \n",
       "1                      2010.0  ...        13          6        13   \n",
       "2                      2010.0  ...         9         15         5   \n",
       "\n",
       "  Proliferation_score  T_cell_score  Neuroendocrine_score score_prolif  \\\n",
       "0           -1.281150     -3.294353              2.421188    -2.581529   \n",
       "1           -1.226665      0.957415             -0.122859    -3.387491   \n",
       "2            1.434147     -0.939168             -0.041094     2.428172   \n",
       "\n",
       "    S_score G2M_score phase  \n",
       "0 -0.847903 -0.972469    G1  \n",
       "1 -0.637804 -1.100115    G1  \n",
       "2  1.475654  1.156046     S  \n",
       "\n",
       "[3 rows x 1085 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_final = pd.read_csv(TCGA_BULLKpy + \"251225_TCGA_metadata.tsv\", sep=\"\\t\",\n",
    "                            low_memory = False)\n",
    "print(metadata_final.shape)\n",
    "metadata_final.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "da61c79e-af0e-4732-ac70-769b5f83e33f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>UBE2Q2</th>\n",
       "      <th>CHMP1B</th>\n",
       "      <th>PSMA2P1</th>\n",
       "      <th>SHQ1P1</th>\n",
       "      <th>CPHL1P</th>\n",
       "      <th>SSXP10</th>\n",
       "      <th>REM1</th>\n",
       "      <th>TCOF1</th>\n",
       "      <th>NSRP1</th>\n",
       "      <th>OPA6</th>\n",
       "      <th>...</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>OR1E5</th>\n",
       "      <th>RP11-390F4.3</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>PTRF</th>\n",
       "      <th>DIAPH2-AS1</th>\n",
       "      <th>SELV</th>\n",
       "      <th>NFIX</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0003-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0033-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 40543 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample           UBE2Q2  CHMP1B  PSMA2P1  SHQ1P1  CPHL1P  SSXP10  REM1  TCOF1  \\\n",
       "TCGA-02-0003-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0033-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0047-01       0       0        0       0       0       0     0      0   \n",
       "\n",
       "sample           NSRP1  OPA6  ...  TULP2  OR1E5  RP11-390F4.3  GNGT2  GNGT1  \\\n",
       "TCGA-02-0003-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0033-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0047-01      0     0  ...      0      0             0      0      0   \n",
       "\n",
       "sample           PTRF  DIAPH2-AS1  SELV  NFIX  SELP  \n",
       "TCGA-02-0003-01     0           0     0     0     0  \n",
       "TCGA-02-0033-01     0           0     0     0     0  \n",
       "TCGA-02-0047-01     0           0     0     0     0  \n",
       "\n",
       "[3 rows x 40543 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc3.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "14a31ff5-1e8e-4e63-b60c-56f8fb4f9a66",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_genes_lung = [\"TP53\", \"RB1\", \"KRAS\", \"EGFR\", \"KEAP1\", \"NFE2L2\", \"STK11\", \"PIK3CA\",\n",
    "                 \"SMARCA4\", \"BRAF\", \"PTEN\", \"CDKN2A\", \"CCND1\", \"NOTCH1\", \"CREBBP\", \"EP300\",\n",
    "                 \"KMT2D\", \"KMT2C\", \"NOTCH1\", \"NOTCH2\", \"NOTCH3\", \"SLIT2\", \"MYC\", \"MYCL\", \"MYCN\",\n",
    "                 \"SOX2\", \"ASCL1\", \"NEUROD1\", \"POU2F3\", \"YAP1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "92e5f755-9c11-41fb-a19a-d4a57bb679e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9104, 30)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>TP53</th>\n",
       "      <th>RB1</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>KEAP1</th>\n",
       "      <th>NFE2L2</th>\n",
       "      <th>STK11</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>SMARCA4</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>...</th>\n",
       "      <th>NOTCH3</th>\n",
       "      <th>SLIT2</th>\n",
       "      <th>MYC</th>\n",
       "      <th>MYCL</th>\n",
       "      <th>MYCN</th>\n",
       "      <th>SOX2</th>\n",
       "      <th>ASCL1</th>\n",
       "      <th>NEUROD1</th>\n",
       "      <th>POU2F3</th>\n",
       "      <th>YAP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0003-01</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0033-01</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample           TP53  RB1  KRAS  EGFR  KEAP1  NFE2L2  STK11  PIK3CA  SMARCA4  \\\n",
       "TCGA-02-0003-01     1    0     0     1      0       0      0       0        0   \n",
       "TCGA-02-0033-01     1    1     0     0      0       0      0       1        0   \n",
       "TCGA-02-0047-01     0    0     0     0      0       0      0       1        0   \n",
       "\n",
       "sample           BRAF  ...  NOTCH3  SLIT2  MYC  MYCL  MYCN  SOX2  ASCL1  \\\n",
       "TCGA-02-0003-01     0  ...       0      0    0     0     0     0      0   \n",
       "TCGA-02-0033-01     0  ...       0      0    0     0     0     0      0   \n",
       "TCGA-02-0047-01     0  ...       0      0    0     0     0     0      0   \n",
       "\n",
       "sample           NEUROD1  POU2F3  YAP1  \n",
       "TCGA-02-0003-01        0       0     0  \n",
       "TCGA-02-0033-01        0       0     0  \n",
       "TCGA-02-0047-01        0       0     0  \n",
       "\n",
       "[3 rows x 30 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc3_genes_lung = mc3[list_genes_lung]\n",
    "print(mc3_genes_lung.shape)\n",
    "mc3_genes_lung.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7635f8db-3d4d-4fed-abfd-a371c6d5cf6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mc3_genes_lung.to_csv(TCGA_BULLKpy + \"251225_TCGA_lung_mutant_genes.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7d192133-a949-4252-8317-bdc730fb439b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10534, 1115)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>clinical_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>histological_grade</th>\n",
       "      <th>initial_pathologic_dx_year</th>\n",
       "      <th>...</th>\n",
       "      <th>NOTCH3</th>\n",
       "      <th>SLIT2</th>\n",
       "      <th>MYC</th>\n",
       "      <th>MYCL</th>\n",
       "      <th>MYCN</th>\n",
       "      <th>SOX2</th>\n",
       "      <th>ASCL1</th>\n",
       "      <th>NEUROD1</th>\n",
       "      <th>POU2F3</th>\n",
       "      <th>YAP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-S9-A7J2-01</td>\n",
       "      <td>TCGA-S9-A7J2</td>\n",
       "      <td>LGG</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>G3</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-G3-A3CH-11</td>\n",
       "      <td>TCGA-G3-A3CH</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>MALE</td>\n",
       "      <td>ASIAN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocellular Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EK-A2RE-01</td>\n",
       "      <td>TCGA-EK-A2RE</td>\n",
       "      <td>CESC</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Cervical Squamous Cell Carcinoma</td>\n",
       "      <td>G2</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1115 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample_ID    Patient_ID Project_ID  gender   race  \\\n",
       "0  TCGA-S9-A7J2-01  TCGA-S9-A7J2        LGG    MALE  WHITE   \n",
       "1  TCGA-G3-A3CH-11  TCGA-G3-A3CH       LIHC    MALE  ASIAN   \n",
       "2  TCGA-EK-A2RE-01  TCGA-EK-A2RE       CESC  FEMALE  WHITE   \n",
       "\n",
       "  ajcc_pathologic_tumor_stage clinical_stage  \\\n",
       "0                         NaN            NaN   \n",
       "1                  Stage IIIA            NaN   \n",
       "2                         NaN      Stage IIA   \n",
       "\n",
       "                  histological_type histological_grade  \\\n",
       "0                 Oligodendroglioma                 G3   \n",
       "1          Hepatocellular Carcinoma                 G2   \n",
       "2  Cervical Squamous Cell Carcinoma                 G2   \n",
       "\n",
       "   initial_pathologic_dx_year  ... NOTCH3  SLIT2  MYC MYCL  MYCN  SOX2 ASCL1  \\\n",
       "0                      2013.0  ...    0.0    0.0  0.0  0.0   0.0   0.0   0.0   \n",
       "1                      2010.0  ...    NaN    NaN  NaN  NaN   NaN   NaN   NaN   \n",
       "2                      2010.0  ...    0.0    0.0  0.0  0.0   0.0   0.0   0.0   \n",
       "\n",
       "  NEUROD1 POU2F3 YAP1  \n",
       "0     0.0    0.0  0.0  \n",
       "1     NaN    NaN  NaN  \n",
       "2     0.0    0.0  0.0  \n",
       "\n",
       "[3 rows x 1115 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata with 30 lung cancer mutations\n",
    "metadata_final2 = metadata_final.merge(mc3_genes_lung, how=\"left\", left_on=\"Sample_ID\", right_index=True)\n",
    "print(metadata_final2.shape)\n",
    "metadata_final2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e12a9cb8-e0e2-4f6f-9df3-6967a3c64613",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_final2.to_csv(TCGA_BULLKpy + \"251225_TCGA_metadata_lung_mutant_genes.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c7f2b1e-3516-4de0-9055-e89d76f3cee1",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# PENDING\n",
    "2.2. Add mutations at the gene level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "77233849-6d2b-4e59-9e63-91948f7d0ac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <th>ENSG00000000457.12</th>\n",
       "      <th>ENSG00000000460.15</th>\n",
       "      <th>ENSG00000000938.11</th>\n",
       "      <th>ENSG00000000971.14</th>\n",
       "      <th>ENSG00000001036.12</th>\n",
       "      <th>ENSG00000001084.9</th>\n",
       "      <th>ENSG00000001167.13</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000281760.1</th>\n",
       "      <th>ENSG00000281781.1</th>\n",
       "      <th>ENSG00000281817.1</th>\n",
       "      <th>ENSG00000281844.1</th>\n",
       "      <th>ENSG00000281855.1</th>\n",
       "      <th>ENSG00000281873.1</th>\n",
       "      <th>ENSG00000281883.1</th>\n",
       "      <th>ENSG00000281887.1</th>\n",
       "      <th>ENSG00000281889.1</th>\n",
       "      <th>ENSG00000281899.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KO-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KZ-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LA-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LP-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LR-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-V4-A9EU-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-V4-A9F1-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-WC-A885-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-WC-A87W-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-YZ-A983-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11368 rows × 19729 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample            ENSG00000000003.13  ENSG00000000005.5  ENSG00000000419.11  \\\n",
       "TCGA-OR-A5KO-01A                   0                  0                   0   \n",
       "TCGA-OR-A5KZ-01A                   0                  0                   0   \n",
       "TCGA-OR-A5LA-01A                   0                  0                   0   \n",
       "TCGA-OR-A5LP-01A                   0                  0                   0   \n",
       "TCGA-OR-A5LR-01A                   0                  0                   0   \n",
       "...                              ...                ...                 ...   \n",
       "TCGA-V4-A9EU-01A                   0                  0                   0   \n",
       "TCGA-V4-A9F1-01A                   0                  0                   0   \n",
       "TCGA-WC-A885-01A                   0                  0                   0   \n",
       "TCGA-WC-A87W-01A                   0                  0                   0   \n",
       "TCGA-YZ-A983-01A                   0                  0                   0   \n",
       "\n",
       "sample            ENSG00000000457.12  ENSG00000000460.15  ENSG00000000938.11  \\\n",
       "TCGA-OR-A5KO-01A                   0                   0                   0   \n",
       "TCGA-OR-A5KZ-01A                   0                   0                   0   \n",
       "TCGA-OR-A5LA-01A                   0                   0                   0   \n",
       "TCGA-OR-A5LP-01A                   0                   0                   0   \n",
       "TCGA-OR-A5LR-01A                   0                   0                   0   \n",
       "...                              ...                 ...                 ...   \n",
       "TCGA-V4-A9EU-01A                   0                   0                   0   \n",
       "TCGA-V4-A9F1-01A                   0                   0                   0   \n",
       "TCGA-WC-A885-01A                   0                   0                   0   \n",
       "TCGA-WC-A87W-01A                   0                   0                   0   \n",
       "TCGA-YZ-A983-01A                   0                   0                   0   \n",
       "\n",
       "sample            ENSG00000000971.14  ENSG00000001036.12  ENSG00000001084.9  \\\n",
       "TCGA-OR-A5KO-01A                   0                   1                  0   \n",
       "TCGA-OR-A5KZ-01A                   0                   0                  0   \n",
       "TCGA-OR-A5LA-01A                   0                   0                  0   \n",
       "TCGA-OR-A5LP-01A                   0                   0                  0   \n",
       "TCGA-OR-A5LR-01A                   0                   0                  0   \n",
       "...                              ...                 ...                ...   \n",
       "TCGA-V4-A9EU-01A                   0                   0                  0   \n",
       "TCGA-V4-A9F1-01A                   0                   0                  0   \n",
       "TCGA-WC-A885-01A                   0                   0                  0   \n",
       "TCGA-WC-A87W-01A                   0                   0                  0   \n",
       "TCGA-YZ-A983-01A                   0                   0                  0   \n",
       "\n",
       "sample            ENSG00000001167.13  ...  ENSG00000281760.1  \\\n",
       "TCGA-OR-A5KO-01A                   0  ...                  0   \n",
       "TCGA-OR-A5KZ-01A                   0  ...                  0   \n",
       "TCGA-OR-A5LA-01A                   0  ...                  0   \n",
       "TCGA-OR-A5LP-01A                   0  ...                  0   \n",
       "TCGA-OR-A5LR-01A                   0  ...                  0   \n",
       "...                              ...  ...                ...   \n",
       "TCGA-V4-A9EU-01A                   0  ...                  0   \n",
       "TCGA-V4-A9F1-01A                   0  ...                  0   \n",
       "TCGA-WC-A885-01A                   1  ...                  0   \n",
       "TCGA-WC-A87W-01A                   0  ...                  0   \n",
       "TCGA-YZ-A983-01A                   0  ...                  0   \n",
       "\n",
       "sample            ENSG00000281781.1  ENSG00000281817.1  ENSG00000281844.1  \\\n",
       "TCGA-OR-A5KO-01A                  0                  1                  0   \n",
       "TCGA-OR-A5KZ-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LA-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LP-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LR-01A                  0                  0                  0   \n",
       "...                             ...                ...                ...   \n",
       "TCGA-V4-A9EU-01A                  0                  0                  0   \n",
       "TCGA-V4-A9F1-01A                  0                  0                  0   \n",
       "TCGA-WC-A885-01A                  0                  0                  1   \n",
       "TCGA-WC-A87W-01A                  0                  0                  0   \n",
       "TCGA-YZ-A983-01A                  0                  0                  0   \n",
       "\n",
       "sample            ENSG00000281855.1  ENSG00000281873.1  ENSG00000281883.1  \\\n",
       "TCGA-OR-A5KO-01A                  0                  0                  0   \n",
       "TCGA-OR-A5KZ-01A                  0                  1                  0   \n",
       "TCGA-OR-A5LA-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LP-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LR-01A                  0                  0                  0   \n",
       "...                             ...                ...                ...   \n",
       "TCGA-V4-A9EU-01A                  0                  0                  0   \n",
       "TCGA-V4-A9F1-01A                  0                  0                  0   \n",
       "TCGA-WC-A885-01A                  1                  0                  0   \n",
       "TCGA-WC-A87W-01A                  0                  0                  0   \n",
       "TCGA-YZ-A983-01A                  0                  0                  0   \n",
       "\n",
       "sample            ENSG00000281887.1  ENSG00000281889.1  ENSG00000281899.1  \n",
       "TCGA-OR-A5KO-01A                  0                  0                  0  \n",
       "TCGA-OR-A5KZ-01A                  0                  0                  0  \n",
       "TCGA-OR-A5LA-01A                  0                  0                  0  \n",
       "TCGA-OR-A5LP-01A                  0                  0                  0  \n",
       "TCGA-OR-A5LR-01A                  0                  0                  0  \n",
       "...                             ...                ...                ...  \n",
       "TCGA-V4-A9EU-01A                  0                  0                  0  \n",
       "TCGA-V4-A9F1-01A                  0                  0                  0  \n",
       "TCGA-WC-A885-01A                  0                  0                  0  \n",
       "TCGA-WC-A87W-01A                  0                  0                  0  \n",
       "TCGA-YZ-A983-01A                  0                  0                  0  \n",
       "\n",
       "[11368 rows x 19729 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gistic2 = gistic.rename(columns={\"Unnamed: 0\": \"sample\"}).set_index(\"sample\").T\n",
    "gistic2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "cbb341ac-5ce7-4653-890c-c971256cb909",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11368, 11939)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <th>ENSG00000000457.12</th>\n",
       "      <th>ENSG00000000460.15</th>\n",
       "      <th>ENSG00000000971.14</th>\n",
       "      <th>ENSG00000001084.9</th>\n",
       "      <th>ENSG00000001167.13</th>\n",
       "      <th>ENSG00000001497.15</th>\n",
       "      <th>ENSG00000001561.6</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000281656.1</th>\n",
       "      <th>ENSG00000281674.1</th>\n",
       "      <th>ENSG00000281676.1</th>\n",
       "      <th>ENSG00000281700.1</th>\n",
       "      <th>ENSG00000281741.1</th>\n",
       "      <th>ENSG00000281781.1</th>\n",
       "      <th>ENSG00000281817.1</th>\n",
       "      <th>ENSG00000281844.1</th>\n",
       "      <th>ENSG00000281855.1</th>\n",
       "      <th>ENSG00000281889.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KO-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KZ-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LA-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11939 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample            ENSG00000000003.13  ENSG00000000005.5  ENSG00000000419.11  \\\n",
       "TCGA-OR-A5KO-01A                   0                  0                   0   \n",
       "TCGA-OR-A5KZ-01A                   0                  0                   0   \n",
       "TCGA-OR-A5LA-01A                   0                  0                   0   \n",
       "\n",
       "sample            ENSG00000000457.12  ENSG00000000460.15  ENSG00000000971.14  \\\n",
       "TCGA-OR-A5KO-01A                   0                   0                   0   \n",
       "TCGA-OR-A5KZ-01A                   0                   0                   0   \n",
       "TCGA-OR-A5LA-01A                   0                   0                   0   \n",
       "\n",
       "sample            ENSG00000001084.9  ENSG00000001167.13  ENSG00000001497.15  \\\n",
       "TCGA-OR-A5KO-01A                  0                   0                   0   \n",
       "TCGA-OR-A5KZ-01A                  0                   0                   0   \n",
       "TCGA-OR-A5LA-01A                  0                   0                   0   \n",
       "\n",
       "sample            ENSG00000001561.6  ...  ENSG00000281656.1  \\\n",
       "TCGA-OR-A5KO-01A                  0  ...                  0   \n",
       "TCGA-OR-A5KZ-01A                  0  ...                  0   \n",
       "TCGA-OR-A5LA-01A                  0  ...                  0   \n",
       "\n",
       "sample            ENSG00000281674.1  ENSG00000281676.1  ENSG00000281700.1  \\\n",
       "TCGA-OR-A5KO-01A                  0                  0                  0   \n",
       "TCGA-OR-A5KZ-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LA-01A                  0                  0                  0   \n",
       "\n",
       "sample            ENSG00000281741.1  ENSG00000281781.1  ENSG00000281817.1  \\\n",
       "TCGA-OR-A5KO-01A                  0                  0                  1   \n",
       "TCGA-OR-A5KZ-01A                  0                  0                  0   \n",
       "TCGA-OR-A5LA-01A                  0                  0                  0   \n",
       "\n",
       "sample            ENSG00000281844.1  ENSG00000281855.1  ENSG00000281889.1  \n",
       "TCGA-OR-A5KO-01A                  0                  0                  0  \n",
       "TCGA-OR-A5KZ-01A                  0                  0                  0  \n",
       "TCGA-OR-A5LA-01A                  0                  0                  0  \n",
       "\n",
       "[3 rows x 11939 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gistic_clean = gistic2.loc[:, gistic2.sum(axis=0) > 0]\n",
    "print(gistic_clean.shape)\n",
    "gistic_clean.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f613715-49fa-460a-acf7-d337a90d8310",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d86f8d-6d64-4585-8ff4-ec35be6973d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d613ab4-a1ed-4d96-ac53-da171b7e29fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90c179f6-9bd8-41b5-bb32-ed6b8aec1952",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "6532c146-f5ec-46b6-a8c2-0cc058495acb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001497</th>\n",
       "      <th>ENSG00000001561</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000281656</th>\n",
       "      <th>ENSG00000281674</th>\n",
       "      <th>ENSG00000281676</th>\n",
       "      <th>ENSG00000281700</th>\n",
       "      <th>ENSG00000281741</th>\n",
       "      <th>ENSG00000281781</th>\n",
       "      <th>ENSG00000281817</th>\n",
       "      <th>ENSG00000281844</th>\n",
       "      <th>ENSG00000281855</th>\n",
       "      <th>ENSG00000281889</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KO-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KZ-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LA-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 11939 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample            ENSG00000000003  ENSG00000000005  ENSG00000000419  \\\n",
       "TCGA-OR-A5KO-01A                0                0                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000000457  ENSG00000000460  ENSG00000000971  \\\n",
       "TCGA-OR-A5KO-01A                0                0                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000001084  ENSG00000001167  ENSG00000001497  \\\n",
       "TCGA-OR-A5KO-01A                0                0                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000001561  ...  ENSG00000281656  ENSG00000281674  \\\n",
       "TCGA-OR-A5KO-01A                0  ...                0                0   \n",
       "TCGA-OR-A5KZ-01A                0  ...                0                0   \n",
       "TCGA-OR-A5LA-01A                0  ...                0                0   \n",
       "\n",
       "sample            ENSG00000281676  ENSG00000281700  ENSG00000281741  \\\n",
       "TCGA-OR-A5KO-01A                0                0                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000281781  ENSG00000281817  ENSG00000281844  \\\n",
       "TCGA-OR-A5KO-01A                0                1                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000281855  ENSG00000281889  \n",
       "TCGA-OR-A5KO-01A                0                0  \n",
       "TCGA-OR-A5KZ-01A                0                0  \n",
       "TCGA-OR-A5LA-01A                0                0  \n",
       "\n",
       "[3 rows x 11939 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gistic_clean.columns = gistic_clean.columns.str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
    "gistic_clean.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "14ed0f01-ba1a-4038-a66c-0e26674b3213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>FIRRM</th>\n",
       "      <th>CFH</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>LAS1L</th>\n",
       "      <th>ENPP4</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000281656</th>\n",
       "      <th>ENSG00000281674</th>\n",
       "      <th>ENSG00000281676</th>\n",
       "      <th>ENSG00000281700</th>\n",
       "      <th>ENSG00000281741</th>\n",
       "      <th>ENSG00000281781</th>\n",
       "      <th>ENSG00000281817</th>\n",
       "      <th>ENSG00000281844</th>\n",
       "      <th>ENSG00000281855</th>\n",
       "      <th>ENSG00000281889</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KO-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5KZ-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LA-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LP-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LR-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-V4-A9EU-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-V4-A9F1-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-WC-A885-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-WC-A87W-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-YZ-A983-01A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11368 rows × 11939 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample            TSPAN6  TNMD  DPM1  SCYL3  FIRRM  CFH  GCLC  NFYA  LAS1L  \\\n",
       "TCGA-OR-A5KO-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-OR-A5KZ-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-OR-A5LA-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-OR-A5LP-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-OR-A5LR-01A       0     0     0      0      0    0     0     0      0   \n",
       "...                  ...   ...   ...    ...    ...  ...   ...   ...    ...   \n",
       "TCGA-V4-A9EU-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-V4-A9F1-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-WC-A885-01A       0     0     0      0      0    0     0     1      0   \n",
       "TCGA-WC-A87W-01A       0     0     0      0      0    0     0     0      0   \n",
       "TCGA-YZ-A983-01A       0     0     0      0      0    0     0     0      0   \n",
       "\n",
       "sample            ENPP4  ...  ENSG00000281656  ENSG00000281674  \\\n",
       "TCGA-OR-A5KO-01A      0  ...                0                0   \n",
       "TCGA-OR-A5KZ-01A      0  ...                0                0   \n",
       "TCGA-OR-A5LA-01A      0  ...                0                0   \n",
       "TCGA-OR-A5LP-01A      0  ...                0                0   \n",
       "TCGA-OR-A5LR-01A      0  ...                0                0   \n",
       "...                 ...  ...              ...              ...   \n",
       "TCGA-V4-A9EU-01A      0  ...                0                0   \n",
       "TCGA-V4-A9F1-01A      0  ...                0                0   \n",
       "TCGA-WC-A885-01A      1  ...                0                1   \n",
       "TCGA-WC-A87W-01A      0  ...                0                0   \n",
       "TCGA-YZ-A983-01A      0  ...                0                0   \n",
       "\n",
       "sample            ENSG00000281676  ENSG00000281700  ENSG00000281741  \\\n",
       "TCGA-OR-A5KO-01A                0                0                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "TCGA-OR-A5LP-01A                0                0                0   \n",
       "TCGA-OR-A5LR-01A                0                0                0   \n",
       "...                           ...              ...              ...   \n",
       "TCGA-V4-A9EU-01A                0                0                0   \n",
       "TCGA-V4-A9F1-01A                0                0                0   \n",
       "TCGA-WC-A885-01A                0                0                0   \n",
       "TCGA-WC-A87W-01A                0                0                0   \n",
       "TCGA-YZ-A983-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000281781  ENSG00000281817  ENSG00000281844  \\\n",
       "TCGA-OR-A5KO-01A                0                1                0   \n",
       "TCGA-OR-A5KZ-01A                0                0                0   \n",
       "TCGA-OR-A5LA-01A                0                0                0   \n",
       "TCGA-OR-A5LP-01A                0                0                0   \n",
       "TCGA-OR-A5LR-01A                0                0                0   \n",
       "...                           ...              ...              ...   \n",
       "TCGA-V4-A9EU-01A                0                0                0   \n",
       "TCGA-V4-A9F1-01A                0                0                0   \n",
       "TCGA-WC-A885-01A                0                0                1   \n",
       "TCGA-WC-A87W-01A                0                0                0   \n",
       "TCGA-YZ-A983-01A                0                0                0   \n",
       "\n",
       "sample            ENSG00000281855  ENSG00000281889  \n",
       "TCGA-OR-A5KO-01A                0                0  \n",
       "TCGA-OR-A5KZ-01A                0                0  \n",
       "TCGA-OR-A5LA-01A                0                0  \n",
       "TCGA-OR-A5LP-01A                0                0  \n",
       "TCGA-OR-A5LR-01A                0                0  \n",
       "...                           ...              ...  \n",
       "TCGA-V4-A9EU-01A                0                0  \n",
       "TCGA-V4-A9F1-01A                0                0  \n",
       "TCGA-WC-A885-01A                1                0  \n",
       "TCGA-WC-A87W-01A                0                0  \n",
       "TCGA-YZ-A983-01A                0                0  \n",
       "\n",
       "[11368 rows x 11939 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gistic_FINAL = gistic_clean.rename(columns=mapping)\n",
    "gistic_FINAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "b2647448-dd91-4776-af5e-82e5bbe7a7eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gistic_FINAL.to_csv(DESKTOP + \"gistic.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "f9361a16-fa1b-4893-9df3-84f5cc1c40b1",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Index(['TP53', 'RB1'], dtype='object', name='sample')] are in the [columns]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[86], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m gistic_FINAL_selected \u001b[38;5;241m=\u001b[39m \u001b[43mgistic_FINAL\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTP53\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mRB1\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m      2\u001b[0m gistic_FINAL_selected\n",
      "File \u001b[0;32m~/mambaforge/envs/biopython/lib/python3.10/site-packages/pandas/core/frame.py:3813\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3811\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3812\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 3813\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3815\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/mambaforge/envs/biopython/lib/python3.10/site-packages/pandas/core/indexes/base.py:6070\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6067\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6068\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6070\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6072\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6073\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6074\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/biopython/lib/python3.10/site-packages/pandas/core/indexes/base.py:6130\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6128\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m use_interval_msg:\n\u001b[1;32m   6129\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 6130\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6132\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[1;32m   6133\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of [Index(['TP53', 'RB1'], dtype='object', name='sample')] are in the [columns]\""
     ]
    }
   ],
   "source": [
    "gistic_FINAL_selected = gistic_FINAL[[\"TP53\", \"RB1\"]]\n",
    "gistic_FINAL_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4617666f-894c-4917-b572-13fef358b388",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1bb8ed43-e434-4327-9c3f-04482a052c10",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Tools: Gene IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81623834-6400-4447-a915-d9b14626e021",
   "metadata": {},
   "source": [
    "Ensembl → BioMart → Ensembl Genes database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c7f936f4-7eba-46b7-bb60-40c7a2d54392",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yw/q_pg1y917pz9nzm99n51ctzh0000gp/T/ipykernel_1467/22589141.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  annot = pd.read_csv(DESKTOP + \"251224_biomart_export.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Source of gene name</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>Source of transcript name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>Gene Synonym</th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>HGNC ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-TF-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MTTF</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>HGNC:7481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-TF-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>TRNF</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>HGNC:7481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>1</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-RNR1-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>12S</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>HGNC:7470</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0  ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1  ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "2  ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "\n",
       "  Transcript stable ID version Chromosome/scaffold name  Gene start (bp)  \\\n",
       "0            ENST00000387314.1                       MT              577   \n",
       "1            ENST00000387314.1                       MT              577   \n",
       "2            ENST00000389680.2                       MT              648   \n",
       "\n",
       "   Gene end (bp)  Strand Gene name Source of gene name Transcript name  \\\n",
       "0            647       1     MT-TF         HGNC Symbol       MT-TF-201   \n",
       "1            647       1     MT-TF         HGNC Symbol       MT-TF-201   \n",
       "2           1601       1   MT-RNR1         HGNC Symbol     MT-RNR1-201   \n",
       "\n",
       "  Source of transcript name Gene type Gene Synonym HGNC symbol    HGNC ID  \n",
       "0           Transcript name   Mt_tRNA         MTTF       MT-TF  HGNC:7481  \n",
       "1           Transcript name   Mt_tRNA         TRNF       MT-TF  HGNC:7481  \n",
       "2           Transcript name   Mt_rRNA          12S     MT-RNR1  HGNC:7470  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot = pd.read_csv(DESKTOP + \"251224_biomart_export.tsv\", sep=\"\\t\")\n",
    "annot.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a46c5ae3-e303-47d7-94ad-a2359e68b23a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(86369, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000210049</th>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210077</th>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene name\n",
       "Gene stable ID           \n",
       "ENSG00000210049     MT-TF\n",
       "ENSG00000211459   MT-RNR1\n",
       "ENSG00000210077     MT-TV"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = annot[[\"Gene stable ID\", \"Gene name\"]].drop_duplicates().set_index(\"Gene stable ID\")\n",
    "print(gene_names.shape)\n",
    "gene_names.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "20e5623b-d11e-4deb-9ca5-8d2ce42d6026",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(41452, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000210049</th>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210077</th>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene name\n",
       "Gene stable ID           \n",
       "ENSG00000210049     MT-TF\n",
       "ENSG00000211459   MT-RNR1\n",
       "ENSG00000210077     MT-TV"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names2 = gene_names.drop_duplicates()\n",
    "print(gene_names2.shape)\n",
    "gene_names2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "24f91ebc-1610-4478-b498-a321e67e33de",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = gene_names2[\"Gene name\"].to_dict()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
