
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Filter samples &#8212; BULLKpy 0.0.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=f539c95a"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/pp/filter_samples';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Find bad Obs columns" href="find_bad_obs_cols_by_write.html" />
    <link rel="prev" title="Filter genes" href="filter_genes.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.0.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">BULLKpy 0.0.1 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../notebooks/index.html">
    Tutorials (notebooks)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malumbres/BULLKpy" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../bullkpy.html">bullkpy</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../io/index.html">Input Output</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../io/read_counts.html">Read counts</a></li>



<li class="toctree-l2"><a class="reference internal" href="../io/add_metadata.html">Add metadata</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">Pre-processing</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="batch_correct_combat.html">Batch correction with Combat</a></li>
<li class="toctree-l2"><a class="reference internal" href="filter_genes.html">Filter genes</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Filter samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="find_bad_obs_cols_by_write.html">Find bad Obs columns</a></li>
<li class="toctree-l2"><a class="reference internal" href="highly_variable_genes.html">Highly variable genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="log1p.html">log1p</a></li>
<li class="toctree-l2"><a class="reference internal" href="make_obs_h5ad_safe.html">Make Obs .h5ad safe</a></li>
<li class="toctree-l2"><a class="reference internal" href="normalize_cpm.html">Normalize cpm</a></li>
<li class="toctree-l2"><a class="reference internal" href="qc_metrics.html">QC metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="sanitize_metadata.html">Sanitize metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="set_raw_counts.html">Set raw counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="_is_integerish.html">Is integerish</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../tl/index.html">Tools</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../tl/adjusted_rand_index.html">Adjusted rand index</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/association.html">Association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/categorical_association.html">Categorical association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/cluster.html">Clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/cluster_metrics.html">Cluster metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/de.html">Differential expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/de_glm.html">Negative binomial GLM differential expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/gene_categorical_association.html">Gene-Category association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/gene_gene_correlations.html">Gene-gene correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/gsea_preranked.html">GSEA preranked</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/leiden_resolution_scan.html">Leiden resolution scan</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/list_enrichr_libraries.html">List enrichr libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/neighbors.html">Neighbors</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/obs_categorical_association.html">Obs-Category association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/obs_obs_corr_matrix.html">Obs-Obs correlation matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/pairwise_posthoc.html">Pairwise postdoc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/partial_corr.html">Partial correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/pca.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/pca_loadings.html">PCA loadings</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/plot_corr_heatmap.html">Plot Correlation heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/plot_corr_scatter.html">Plot Correlation scatter</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/posthoc_per_gene.html">Posthoc per gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/rank_genes_categorical.html">Rank genes categorical</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/rank_genes_groups.html">Rank genes groups</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/score_genes.html">Score genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/score_genes_cell_cycle.html">Score cell cycle genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/score_genes_dict.html">Score genes from a dictionary of genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/top_gene_gene_correlations.html">Top gene-gene correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/top_gene_obs_correlations.html">Top Gene-Obs correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/top_obs_obs_correlations.html">Top Obs-Obs correlations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/umap.html">UMAP embedding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tl/umap_graph.html">UMAP graph</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../pl/index.html">Plotting</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../pl/ari_resolution_heatmap.html">ARI resolution heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/association_heatmap.html">Association heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/boxplot_with_stats.html">Boxplot with statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/categorical_confusion.html">Categorical confusion</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/corr_heatmap.html">Correlation heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/corrplot_obs.html">Correlation plot Observations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/dotplot.html">Dotplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/dotplot_association.html">Dotplot association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gene_association.html">Gene Association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gene_association_volcano.html">Gene Association Volcano</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gene_plot.html">Gene Plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/genes_vs_mt_fraction.html">Genes vs. mitochondrial fraction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gsea_bubbleplot.html">GSEA bubbleplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/gsea_leading_edge_heatmap.html">GSEA Leading edge heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/heatmap_association.html">Heatmap association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/heatmap_de.html">Heatmap Differential Expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/leading_edge_jaccard_heatmap.html">Leading edge Jaccard heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/leading_edge_overlap_matrix.html">Leading edge overlap matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/library_size_vs_genes.html">Library size versus genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/ma.html">MA plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/mt_fraction_vs_counts.html">Mitochondrial fraction vs. counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/oncoprint.html">oncoprint</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/pca_loadings_bar.html">PCA loadings bar plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/pca_loadings_heatmap.html">PCA loadings heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/pca_scatter.html">PCA scatter plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/pca_variance_ratio.html">PCA variance ratio</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/qc_by_group.html">QC by group</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/qc_metrics.html">QC metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/qc_pairplot.html">QC pairplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/qc_scatter_panel.html">QC scatter panel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/rankplot.html">Rankplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/rankplot_association.html">Rankplot association</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/rank_genes_groups.html">Rank genes groups</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/rank_genes_groups_dotplot.html">Rank genes groups dotplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/sample_distances.html">Sample distances</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/umap.html">UMAP plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/violin.html">Violin plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/volcano.html">Volcano plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pl/volcano_categorical.html">Volcano for categorial comparisons</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">API reference</a></li>
    
    
    <li class="breadcrumb-item"><a href="index.html" class="nav-link">Pre-processing</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Filter samples</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="filter-samples">
<h1>Filter samples<a class="headerlink" href="#filter-samples" title="Link to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="bullkpy.pp.filter_samples">
<span class="sig-prename descclassname"><span class="pre">bullkpy.pp.</span></span><span class="sig-name descname"><span class="pre">filter_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_pct_mt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_pct_ribo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expr_threshold_for_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mt_var_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ribo_var_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/bullkpy/pp/filter.html#filter_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bullkpy.pp.filter_samples" title="Link to this definition">#</a></dt>
<dd><p>Filter samples using QC metrics in <cite>adata.obs</cite> if present; otherwise compute from matrix.</p>
<dl class="simple">
<dt>No hard requirement that any of these columns exist:</dt><dd><ul class="simple">
<li><p>total_counts</p></li>
<li><p>n_genes_detected</p></li>
<li><p>pct_counts_mt</p></li>
<li><p>pct_counts_ribo</p></li>
</ul>
</dd>
</dl>
<p>If a requested threshold is provided but cannot be computed, it will be skipped with a warning.</p>
</dd></dl>

<p>Filter samples based on QC thresholds (with graceful fallback computation).</p>
<p><code class="docutils literal notranslate"><span class="pre">bk.pp.filter_samples</span></code> removes low-quality samples using QC metrics stored in
<code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> <strong>if available</strong>, and <strong>computes missing metrics from the expression matrix</strong>
when needed (Scanpy-like behavior, but adapted for bulk RNA-seq).</p>
<p>A key design choice: <strong>there are no hard requirements</strong> that any QC columns exist.
If a threshold is requested but the corresponding metric cannot be computed, the
function <strong>skips that threshold with a warning</strong> instead of failing.</p>
<hr class="docutils" />
<section id="what-it-does">
<h2>What it does<a class="headerlink" href="#what-it-does" title="Link to this heading">#</a></h2>
<p>Given one or more thresholds, the function builds a boolean mask over samples and keeps
only samples that pass all requested filters.</p>
<p>Supported thresholds:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">min_counts</span></code> → keeps samples with <code class="docutils literal notranslate"><span class="pre">total_counts</span> <span class="pre">&gt;=</span> <span class="pre">min_counts</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">min_genes</span></code> → keeps samples with <code class="docutils literal notranslate"><span class="pre">n_genes_detected</span> <span class="pre">&gt;=</span> <span class="pre">min_genes</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max_pct_mt</span></code> → keeps samples with <code class="docutils literal notranslate"><span class="pre">pct_counts_mt</span> <span class="pre">&lt;=</span> <span class="pre">max_pct_mt</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max_pct_ribo</span></code> → keeps samples with <code class="docutils literal notranslate"><span class="pre">pct_counts_ribo</span> <span class="pre">&lt;=</span> <span class="pre">max_pct_ribo</span></code></p></li>
</ul>
<p>If these columns are missing in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>, the function attempts to compute them from
<code class="docutils literal notranslate"><span class="pre">adata.layers[layer]</span></code> (default <code class="docutils literal notranslate"><span class="pre">&quot;counts&quot;</span></code>) or <code class="docutils literal notranslate"><span class="pre">adata.X</span></code> if <code class="docutils literal notranslate"><span class="pre">layer=None</span></code>.</p>
<p>Computed metrics are <strong>stored back into <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code></strong> so they can be reused by other functions.</p>
</section>
<hr class="docutils" />
<section id="inputs-and-requirements">
<h2>Inputs and requirements<a class="headerlink" href="#inputs-and-requirements" title="Link to this heading">#</a></h2>
<section id="expression-matrix-source">
<h3>Expression matrix source<a class="headerlink" href="#expression-matrix-source" title="Link to this heading">#</a></h3>
<p>By default, metrics are computed from:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata.layers[&quot;counts&quot;]</span></code> (recommended for raw counts), otherwise</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">adata.X</span></code> if <code class="docutils literal notranslate"><span class="pre">layer=None</span></code></p></li>
</ul>
<p>If your data are already normalized or log-transformed, the filters still run,
but interpretation changes (see below).</p>
</section>
<section id="gene-masks-for-mt-and-ribosomal-genes">
<h3>Gene masks for MT and ribosomal genes.<a class="headerlink" href="#gene-masks-for-mt-and-ribosomal-genes" title="Link to this heading">#</a></h3>
<p>To compute mitochondrial/ribosomal percentages, the function needs to identify
mitochondrial and ribosomal genes. It does this using:</p>
<ul class="simple">
<li><p>provided masks in <code class="docutils literal notranslate"><span class="pre">adata.var</span></code> (recommended), e.g. <code class="docutils literal notranslate"><span class="pre">mt_var_key=&quot;mt&quot;</span></code></p></li>
<li><p>otherwise, heuristics based on gene names (implementation-dependent)</p></li>
</ul>
<p>If gene masks cannot be inferred, <code class="docutils literal notranslate"><span class="pre">%mt</span></code> / <code class="docutils literal notranslate"><span class="pre">%ribo</span></code> filters are skipped with a warning.</p>
</section>
</section>
<hr class="docutils" />
<section id="basic-usage">
<h2>Basic usage<a class="headerlink" href="#basic-usage" title="Link to this heading">#</a></h2>
<section id="filter-only-by-detected-genes">
<h3>Filter only by detected genes<a class="headerlink" href="#filter-only-by-detected-genes" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bk</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">min_genes</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="filter-by-multiple-qc-criteria">
<h3>Filter by multiple QC criteria<a class="headerlink" href="#filter-by-multiple-qc-criteria" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bk</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">min_counts</span><span class="o">=</span><span class="mf">1e6</span><span class="p">,</span>
    <span class="n">min_genes</span><span class="o">=</span><span class="mi">8000</span><span class="p">,</span>
    <span class="n">max_pct_mt</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">max_pct_ribo</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="using-precomputed-qc-metrics">
<h3>Using precomputed QC metrics<a class="headerlink" href="#using-precomputed-qc-metrics" title="Link to this heading">#</a></h3>
<p>If you already ran:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bk</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
<p>then filter_samples will reuse the columns in adata.obs instead of recomputing.</p>
<p>This is faster and guarantees consistency with your QC plots.</p>
</section>
<section id="working-with-non-count-data">
<h3>Working with non-count data<a class="headerlink" href="#working-with-non-count-data" title="Link to this heading">#</a></h3>
<p>If your matrix is already log-transformed (e.g. “log1p_cpm”), you can still filter,
but some thresholds become less meaningful:
•	min_genes remains meaningful (counts non-zero genes above threshold)
•	min_counts depends on the scale of the data (sum of transformed values)
•	%mt and %ribo become relative fractions of transformed values</p>
<p>For normalized/log data, filtering is best done using relative thresholds
or distribution inspection (e.g. percentile cutoffs).</p>
</section>
<section id="custom-definition-of-detected-genes">
<h3>Custom definition of “detected genes”<a class="headerlink" href="#custom-definition-of-detected-genes" title="Link to this heading">#</a></h3>
<p>The number of detected genes is computed as:</p>
<p>number of genes with expression &gt; expr_threshold_for_genes</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bk</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">min_genes</span><span class="o">=</span><span class="mi">6000</span><span class="p">,</span>
    <span class="n">expr_threshold_for_genes</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;log1p_cpm&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="in-place-vs-copy">
<h3>In-place vs copy<a class="headerlink" href="#in-place-vs-copy" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># By default, filtering is done in-place (faster):</span>
<span class="n">bk</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># To return a new AnnData without modifying the original:</span>
<span class="n">adata_filt</span> <span class="o">=</span> <span class="n">bk</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

</pre></div>
</div>
</section>
<section id="stored-metadata">
<h3>Stored metadata<a class="headerlink" href="#stored-metadata" title="Link to this heading">#</a></h3>
<p>This function records filter parameters and before/after counts in:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;pp&quot;</span><span class="p">][</span><span class="s2">&quot;filter_samples&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>This is useful for reproducibility and reporting.</p>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="filter_genes.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Filter genes</p>
      </div>
    </a>
    <a class="right-next"
       href="find_bad_obs_cols_by_write.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Find bad Obs columns</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bullkpy.pp.filter_samples"><code class="docutils literal notranslate"><span class="pre">filter_samples()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-it-does">What it does</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inputs-and-requirements">Inputs and requirements</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-matrix-source">Expression matrix source</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-masks-for-mt-and-ribosomal-genes">Gene masks for MT and ribosomal genes.</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-usage">Basic usage</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#filter-only-by-detected-genes">Filter only by detected genes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#filter-by-multiple-qc-criteria">Filter by multiple QC criteria</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#using-precomputed-qc-metrics">Using precomputed QC metrics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#working-with-non-count-data">Working with non-count data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-definition-of-detected-genes">Custom definition of “detected genes”</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#in-place-vs-copy">In-place vs copy</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#stored-metadata">Stored metadata</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2026, Marcos Malumbres.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>